// @vue/runtime-vapor/dist/runtime-vapor.esm-bundler.js
/**
  * @vue/runtime-vapor v3.6.0-beta.10
  * (c) 2018-present Yuxi (Evan) You and Vue contributors
  * @license MIT
  **/
import { EffectScope, Fragment, NULL_DYNAMIC_COMPONENT, Static, TransitionPropsValidators, VueElementBase, activate, baseApplyTranslation, baseEmit, baseNormalizePropsOptions, baseResolveTransitionHooks, baseUseCssVars, callPendingCbs, callWithAsyncErrorHandling, callWithErrorHandling, checkTransitionMode, createAppAPI, createAsyncComponentContext, createCanSetSetupRefChecker, createInternalObject, createVNode, currentInstance, currentInstance as currentInstance$1, deactivate, devtoolsComponentAdded, endMeasure, ensureHydrationRenderer, ensureRenderer, ensureVaporSlotFallback, expose, flushOnAppMount, forceReflow, getAttributeMismatch, getComponentName, getFunctionalFallthrough, getInheritedScopeIds, getTransitionRawChildren, handleError, handleMovedChildren, hasCSSTransform, initFeatureFlags, invalidateMount, invokeDirectiveHook, isAsyncWrapper, isEmitListener, isHydrating, isKeepAlive, isMapEqual, isMismatchAllowed, isRef, isSetEqual, isTeleportDeferred, isTeleportDisabled, isTemplateNode, isTemplateRefKey, isVNode, isValidHtmlOrSvgAttribute, knownTemplateRefs, leaveCbKey, markAsyncBoundary, matches, mergeProps, nextUid, normalizeContainer, normalizeRef, onBeforeMount, onBeforeUnmount, onBeforeUpdate, onMounted, onScopeDispose, onUpdated, patchStyle, performAsyncHydrate, performTransitionEnter, performTransitionLeave, popWarningContext, pushWarningContext, queueJob, queuePostFlushCb, registerHMR, renderSlot, resetShapeFlag, resolveDynamicComponent, resolvePropValue, resolveTeleportTarget, resolveTransitionProps, setCurrentInstance, setCurrentRenderingInstance, setDevtoolsHook, setRef, setTransitionHooks, setVarsOnNode, shallowReactive, shallowRef, shouldSetAsProp, shouldSetAsPropForVueCE, shouldUpdateComponent, simpleSetCurrentInstance, startMeasure, toClassSet, toStyleMap, unregisterHMR, unsafeToTrustedHTML, useAsyncComponentState, useTransitionState, vModelCheckboxInit, vModelCheckboxUpdate, vModelGetValue, vModelSelectInit, vModelSetSelected, vModelTextInit, vModelTextUpdate, vShowHidden, vShowOriginalDisplay, validateProps, warn, warnExtraneousAttributes, warnPropMismatch, watch, withAsyncContext as withAsyncContext$1, xlinkNS } from "@vue/runtime-dom";
import { EMPTY_ARR, EMPTY_OBJ, NO, NOOP, camelize, canSetValueDirectly, extend, getEscapedCssVarName, getGlobalThis, hasOwn, includeBooleanAttr, invokeArrayFns, isArray, isFunction, isObject, isOn, isPlainObject, isPromise, isReservedProp, isString, looseEqual, normalizeClass, normalizeCssVarValue, normalizeStyle, parseStringStyle, remove as remove$1, stringifyStyle, toDisplayString } from "@vue/shared";
import { EffectScope as EffectScope$1, ReactiveEffect, computed, effectScope, isReactive, isReadonly, isRef as isRef$1, isShallow, markRaw, onEffectCleanup, onScopeDispose as onScopeDispose$1, pauseTracking, proxyRefs, resetTracking, setActiveSub, shallowReadArray, shallowRef as shallowRef$1, toRaw, toReactive, toReadonly, traverse, unref, watch as watch$1 } from "@vue/reactivity";
//#region packages/runtime-vapor/src/insertionState.ts
let insertionParent;
let insertionAnchor;
let insertionIndex;
let isLastInsertion;
/**
* This function is called before a block type that requires insertion
* (component, slot outlet, if, for) is created. The state is used for actual
* insertion on client-side render, and used for node adoption during hydration.
*/
function setInsertionState(parent, anchor, logicalIndex, last) {
	insertionParent = parent;
	isLastInsertion = last;
	insertionIndex = logicalIndex;
	if (anchor !== void 0) if (isHydrating$1) insertionAnchor = void 0;
	else {
		insertionAnchor = anchor;
		if (anchor === 0 && !parent.$fc) parent.$fc = parent.firstChild;
	}
	else insertionAnchor = void 0;
}
function resetInsertionState() {
	insertionParent = insertionAnchor = insertionIndex = isLastInsertion = void 0;
}
//#endregion
//#region packages/runtime-vapor/src/dom/hydration.ts
let isHydratingEnabled = false;
function setIsHydratingEnabled(value) {
	isHydratingEnabled = value;
}
let currentHydrationNode = null;
let isHydrating$1 = false;
function setIsHydrating(value) {
	if (!isHydratingEnabled && !isHydrating) return false;
	try {
		return isHydrating$1;
	} finally {
		isHydrating$1 = value;
	}
}
function runWithoutHydration(fn) {
	const prev = setIsHydrating(false);
	try {
		return fn();
	} finally {
		setIsHydrating(prev);
	}
}
let isOptimized$1 = false;
function performHydration(fn, setup, cleanup) {
	if (!isOptimized$1) {
		adoptTemplate = adoptTemplateImpl;
		locateHydrationNode = locateHydrationNodeImpl;
		Comment.prototype.$fe = void 0;
		Node.prototype.$idx = void 0;
		Node.prototype.$llc = void 0;
		Node.prototype.$vha = void 0;
		isOptimized$1 = true;
	}
	const prev = setIsHydrating(true);
	const prevHydrationNode = currentHydrationNode;
	currentHydrationNode = null;
	try {
		setup();
		return fn();
	} finally {
		cleanup();
		currentHydrationNode = prevHydrationNode;
		setIsHydrating(prev);
	}
}
function withHydration(container, fn) {
	const setup = () => setInsertionState(container);
	const cleanup = () => resetInsertionState();
	return performHydration(fn, setup, cleanup);
}
function hydrateNode(node, fn) {
	const setup = () => currentHydrationNode = node;
	const cleanup = () => {};
	return performHydration(fn, setup, cleanup);
}
function enterHydration(node) {
	const prevHydrationEnabled = isHydratingEnabled;
	if (!prevHydrationEnabled) setIsHydratingEnabled(true);
	const prev = setIsHydrating(true);
	const prevHydrationNode = currentHydrationNode;
	currentHydrationNode = node;
	return () => {
		currentHydrationNode = prevHydrationNode;
		setIsHydrating(prev);
		if (!prevHydrationEnabled) setIsHydratingEnabled(false);
	};
}
let adoptTemplate;
let locateHydrationNode;
const isComment = (node, data) => node.nodeType === 8 && node.data === data;
function setCurrentHydrationNode(node) {
	currentHydrationNode = node;
}
/* @__NO_SIDE_EFFECTS__ */
function locateNextSiblingOfParent(n) {
	if (!n.parentNode) return null;
	return /* @__PURE__ */ _next(n.parentNode) || /* @__PURE__ */ locateNextSiblingOfParent(n.parentNode);
}
function advanceHydrationNode(node) {
	const ret = /* @__PURE__ */ _next(node) || /* @__PURE__ */ locateNextSiblingOfParent(node);
	if (ret) setCurrentHydrationNode(ret);
}
/**
* Locate the first non-fragment-comment node and locate the next node
* while handling potential fragments.
*/
function adoptTemplateImpl(node, template) {
	if (!(template[0] === "<" && template[1] === "!")) {
		if (template.trim() === "" && isComment(node, "]") && isComment(node.previousSibling, "[")) node.before(node = /* @__PURE__ */ createTextNode());
		node = resolveHydrationTarget(node);
	}
	const type = node.nodeType;
	if (type === 8 && !template.startsWith("<!") || type === 1 && !template.startsWith(`<` + node.tagName.toLowerCase())) node = handleMismatch(node, template);
	advanceHydrationNode(node);
	return node;
}
function locateNextNode(node) {
	return isComment(node, "[") ? /* @__PURE__ */ _next(locateEndAnchor(node)) : isComment(node, "teleport start") ? /* @__PURE__ */ _next(locateEndAnchor(node, "teleport start", "teleport end")) : /* @__PURE__ */ _next(node);
}
function locateHydrationNodeImpl(consumeFragmentStart = false) {
	let node;
	if (insertionIndex !== void 0) node = locateChildByLogicalIndex(insertionParent, insertionIndex);
	else if (insertionParent) node = insertionParent.firstChild;
	else node = currentHydrationNode;
	if (consumeFragmentStart && node && isComment(node, "[")) node = node.nextSibling;
	if (!!(process.env.NODE_ENV !== "production") && !node) throw new Error("No current hydration node was found.\nthis is likely a Vue internal bug.");
	resetInsertionState();
	currentHydrationNode = node;
}
function locateEndAnchor(node, open = "[", close = "]") {
	if (node.$fe) return node.$fe;
	const stack = [node];
	while ((node = /* @__PURE__ */ _next(node)) && stack.length > 0) if (node.nodeType === 8) {
		if (node.data === open) stack.push(node);
		else if (node.data === close) {
			const matchingOpen = stack.pop();
			matchingOpen.$fe = node;
			if (stack.length === 0) return node;
		}
	}
	return null;
}
function locateHydrationBoundaryClose(node, closeHint = null) {
	let close = closeHint;
	if (!close || !isComment(close, "]")) if (isComment(node, "]")) close = node;
	else {
		let candidate = locateNextNode(node);
		while (candidate && !isComment(candidate, "]")) candidate = locateNextNode(candidate);
		close = candidate;
	}
	if (!close) return node;
	return close;
}
function handleMismatch(node, template) {
	if (!isMismatchAllowed(node.parentElement, 1)) {
		(process.env.NODE_ENV !== "production" || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && warn(`Hydration node mismatch:\n- rendered on server:`, node, node.nodeType === 3 ? `(text)` : isComment(node, "[[") ? `(start of block node)` : ``, `\n- expected on client:`, template);
		logMismatchError();
	}
	if (isComment(node, "[")) removeFragmentNodes(node);
	const shouldPreserveAnchor = isHydrationAnchor(node);
	const container = /* @__PURE__ */ parentNode(node);
	const next = shouldPreserveAnchor ? node : /* @__PURE__ */ _next(node);
	if (!shouldPreserveAnchor) remove(node, container);
	if (template[0] !== "<") return container.insertBefore(/* @__PURE__ */ createTextNode(template), next);
	const t = /* @__PURE__ */ createElement("template");
	t.innerHTML = template;
	const newNode = (/* @__PURE__ */ _child(t.content)).cloneNode(true);
	if (node.nodeType === 1) {
		newNode.innerHTML = node.innerHTML;
		Array.from(node.attributes).forEach((attr) => {
			newNode.setAttribute(attr.name, attr.value);
		});
	}
	container.insertBefore(newNode, next);
	return newNode;
}
let hasLoggedMismatchError = false;
const logMismatchError = () => {
	if (hasLoggedMismatchError) return;
	console.error("Hydration completed but contains mismatches.");
	hasLoggedMismatchError = true;
};
function removeFragmentNodes(node, endAnchor) {
	const parent = /* @__PURE__ */ parentNode(node);
	if (!parent) return;
	const end = endAnchor || locateEndAnchor(node);
	while (true) {
		const next = /* @__PURE__ */ _next(node);
		if (next && next !== end) remove(next, parent);
		else break;
	}
}
function removeHydrationNode(node, close = null) {
	const parent = /* @__PURE__ */ parentNode(node);
	if (!parent) return;
	if (isComment(node, "[")) {
		const end = locateEndAnchor(node);
		removeFragmentNodes(node, end || void 0);
		const endParent = end && /* @__PURE__ */ parentNode(end);
		if (end && end !== close && endParent) remove(end, endParent);
	} else if (isComment(node, "teleport start")) {
		const end = locateEndAnchor(node, "teleport start", "teleport end");
		removeFragmentNodes(node, end || void 0);
		const endParent = end && /* @__PURE__ */ parentNode(end);
		if (end && end !== close && endParent) remove(end, endParent);
	}
	remove(node, parent);
}
function cleanupHydrationTail(node) {
	const container = node.parentElement;
	if (container) warnHydrationChildrenMismatch(container);
	removeHydrationNode(node);
}
function markHydrationAnchor(node) {
	node.$vha = 1;
	return node;
}
function isHydrationAnchor(node) {
	return !!node && node.$vha === 1;
}
function resolveHydrationTarget(node) {
	while (true) {
		if (isHydrationAnchor(node)) return node;
		if (node.nodeType === 8 && (node.data === "[" || node.data === "]" || node.data === "teleport start" || node.data === "teleport end")) {
			const next = node.nextSibling;
			if (!next) return node;
			node = next;
			continue;
		}
		return node;
	}
}
function finalizeHydrationBoundary(close) {
	let node = currentHydrationNode;
	if (!close || !node || node === close) return;
	let cur = node;
	let hasRemovableNode = false;
	while (cur && cur !== close) {
		if (!isHydrationAnchor(cur)) hasRemovableNode = true;
		cur = locateNextNode(cur);
	}
	if (!cur) return;
	if (!hasRemovableNode) {
		setCurrentHydrationNode(close);
		return;
	}
	warnHydrationChildrenMismatch(close.parentElement);
	while (node && node !== close) {
		const next = locateNextNode(node);
		if (!isHydrationAnchor(node)) removeHydrationNode(node, close);
		node = next;
	}
	setCurrentHydrationNode(close);
}
function warnHydrationChildrenMismatch(container) {
	if (container && !isMismatchAllowed(container, 1)) {
		(process.env.NODE_ENV !== "production" || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && warn(`Hydration children mismatch on`, container, `\nServer rendered element contains more child nodes than client nodes.`);
		logMismatchError();
	}
}
function enterHydrationBoundary(close) {
	return () => {
		finalizeHydrationBoundary(close);
	};
}
//#endregion
//#region packages/runtime-vapor/src/dom/node.ts
/* @__NO_SIDE_EFFECTS__ */
function createElement(tagName) {
	return document.createElement(tagName);
}
/* @__NO_SIDE_EFFECTS__ */
function createTextNode(value = "") {
	return document.createTextNode(value);
}
/* @__NO_SIDE_EFFECTS__ */
function createComment(data) {
	return document.createComment(data);
}
/* @__NO_SIDE_EFFECTS__ */
function querySelector(selectors) {
	return document.querySelector(selectors);
}
/* @__NO_SIDE_EFFECTS__ */
function parentNode(node) {
	return node.parentNode;
}
/* @__NO_SIDE_EFFECTS__ */
function txt(node) {
	if (isHydrating$1) {
		let n = /* @__PURE__ */ _child(node);
		if (!n) return node.appendChild(/* @__PURE__ */ createTextNode());
		return n;
	}
	return /* @__PURE__ */ _child(node);
}
/* @__NO_SIDE_EFFECTS__ */
function child(node, logicalIndex) {
	if (isHydrating$1) return locateChildByLogicalIndex(node, logicalIndex !== null && logicalIndex !== void 0 ? logicalIndex : 0);
	return /* @__PURE__ */ _child(node);
}
/* @__NO_SIDE_EFFECTS__ */
function nthChild(node, i) {
	if (isHydrating$1) return locateChildByLogicalIndex(node, i);
	return node.childNodes[i];
}
/* @__NO_SIDE_EFFECTS__ */
function next(node, logicalIndex) {
	if (isHydrating$1) return locateChildByLogicalIndex(node.parentNode, logicalIndex);
	return /* @__PURE__ */ _next(node);
}
/* @__NO_SIDE_EFFECTS__ */
function _child(node) {
	return node.firstChild;
}
/* @__NO_SIDE_EFFECTS__ */
function _next(node) {
	return node.nextSibling;
}
function locateChildByLogicalIndex(parent, logicalIndex) {
	let child = parent.$llc || parent.firstChild;
	let fromIndex = child.$idx || 0;
	if (logicalIndex < fromIndex) {
		child = parent.firstChild;
		fromIndex = 0;
	}
	while (child) {
		if (fromIndex === logicalIndex) {
			child.$idx = logicalIndex;
			return parent.$llc = child;
		}
		child = isComment(child, "[") ? locateEndAnchor(child).nextSibling : child.nextSibling;
		fromIndex++;
	}
	return null;
}
//#endregion
//#region packages/runtime-vapor/src/dom/event.ts
function addEventListener(el, event, handler, options) {
	el.addEventListener(event, handler, options);
	return () => el.removeEventListener(event, handler, options);
}
function on(el, event, handler, options = {}) {
	if (isArray(handler)) handler.forEach((fn) => on(el, event, fn, options));
	else {
		if (!handler) return;
		addEventListener(el, event, handler, options);
		if (options.effect) onEffectCleanup(() => {
			el.removeEventListener(event, handler, options);
		});
	}
}
function delegate(el, event, handler) {
	const key = `$evt${event}`;
	const existing = el[key];
	if (existing) if (isArray(existing)) existing.push(handler);
	else el[key] = [existing, handler];
	else el[key] = handler;
}
/**
* Event delegation borrowed from solid
*/
const delegatedEvents = /* @__PURE__ */ Object.create(null);
const delegateEvents = (...names) => {
	for (const name of names) if (!delegatedEvents[name]) {
		delegatedEvents[name] = true;
		document.addEventListener(name, delegatedEventHandler);
	}
};
const delegatedEventHandler = (e) => {
	let node = e.composedPath && e.composedPath()[0] || e.target;
	if (e.target !== node) Object.defineProperty(e, "target", {
		configurable: true,
		value: node
	});
	Object.defineProperty(e, "currentTarget", {
		configurable: true,
		get() {
			return node || document;
		}
	});
	while (node !== null) {
		const handlers = node[`$evt${e.type}`];
		if (handlers) if (isArray(handlers)) {
			for (const handler of handlers) if (!node.disabled) {
				handler(e);
				if (e.cancelBubble) return;
			}
		} else {
			handlers(e);
			if (e.cancelBubble) return;
		}
		node = node.host && node.host !== node && node.host instanceof Node ? node.host : node.parentNode;
	}
};
function setDynamicEvents(el, events) {
	for (const name in events) on(el, name, events[name], { effect: true });
}
function createInvoker(handler) {
	const i = currentInstance;
	return (...args) => callWithAsyncErrorHandling(handler, i, 5, args);
}
//#endregion
//#region packages/runtime-vapor/src/dom/prop.ts
const hasFallthroughKey = (key) => currentInstance.hasFallthrough && key in currentInstance.attrs;
function setProp(el, key, value) {
	if (key in el) setDOMProp(el, key, value);
	else setAttr(el, key, value);
}
function setAttr(el, key, value, isSVG = false) {
	if (!isApplyingFallthroughProps && el.$root && hasFallthroughKey(key)) return;
	if (key === "true-value") el._trueValue = value;
	else if (key === "false-value") el._falseValue = value;
	if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1 && !attributeHasMismatch(el, key, value)) {
		el[`$${key}`] = value;
		return;
	}
	if (value !== el[`$${key}`]) {
		el[`$${key}`] = value;
		if (isSVG && key.startsWith("xlink:")) if (value != null) el.setAttributeNS(xlinkNS, key, value);
		else el.removeAttributeNS(xlinkNS, key.slice(6, key.length));
		else if (value != null) el.setAttribute(key, value);
		else el.removeAttribute(key);
	}
}
function setDOMProp(el, key, value, forceHydrate = false, attrName) {
	if (!isApplyingFallthroughProps && el.$root && hasFallthroughKey(key)) return;
	if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1 && !attributeHasMismatch(el, key, value) && !shouldForceHydrate(el, key) && !forceHydrate) return;
	const prev = el[key];
	if (value === prev) return;
	let needRemove = false;
	if (value === "" || value == null) {
		const type = typeof prev;
		if (type === "boolean") value = includeBooleanAttr(value);
		else if (value == null && type === "string") {
			value = "";
			needRemove = true;
		} else if (type === "number") {
			value = 0;
			needRemove = true;
		}
	}
	try {
		el[key] = value;
	} catch (e) {
		if (!!(process.env.NODE_ENV !== "production") && !needRemove) warn(`Failed setting prop "${key}" on <${el.tagName.toLowerCase()}>: value ${value} is invalid.`, e);
	}
	needRemove && el.removeAttribute(attrName || key);
}
function setClass(el, value, isSVG = false) {
	if (el.$root) setClassIncremental(el, value);
	else {
		value = normalizeClass(value);
		if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1 && !classHasMismatch(el, value, false)) {
			el.$cls = value;
			return;
		}
		if (value !== el.$cls) if (isSVG) el.setAttribute("class", el.$cls = value);
		else el.className = el.$cls = value;
	}
}
function setClassIncremental(el, value) {
	const cacheKey = `$clsi${isApplyingFallthroughProps ? "$" : ""}`;
	const normalizedValue = normalizeClass(value);
	if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1 && !classHasMismatch(el, normalizedValue, true)) {
		el[cacheKey] = normalizedValue;
		return;
	}
	const prev = el[cacheKey];
	if ((value = el[cacheKey] = normalizedValue) !== prev) {
		const nextList = value.split(/\s+/);
		if (value) el.classList.add(...nextList);
		if (prev) {
			for (const cls of prev.split(/\s+/)) if (!nextList.includes(cls)) el.classList.remove(cls);
		}
	}
}
/**
* dev only
* defer style matching checks until hydration completes (instance.block is set) if
* the component uses style v-bind or the element contains CSS variables, to correctly
* verify if the element is the component root.
*/
function shouldDeferCheckStyleMismatch(el) {
	return !!(process.env.NODE_ENV !== "production") && (!!currentInstance.getCssVars || Object.values(el.style).some((v) => v.startsWith("--")));
}
function setStyle(el, value) {
	if (el.$root) setStyleIncremental(el, value);
	else {
		const normalizedValue = normalizeStyle(value);
		if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1) {
			if (shouldDeferCheckStyleMismatch(el)) {
				const instance = currentInstance;
				queuePostFlushCb(() => {
					if (!styleHasMismatch(el, value, normalizedValue, false, instance)) {
						el.$sty = normalizedValue;
						return;
					}
					patchStyle(el, el.$sty, el.$sty = normalizedValue);
				});
				return;
			} else if (!styleHasMismatch(el, value, normalizedValue, false)) {
				el.$sty = normalizedValue;
				return;
			}
		}
		patchStyle(el, el.$sty, el.$sty = normalizedValue);
	}
}
function setStyleIncremental(el, value) {
	const cacheKey = `$styi${isApplyingFallthroughProps ? "$" : ""}`;
	const normalizedValue = isString(value) ? parseStringStyle(value) : normalizeStyle(value);
	if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1) {
		if (shouldDeferCheckStyleMismatch(el)) {
			const instance = currentInstance;
			queuePostFlushCb(() => {
				if (!styleHasMismatch(el, value, normalizedValue, true, instance)) {
					el[cacheKey] = normalizedValue;
					return;
				}
				patchStyle(el, el[cacheKey], el[cacheKey] = normalizedValue);
			});
			return;
		} else if (!styleHasMismatch(el, value, normalizedValue, true)) {
			el[cacheKey] = normalizedValue;
			return;
		}
	}
	patchStyle(el, el[cacheKey], el[cacheKey] = normalizedValue);
}
function setValue(el, value, forceHydrate = false) {
	if (!isApplyingFallthroughProps && el.$root && hasFallthroughKey("value")) return;
	el._value = value;
	if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1 && !attributeHasMismatch(el, "value", getClientText(el, value)) && !shouldForceHydrate(el, "value") && !forceHydrate) return;
	const oldValue = el.tagName === "OPTION" ? el.getAttribute("value") : el.value;
	const newValue = value == null ? "" : value;
	if (oldValue !== newValue) el.value = newValue;
	if (value == null) el.removeAttribute("value");
}
/**
* Only called on text nodes!
* Compiler should also ensure value passed here is already converted by
* `toDisplayString`
*/
function setText(el, value) {
	if (isHydrating$1) {
		const clientText = getClientText(el.parentNode, value);
		if (el.nodeValue == clientText) {
			el.$txt = clientText;
			return;
		}
		const parent = el.parentElement;
		if (parent && !isMismatchAllowed(parent, 0)) {
			(process.env.NODE_ENV !== "production" || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && warn(`Hydration text mismatch in`, el.parentNode, `\n  - rendered on server: ${JSON.stringify(el.data)}\n  - expected on client: ${JSON.stringify(value)}`);
			logMismatchError();
		}
	}
	if (el.$txt !== value) el.nodeValue = el.$txt = value;
}
/**
* Used by setDynamicProps only, so need to guard with `toDisplayString`
*/
function setElementText(el, value) {
	value = toDisplayString(value);
	if (isHydrating$1) {
		let clientText = getClientText(el, value);
		if (el.textContent === clientText) {
			el.$txt = clientText;
			return;
		}
		if (!isMismatchAllowed(el, 0)) {
			(process.env.NODE_ENV !== "production" || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && warn(`Hydration text content mismatch on`, el, `\n  - rendered on server: ${el.textContent}\n  - expected on client: ${clientText}`);
			logMismatchError();
		}
	}
	if (el.$txt !== value) el.textContent = el.$txt = value;
}
function setBlockText(block, value) {
	value = value == null ? "" : value;
	if (block.$txt !== value) setTextToBlock(block, block.$txt = value);
}
/**
* dev only
*/
function warnCannotSetProp(prop) {
	warn(`Extraneous non-props attributes (${prop}) were passed to component but could not be automatically inherited because component renders text or multiple root nodes.`);
}
function setTextToBlock(block, value) {
	if (block instanceof Node) {
		if (block instanceof Element) block.textContent = value;
		else if (!!(process.env.NODE_ENV !== "production")) warnCannotSetProp("textContent");
	} else if (isVaporComponent(block)) setTextToBlock(block.block, value);
	else if (isArray(block)) {
		if (!!(process.env.NODE_ENV !== "production")) warnCannotSetProp("textContent");
	} else setTextToBlock(block.nodes, value);
}
function setHtml(el, value) {
	value = value == null ? "" : unsafeToTrustedHTML(value);
	if (el.$html !== value) el.innerHTML = el.$html = value;
}
function setBlockHtml(block, value) {
	value = value == null ? "" : unsafeToTrustedHTML(value);
	if (block.$html !== value) setHtmlToBlock(block, block.$html = value);
}
function setHtmlToBlock(block, value) {
	if (block instanceof Node) {
		if (block instanceof Element) block.innerHTML = value;
		else if (!!(process.env.NODE_ENV !== "production")) warnCannotSetProp("innerHTML");
	} else if (isVaporComponent(block)) setHtmlToBlock(block.block, value);
	else if (isArray(block)) {
		if (!!(process.env.NODE_ENV !== "production")) warnCannotSetProp("innerHTML");
	} else setHtmlToBlock(block.nodes, value);
}
function setDynamicProps(el, args, isSVG) {
	const props = args.length > 1 ? mergeProps(...args) : args[0];
	const cacheKey = `$dprops${isApplyingFallthroughProps ? "$" : ""}`;
	const prevKeys = el[cacheKey];
	if (prevKeys) {
		for (const key of prevKeys) if (!(key in props)) setDynamicProp(el, key, null, isSVG);
	}
	for (const key of el[cacheKey] = Object.keys(props)) setDynamicProp(el, key, props[key], isSVG);
}
/**
* @internal
*/
function setDynamicProp(el, key, value, isSVG = false) {
	let forceHydrate = false;
	if (key === "class") setClass(el, value, isSVG);
	else if (key === "style") setStyle(el, value);
	else if (isOn(key)) on(el, key[2].toLowerCase() + key.slice(3), value, { effect: true });
	else if ((forceHydrate = key[0] === ".") ? (key = key.slice(1), true) : key[0] === "^" ? (key = key.slice(1), false) : shouldSetAsProp(el, key, value, isSVG)) if (key === "innerHTML") setHtml(el, value);
	else if (key === "textContent") setElementText(el, value);
	else if (key === "value" && canSetValueDirectly(el.tagName)) setValue(el, value, forceHydrate);
	else setDOMProp(el, key, value, forceHydrate);
	else if (el._isVueCE && (shouldSetAsPropForVueCE(el, key) || el._def.__asyncLoader && (/[A-Z]/.test(key) || !isString(value)))) setDOMProp(el, camelize(key), value, forceHydrate, key);
	else setAttr(el, key, value, isSVG);
	return value;
}
let isOptimized = false;
/**
* Optimize property lookup for cache properties on Element and Text nodes
*/
function optimizePropertyLookup() {
	if (isOptimized) return;
	isOptimized = true;
	const proto = Element.prototype;
	proto.$transition = void 0;
	proto.$key = void 0;
	proto.$fc = proto.$evtclick = void 0;
	proto.$root = false;
	proto.$html = proto.$cls = proto.$sty = "";
	Text.prototype.$txt = void 0;
}
function classHasMismatch(el, expected, isIncremental) {
	const actual = el.getAttribute("class");
	const actualClassSet = toClassSet(actual || "");
	const expectedClassSet = toClassSet(expected);
	let hasMismatch = false;
	if (isIncremental) {
		if (expected) hasMismatch = Array.from(expectedClassSet).some((cls) => !actualClassSet.has(cls));
	} else hasMismatch = !isSetEqual(actualClassSet, expectedClassSet);
	if (hasMismatch) {
		warnPropMismatch(el, "class", 2, actual, expected);
		logMismatchError();
		return true;
	}
	return false;
}
function styleHasMismatch(el, value, normalizedValue, isIncremental, instance = currentInstance) {
	const actual = el.getAttribute("style");
	const actualStyleMap = toStyleMap(actual || "");
	const expected = isString(value) ? value : stringifyStyle(normalizedValue);
	const expectedStyleMap = toStyleMap(expected);
	if (el[vShowHidden]) expectedStyleMap.set("display", "none");
	if (instance) resolveCssVars(instance, el, expectedStyleMap);
	let hasMismatch = false;
	if (isIncremental) {
		if (expected) hasMismatch = Array.from(expectedStyleMap.entries()).some(([key, val]) => actualStyleMap.get(key) !== val);
	} else hasMismatch = !isMapEqual(actualStyleMap, expectedStyleMap);
	if (hasMismatch) {
		warnPropMismatch(el, "style", 3, actual, expected);
		logMismatchError();
		return true;
	}
	return false;
}
/**
* dev only
*/
function resolveCssVars(instance, block, expectedMap) {
	if (!instance.isMounted) return;
	const rootBlocks = normalizeBlock(instance);
	if (instance.getCssVars && normalizeBlock(block).every((b) => rootBlocks.includes(b))) {
		const cssVars = instance.getCssVars();
		for (const key in cssVars) {
			const value = normalizeCssVarValue(cssVars[key]);
			expectedMap.set(`--${getEscapedCssVarName(key, false)}`, value);
		}
	}
	if (normalizeBlock(block).every((b) => rootBlocks.includes(b)) && instance.parent) resolveCssVars(instance.parent, instance.block, expectedMap);
}
function attributeHasMismatch(el, key, value) {
	if (isValidHtmlOrSvgAttribute(el, key)) {
		const { actual, expected } = getAttributeMismatch(el, key, value);
		if (actual !== expected) {
			warnPropMismatch(el, key, 4, actual, expected);
			logMismatchError();
			return true;
		}
	}
	return false;
}
function getClientText(el, value) {
	if (value[0] === "\n" && (el.tagName === "PRE" || el.tagName === "TEXTAREA")) value = value.slice(1);
	return value;
}
function shouldForceHydrate(el, key) {
	const { tagName } = el;
	return (tagName === "INPUT" || tagName === "OPTION") && (key.endsWith("value") || key === "indeterminate") || tagName.includes("-");
}
//#endregion
//#region packages/runtime-vapor/src/transition.ts
let _applyTransitionHooks;
let _applyTransitionLeaveHooks;
function registerTransitionHooks(applyHooks, applyLeaveHooks) {
	_applyTransitionHooks = applyHooks;
	_applyTransitionLeaveHooks = applyLeaveHooks;
}
function applyTransitionHooks(block, hooks) {
	return _applyTransitionHooks ? _applyTransitionHooks(block, hooks) : hooks;
}
function applyTransitionLeaveHooks(block, enterHooks, afterLeaveCb) {
	_applyTransitionLeaveHooks && _applyTransitionLeaveHooks(block, enterHooks, afterLeaveCb);
}
//#endregion
//#region packages/runtime-vapor/src/directives/vShow.ts
let currentPendingVShows = null;
function setCurrentPendingVShows(pending) {
	try {
		return currentPendingVShows;
	} finally {
		currentPendingVShows = pending;
	}
}
function applyVShow(target, source) {
	if (isVaporComponent(target)) return applyVShow(target.block, source);
	if (isArray(target) && target.length === 1) return applyVShow(target[0], source);
	if (target instanceof DynamicFragment) {
		const update = target.update;
		target.update = (render, key) => {
			update.call(target, render, key);
			setDisplay(target, source());
		};
	} else if (target instanceof VaporFragment && target.insert) {
		const insert = target.insert;
		target.insert = (parent, anchor) => {
			insert.call(target, parent, anchor);
			setDisplay(target, source());
		};
	}
	renderEffect(() => {
		const value = source();
		if (currentPendingVShows) {
			currentPendingVShows.push({
				target,
				setDisplay: () => setDisplay(target, value)
			});
			return;
		}
		setDisplay(target, value);
	});
}
function setDisplay(target, value) {
	if (isVaporComponent(target)) return setDisplay(target.block, value);
	if (isArray(target)) {
		if (target.length === 0) return;
		if (target.length === 1) return setDisplay(target[0], value);
	}
	if (isFragment(target)) return setDisplay(target.nodes, value);
	if (target instanceof Element) {
		const el = target;
		if (!(vShowOriginalDisplay in el)) el[vShowOriginalDisplay] = el.style.display === "none" ? "" : el.style.display;
		const { $transition } = target;
		if ($transition) if (value) {
			$transition.beforeEnter(target);
			el.style.display = el[vShowOriginalDisplay];
			$transition.enter(target);
		} else if (target.isConnected) $transition.leave(target, () => {
			el.style.display = "none";
		});
		else el.style.display = "none";
		else if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) && isHydrating$1) {
			if (!value && el.style.display !== "none") {
				warnPropMismatch(el, "style", 3, `display: ${el.style.display}`, "display: none");
				logMismatchError();
				el.style.display = "none";
				el[vShowOriginalDisplay] = "";
			}
		} else el.style.display = value ? el[vShowOriginalDisplay] : "none";
		el[vShowHidden] = !value;
	} else if (!!(process.env.NODE_ENV !== "production")) warn("v-show used on component with non-single-element root node and will be ignored.");
}
//#endregion
//#region packages/runtime-vapor/src/vdomInteropState.ts
let isInteropEnabled = false;
function setInteropEnabled() {
	isInteropEnabled = true;
}
//#endregion
//#region packages/runtime-vapor/src/components/Transition.ts
const displayName = "VaporTransition";
let registered = false;
const ensureTransitionHooksRegistered = () => {
	if (!registered) {
		registered = true;
		registerTransitionHooks(applyTransitionHooksImpl, applyTransitionLeaveHooksImpl);
	}
};
const hydrateTransitionImpl = () => {
	if (!currentHydrationNode || !isTemplateNode(currentHydrationNode)) return;
	const { content, parentNode } = currentHydrationNode;
	const { firstChild } = content;
	if (firstChild) {
		let transitionEl;
		for (let node = firstChild; node; node = node.nextSibling) if (node instanceof Element) {
			transitionEl = node;
			break;
		}
		parentNode.insertBefore(content, currentHydrationNode);
		parentNode.removeChild(currentHydrationNode);
		setCurrentHydrationNode(firstChild);
		if (transitionEl instanceof HTMLElement || transitionEl instanceof SVGElement) {
			const originalDisplay = transitionEl.style.display;
			transitionEl.style.display = "none";
			return (hooks) => {
				hooks.beforeEnter(transitionEl);
				transitionEl.style.display = originalDisplay;
				queuePostFlushCb(() => hooks.enter(transitionEl));
			};
		}
	}
};
const decorate$1 = (t) => {
	t.displayName = displayName;
	t.props = TransitionPropsValidators;
	t.__vapor = true;
	return t;
};
const VaporTransition = /* @__PURE__ */ decorate$1((props, { slots, expose }) => {
	expose();
	ensureTransitionHooksRegistered();
	const performAppear = isHydrating$1 ? hydrateTransitionImpl() : void 0;
	const state = useTransitionState();
	const instance = currentInstance;
	const { mode } = props;
	process.env.NODE_ENV !== "production" && checkTransitionMode(mode);
	const resolvedProps = computed(() => resolveTransitionProps(props));
	const propsProxy = new Proxy({}, { get(_, key) {
		return resolvedProps.value[key];
	} });
	const shouldCaptureVShow = !isHydrating$1 && !!props.appear;
	const shouldPerformAppear = !!props.appear && !!performAppear;
	if (instance.rawSlots.$) {
		const frag = new DynamicFragment("transition");
		let isMounted = false;
		renderEffect(() => {
			if (!frag.$transition) frag.$transition = resolveTransitionHooks(frag, propsProxy, state, instance);
			else frag.$transition.mode = resolvedProps.value.mode;
			const [, pendingVShows] = capturePendingVShows(shouldCaptureVShow && !isMounted, () => frag.update(slots.default));
			applyPendingVShows(frag.$transition, resolveTransitionBlock(frag.nodes), pendingVShows);
			if (!isMounted && shouldPerformAppear) performAppear(frag.$transition);
			isMounted = true;
		});
		return frag;
	}
	const [children, pendingVShows] = capturePendingVShows(shouldCaptureVShow, () => slots.default && slots.default() || []);
	const { hooks, root } = applyResolvedTransitionHooks(children, {
		state,
		props: propsProxy,
		instance
	});
	applyPendingVShows(hooks, root, pendingVShows);
	if (shouldPerformAppear) performAppear(hooks);
	return children;
});
const transitionTypeMap = /* @__PURE__ */ new WeakMap();
function getTransitionType(block) {
	const type = transitionTypeMap.get(block);
	if (type !== void 0) return type;
	if (block instanceof Element) return block.localName;
	if (isFragment(block) && block.vnode) {
		const children = getTransitionRawChildren([block.vnode]);
		if (children.length === 1) return children[0].type;
	}
	return block;
}
function getLeavingNodesForType(state, block) {
	const { leavingNodes } = state;
	const type = getTransitionType(block);
	let nodes = leavingNodes.get(type);
	if (!nodes) {
		nodes = Object.create(null);
		leavingNodes.set(type, nodes);
	}
	return nodes;
}
function getLeaveElement(block) {
	if (block instanceof Element) return block;
	if (isInteropEnabled && isFragment(block) && block.vnode) {
		const el = getTransitionElementFromVNode(block.vnode);
		if (el) return el;
	}
	if (isFragment(block) && !isArray(block.nodes) && (block.nodes instanceof Element || isFragment(block.nodes))) return getLeaveElement(block.nodes);
}
const getTransitionHooksContext = (block, props, state, instance, postClone) => {
	const key = String(block.$key);
	const leavingNodes = getLeavingNodesForType(state, block);
	return {
		isLeaving: () => leavingNodes[key] === block,
		setLeavingNodeCache: () => {
			leavingNodes[key] = block;
		},
		unsetLeavingNodeCache: () => {
			if (leavingNodes[key] === block) delete leavingNodes[key];
		},
		earlyRemove: () => {
			const leavingNode = leavingNodes[key];
			const el = leavingNode && getLeaveElement(leavingNode);
			if (el && el[leaveCbKey]) el[leaveCbKey]();
		},
		cloneHooks: (block) => {
			const hooks = resolveTransitionHooks(block, props, state, instance, postClone);
			if (postClone) postClone(hooks);
			return hooks;
		}
	};
};
function resolveTransitionHooks(block, props, state, instance, postClone) {
	const hooks = baseResolveTransitionHooks(getTransitionHooksContext(block, props, state, instance, postClone), props, state, instance);
	hooks.state = state;
	hooks.props = props;
	hooks.instance = instance;
	return hooks;
}
function applyTransitionHooksImpl(block, hooks) {
	return applyResolvedTransitionHooks(block, hooks).hooks;
}
function applyResolvedTransitionHooks(block, hooks) {
	if (isArray(block)) {
		block = block.filter((b) => !(b instanceof Comment));
		if (block.length === 1) block = block[0];
		else if (block.length === 0) return { hooks };
	}
	if (hooks.applyGroup && block instanceof ForFragment) {
		hooks.applyGroup(block, hooks.props, hooks.state, hooks.instance);
		return { hooks };
	}
	const fragments = [];
	const child = resolveTransitionBlock(block, (frag) => fragments.push(frag));
	if (!child) {
		fragments.forEach((f) => f.$transition = hooks);
		if (!!(process.env.NODE_ENV !== "production") && fragments.length === 0) warn("Transition component has no valid child element");
		return { hooks };
	}
	const { props, instance, state, delayedLeave } = hooks;
	let resolvedHooks = resolveTransitionHooks(child, props, state, instance, (hooks) => resolvedHooks = hooks);
	resolvedHooks.persisted = resolvedHooks.persisted || hooks.persisted;
	resolvedHooks.delayedLeave = delayedLeave;
	child.$transition = resolvedHooks;
	fragments.forEach((f) => f.$transition = resolvedHooks);
	return {
		hooks: resolvedHooks,
		root: child
	};
}
function applyTransitionLeaveHooksImpl(block, enterHooks, afterLeaveCb) {
	const leavingBlock = resolveTransitionBlock(block);
	if (!leavingBlock) return void 0;
	const { props, state, instance } = enterHooks;
	const leavingHooks = resolveTransitionHooks(leavingBlock, props, state, instance);
	leavingBlock.$transition = leavingHooks;
	const { mode } = props;
	if (mode === "out-in") {
		state.isLeaving = true;
		leavingHooks.afterLeave = () => {
			state.isLeaving = false;
			afterLeaveCb();
			leavingBlock.$transition = void 0;
			delete leavingHooks.afterLeave;
		};
	} else if (mode === "in-out") leavingHooks.delayLeave = (block, earlyRemove, delayedLeave) => {
		const leavingNodes = getLeavingNodesForType(state, leavingBlock);
		const leavingKey = String(leavingBlock.$key);
		leavingNodes[leavingKey] = leavingBlock;
		const delayedLeaveCb = () => {
			delayedLeave();
			leavingBlock.$transition = void 0;
			if (enterHooks.delayedLeave === delayedLeaveCb) delete enterHooks.delayedLeave;
		};
		block[leaveCbKey] = () => {
			earlyRemove();
			block[leaveCbKey] = void 0;
			leavingBlock.$transition = void 0;
			if (leavingNodes[leavingKey] === leavingBlock) delete leavingNodes[leavingKey];
			if (enterHooks.delayedLeave === delayedLeaveCb) delete enterHooks.delayedLeave;
		};
		enterHooks.delayedLeave = delayedLeaveCb;
	};
}
function resolveTransitionBlock(block, onFragment) {
	let child;
	if (block instanceof Node) {
		if (block instanceof Element) child = block;
	} else if (isVaporComponent(block)) if (isAsyncWrapper(block)) if (!block.type.__asyncResolved) onFragment && onFragment(block.block);
	else {
		child = resolveTransitionBlock(block.block.nodes, onFragment);
		if (child) {
			if (child.$key == null) {
				var _block$$key;
				child.$key = (_block$$key = block.$key) !== null && _block$$key !== void 0 ? _block$$key : block.uid;
			}
			transitionTypeMap.set(child, block.type);
		}
	}
	else {
		if (getComponentName(block.type) === displayName) return void 0;
		child = resolveTransitionBlock(block.block, onFragment);
		if (child) {
			if (child.$key == null) {
				var _block$$key2;
				child.$key = (_block$$key2 = block.$key) !== null && _block$$key2 !== void 0 ? _block$$key2 : block.uid;
			}
			transitionTypeMap.set(child, block.type);
		}
	}
	else if (isArray(block)) {
		let hasFound = false;
		for (const c of block) {
			if (c instanceof Comment) continue;
			const item = resolveTransitionBlock(c, onFragment);
			if (!!(process.env.NODE_ENV !== "production") && hasFound) {
				warn("<transition> can only be used on a single element or component. Use <transition-group> for lists.");
				break;
			}
			child = item;
			hasFound = true;
			if (!!!(process.env.NODE_ENV !== "production")) break;
		}
	} else if (isFragment(block)) if (isInteropEnabled && block.vnode) {
		child = block;
		const children = getTransitionRawChildren([block.vnode]);
		if (children.length === 1) transitionTypeMap.set(child, children[0].type);
	} else {
		if (onFragment) onFragment(block);
		child = resolveTransitionBlock(block.nodes, onFragment);
	}
	return child;
}
function setTransitionHooks$1(block, hooks) {
	if (isVaporComponent(block)) {
		block = resolveTransitionBlock(block.block);
		if (!block) return;
	}
	block.$transition = hooks;
}
function getVNodeKey(vnode) {
	if (!vnode) return;
	const children = getTransitionRawChildren([vnode]);
	return children.length === 1 ? children[0].key : void 0;
}
function getTransitionElementFromVNode(vnode) {
	if (!vnode) return;
	if (vnode.component) return getTransitionElementFromVNode(vnode.component.subTree);
	if (vnode.el instanceof Element) return vnode.el;
	const children = getTransitionRawChildren([vnode]);
	if (children.length === 1 && children[0] !== vnode) return getTransitionElementFromVNode(children[0]);
}
function capturePendingVShows(enabled, render) {
	if (!enabled) return [render(), void 0];
	const pendingVShows = [];
	const prev = setCurrentPendingVShows(pendingVShows);
	try {
		return [render(), pendingVShows];
	} finally {
		setCurrentPendingVShows(prev);
	}
}
function applyPendingVShows(hooks, root, pendingVShows) {
	if (!pendingVShows) return;
	if (root) hooks.persisted = hooks.persisted || pendingVShows.some((pending) => pending.target === root || resolveTransitionBlock(pending.target) === root);
	onBeforeMount(() => {
		for (const pending of pendingVShows) pending.setDisplay();
		pendingVShows.length = 0;
	});
}
//#endregion
//#region packages/runtime-vapor/src/apiDefineComponent.ts
/* @__NO_SIDE_EFFECTS__ */
function defineVaporComponent(comp, extraOptions) {
	if (isFunction(comp)) return extend({ name: comp.name }, extraOptions, {
		setup: comp,
		__vapor: true
	});
	comp.__vapor = true;
	return comp;
}
//#endregion
//#region packages/runtime-vapor/src/refCleanup.ts
/**
* Stores ref cleanup functions keyed by the element/component they are set on.
* Shared between apiTemplateRef.ts (writes) and KeepAlive deactivate (reads).
*/
const refCleanups = /* @__PURE__ */ new WeakMap();
function invalidatePendingRef(el) {
	const c = refCleanups.get(el);
	if (c && c.job) {
		c.job.flags = c.job.flags | 4;
		c.job = void 0;
	}
}
/**
* Synchronously clear the ref for an element being deactivated by KeepAlive.
* In VDOM core, refs are cleared during unmount before the deactivation check.
* Since Vapor's KeepAlive retains scopes (skipping onScopeDispose), we need
* this explicit sync cleanup path.
*/
function unsetRef(el) {
	invalidatePendingRef(el);
	const c = refCleanups.get(el);
	if (c) c.fn();
}
//#endregion
//#region packages/runtime-vapor/src/components/KeepAlive.ts
let currentKeepAliveCtx = null;
function setCurrentKeepAliveCtx(ctx) {
	try {
		return currentKeepAliveCtx;
	} finally {
		currentKeepAliveCtx = ctx;
	}
}
let currentCacheKey;
function withCurrentCacheKey(key, fn) {
	const prev = currentCacheKey;
	currentCacheKey = key;
	try {
		return fn();
	} finally {
		currentCacheKey = prev;
	}
}
const VaporKeepAlive = /* @__PURE__ */ defineVaporComponent({
	name: "VaporKeepAlive",
	__isKeepAlive: true,
	props: {
		include: [
			String,
			RegExp,
			Array
		],
		exclude: [
			String,
			RegExp,
			Array
		],
		max: [String, Number]
	},
	setup(props, { slots, expose }) {
		let exposed;
		expose(exposed);
		if (!slots.default) return;
		const keepAliveInstance = currentInstance;
		const cache = /* @__PURE__ */ new Map();
		const keys = /* @__PURE__ */ new Set();
		const storageContainer = /* @__PURE__ */ createElement("div");
		const keptAliveScopes = /* @__PURE__ */ new Map();
		const resolveCacheKeyFromBlock = (block, interop, branchKey = currentCacheKey) => {
			var _ref2, _$key;
			if (interop && isInteropEnabled) {
				var _ref, _key;
				const frag = block;
				return (_ref = frag.$key !== void 0 ? frag.$key : (_key = frag.vnode.key) !== null && _key !== void 0 ? _key : branchKey) !== null && _ref !== void 0 ? _ref : frag.vnode.type;
			}
			return (_ref2 = (_$key = block.$key) !== null && _$key !== void 0 ? _$key : branchKey) !== null && _ref2 !== void 0 ? _ref2 : block.type;
		};
		let current;
		if (!!(process.env.NODE_ENV !== "production") || __VUE_PROD_DEVTOOLS__) {
			keepAliveInstance.__v_cache = cache;
			keepAliveInstance.__v_keptAliveScopes = keptAliveScopes;
		}
		if (!!(process.env.NODE_ENV !== "production")) {
			const rerender = keepAliveInstance.hmrRerender;
			keepAliveInstance.hmrRerender = () => {
				keepAliveInstance.exposed = null;
				cache.forEach((cached) => unsetShapeFlag(cached));
				cache.clear();
				keys.clear();
				keptAliveScopes.forEach((scope) => scope.stop());
				keptAliveScopes.clear();
				storageContainer.innerHTML = "";
				current = void 0;
				rerender();
			};
		}
		keepAliveInstance.ctx = {
			getStorageContainer: () => storageContainer,
			getCachedComponent: (comp, key) => {
				var _ref4;
				if (isInteropEnabled && isVNode(comp)) {
					var _ref3, _comp$key;
					return cache.get((_ref3 = (_comp$key = comp.key) !== null && _comp$key !== void 0 ? _comp$key : currentCacheKey) !== null && _ref3 !== void 0 ? _ref3 : comp.type);
				}
				return cache.get((_ref4 = key !== null && key !== void 0 ? key : currentCacheKey) !== null && _ref4 !== void 0 ? _ref4 : comp);
			},
			activate: (instance, parentNode, anchor) => {
				current = instance;
				activate$1(instance, parentNode, anchor);
			},
			deactivate: (instance) => {
				current = void 0;
				deactivate$1(instance, storageContainer);
			}
		};
		const innerCacheBlock = (key, block, isCurrent) => {
			const { max } = props;
			if (cache.has(key)) {
				if (isCurrent) {
					keys.delete(key);
					keys.add(key);
				}
			} else {
				keys.add(key);
				if (max && keys.size > parseInt(max, 10)) pruneCacheEntry(keys.values().next().value);
			}
			cache.set(key, block);
			if (isCurrent) current = block;
		};
		const cacheBlock = (block = keepAliveInstance.block) => {
			if (isDynamicFragment(block)) {
				const transition = block.$transition;
				if (transition && transition.mode === "out-in" && transition.state.isLeaving) return;
			}
			const [innerBlock, interop] = getInnerBlock(block);
			if (!innerBlock) return;
			const cacheKey = resolveCacheKeyFromBlock(innerBlock, interop, isDynamicFragment(block) && block.keyed ? block.current : void 0);
			if (!shouldCache(innerBlock, props, interop)) {
				if (cache.has(cacheKey)) pruneCacheEntry(cacheKey);
				return;
			}
			setShapeFlag(innerBlock, interop, cache.has(cacheKey));
			const { currentBlock, currentKey } = getCurrentBlockState();
			innerCacheBlock(cacheKey, innerBlock, currentBlock === innerBlock || currentKey === cacheKey);
		};
		const processShapeFlag = (block) => {
			const [innerBlock, interop] = getInnerBlock(block);
			if (!innerBlock || !shouldCache(innerBlock, props, interop)) return false;
			const cacheKey = resolveCacheKeyFromBlock(innerBlock, interop);
			setShapeFlag(innerBlock, interop, cache.has(cacheKey));
			return cacheKey;
		};
		const pruneCache = (filter) => {
			cache.forEach((cached, key) => {
				const instance = getInstanceFromCache(cached);
				if (!instance) return;
				const name = getComponentName(isAsyncWrapper(instance) ? instance.type.__asyncResolved || {} : instance.type);
				if (name && !filter(name)) pruneCacheEntry(key);
			});
		};
		const deleteScope = (key) => {
			const scope = keptAliveScopes.get(key);
			if (scope) {
				keptAliveScopes.delete(key);
				for (const [k, s] of keptAliveScopes) if (s === scope) {
					keptAliveScopes.delete(k);
					break;
				}
			}
			return scope;
		};
		const pruneCacheEntry = (key) => {
			const cached = cache.get(key);
			if (cached && (!current || cached !== current)) {
				unsetShapeFlag(cached);
				const parentNode = findBlockNode(cached).parentNode;
				if (parentNode) remove(cached, parentNode);
			} else if (current) unsetShapeFlag(current);
			cache.delete(key);
			keys.delete(key);
			const scope = deleteScope(key);
			if (scope) scope.stop();
		};
		watch(() => [props.include, props.exclude], ([include, exclude]) => {
			include && pruneCache((name) => matches(include, name));
			exclude && pruneCache((name) => !matches(exclude, name));
		}, {
			flush: "post",
			deep: true
		});
		onMounted(cacheBlock);
		onUpdated(cacheBlock);
		const getCurrentBlockState = () => {
			const block = keepAliveInstance.block;
			const [currentBlock, interop] = getInnerBlock(block);
			const branchKey = isDynamicFragment(block) && block.keyed ? block.current : currentCacheKey;
			return {
				currentBlock,
				interop,
				currentKey: currentBlock && resolveCacheKeyFromBlock(currentBlock, interop, branchKey)
			};
		};
		onBeforeUnmount(() => {
			const { currentBlock, interop, currentKey } = getCurrentBlockState();
			const deactivateCached = (cached) => {
				unsetShapeFlag(cached);
				const instance = getInstanceFromCache(cached);
				if (instance) {
					const da = instance.da;
					da && queuePostFlushCb(da);
				}
			};
			let matched = false;
			cache.forEach((cached, key) => {
				if (currentKey === key) {
					matched = true;
					deactivateCached(cached);
					return;
				}
				unsetShapeFlag(cached);
				remove(cached, storageContainer);
			});
			if (!matched && currentBlock && isKeptAlive(currentBlock, interop)) deactivateCached(currentBlock);
			keptAliveScopes.forEach((scope) => scope.stop());
			keptAliveScopes.clear();
		});
		const prevCtx = setCurrentKeepAliveCtx({
			processShapeFlag,
			cacheBlock,
			cacheScope(cacheKey, scopeLookupKey, scope) {
				const prevScope = keptAliveScopes.get(cacheKey);
				if (prevScope && prevScope !== scope) {
					const staleScope = deleteScope(cacheKey);
					if (staleScope) staleScope.stop();
				}
				keptAliveScopes.set(cacheKey, scope);
				if (scopeLookupKey !== cacheKey) keptAliveScopes.set(scopeLookupKey, scope);
			},
			getScope(key) {
				return deleteScope(key);
			}
		});
		let children = slots.default();
		setCurrentKeepAliveCtx(prevCtx);
		if (isArray(children)) {
			children = children.filter((child) => !(child instanceof Comment));
			if (children.length > 1) {
				if (!!(process.env.NODE_ENV !== "production")) warn(`KeepAlive should contain exactly one component child.`);
				return children;
			}
		}
		return children;
	}
});
const shouldCache = (block, props, interop = false) => {
	const isAsync = isAsyncWrapper(interop ? block.vnode : block);
	const type = interop && isInteropEnabled ? block.vnode.type : block.type;
	if (isAsync && !type.__asyncResolved) return !props.include;
	const { include, exclude } = props;
	const name = getComponentName(isAsync ? type.__asyncResolved : type);
	return !(include && (!name || !matches(include, name)) || exclude && name && matches(exclude, name));
};
function setShapeFlag(block, interop, cached) {
	if (interop && isInteropEnabled) {
		if (cached) block.vnode.shapeFlag |= 512;
		block.vnode.shapeFlag |= 256;
	} else {
		if (cached) block.shapeFlag |= 512;
		block.shapeFlag |= 256;
	}
}
const unsetShapeFlag = (cached) => {
	if (isVaporComponent(cached)) {
		resetShapeFlag(cached);
		if (isAsyncWrapper(cached)) {
			const [inner] = getInnerBlock(cached.block);
			if (inner && isVaporComponent(inner)) resetShapeFlag(inner);
		}
	} else if (isInteropEnabled) resetShapeFlag(cached.vnode);
};
function isKeptAlive(cached, interop) {
	if (interop && isInteropEnabled && isInteropFragment(cached)) return !!(cached.vnode.shapeFlag & 256);
	return !!(cached.shapeFlag & 256);
}
function getInnerBlock(block) {
	if (isVaporComponent(block)) return [block, false];
	else if (isInteropEnabled && isInteropFragment(block)) return [block, true];
	else if (isFragment(block)) return getInnerBlock(block.nodes);
	return [void 0, false];
}
function isInteropFragment(block) {
	return !!(isFragment(block) && block.vnode);
}
function getInstanceFromCache(cached) {
	if (isVaporComponent(cached)) return cached;
	if (isInteropEnabled) return cached.vnode.component;
}
function activate$1(instance, parentNode, anchor) {
	move(instance.block, parentNode, anchor, 0, instance);
	queuePostFlushCb(() => {
		instance.isDeactivated = false;
		if (instance.a) invokeArrayFns(instance.a);
	});
	if (!!(process.env.NODE_ENV !== "production") || __VUE_PROD_DEVTOOLS__) devtoolsComponentAdded(instance);
}
function deactivate$1(instance, container) {
	unsetRef(instance);
	invalidateMount(instance.m);
	invalidateMount(instance.a);
	move(instance.block, container, null, 1, instance);
	queuePostFlushCb(() => {
		if (instance.da) invokeArrayFns(instance.da);
		instance.isDeactivated = true;
	});
	if (!!(process.env.NODE_ENV !== "production") || __VUE_PROD_DEVTOOLS__) devtoolsComponentAdded(instance);
}
//#endregion
//#region packages/runtime-vapor/src/components/Suspense.ts
let parentSuspense = null;
function setParentSuspense(suspense) {
	try {
		return parentSuspense;
	} finally {
		parentSuspense = suspense;
	}
}
//#endregion
//#region packages/runtime-vapor/src/vdomInterop.ts
const interopKey = Symbol(`interop`);
function filterReservedProps(props) {
	const filtered = {};
	for (const key in props) if (!isReservedProp(key)) filtered[key] = props[key];
	return filtered;
}
const vaporInteropImpl = {
	mount(vnode, container, anchor, parentComponent, parentSuspense, onBeforeMount, onVnodeBeforeMount) {
		let selfAnchor = vnode.anchor = /* @__PURE__ */ createTextNode();
		if (isHydrating$1) queuePostFlushCb(() => container.insertBefore(selfAnchor, anchor));
		else {
			vnode.el = selfAnchor;
			container.insertBefore(selfAnchor, anchor);
		}
		const prev = currentInstance;
		simpleSetCurrentInstance(parentComponent);
		const propsRef = shallowRef(filterReservedProps(vnode.props));
		const slotsRef = shallowRef(vnode.children);
		let prevSuspense = null;
		if (parentSuspense) prevSuspense = setParentSuspense(parentSuspense);
		const dynamicPropSource = [() => propsRef.value];
		dynamicPropSource[interopKey] = true;
		const instance = vnode.component = createComponent(vnode.type, { $: dynamicPropSource }, { _: slotsRef }, void 0, void 0, parentComponent ? parentComponent.appContext : vnode.appContext, true);
		instance.rawPropsRef = propsRef;
		instance.rawSlotsRef = slotsRef;
		ensureVNodeHookState(instance, vnode);
		if (parentComponent && isKeepAlive(parentComponent)) instance.shapeFlag = vnode.shapeFlag;
		if (vnode.transition) setTransitionHooks$1(instance, vnode.transition);
		if (parentSuspense) setParentSuspense(prevSuspense);
		const rootEl = getRootElement(instance);
		if (rootEl) vnode.el = rootEl;
		onVnodeBeforeMount && onVnodeBeforeMount();
		if (vnode.dirs) if (rootEl) onBeforeMount && onBeforeMount();
		else {
			if (!!(process.env.NODE_ENV !== "production")) warn("Runtime directive used on component with non-element root node. The directives will not function as intended.");
			vnode.dirs = null;
		}
		mountComponent(instance, container, selfAnchor);
		simpleSetCurrentInstance(prev);
		return instance;
	},
	update(n1, n2, shouldUpdate, onBeforeUpdate, onVnodeBeforeUpdate) {
		n2.component = n1.component;
		n2.el = n1.el;
		n2.anchor = n1.anchor;
		const instance = n2.component;
		const vnodeHookState = ensureVNodeHookState(instance, n2);
		if (shouldUpdate) {
			const rootEl = getRootElement(instance);
			if (rootEl) n2.el = rootEl;
			onVnodeBeforeUpdate && onVnodeBeforeUpdate();
			if (n2.dirs) if (rootEl) onBeforeUpdate && onBeforeUpdate();
			else n2.dirs = null;
			vnodeHookState.skipVnodeHooks = true;
			instance.rawPropsRef.value = filterReservedProps(n2.props);
			instance.rawSlotsRef.value = n2.children;
			queuePostFlushCb(() => {
				syncVNodeEl(n2, instance);
				if (!instance.isUpdating) vnodeHookState.skipVnodeHooks = false;
			});
		}
	},
	unmount(vnode, doRemove) {
		const container = doRemove ? vnode.anchor.parentNode : void 0;
		const instance = vnode.component;
		if (instance) {
			if (instance.block) unmountComponent(instance, container);
		} else if (vnode.vb) {
			remove(vnode.vb, container);
			stopVaporSlotScope(vnode);
		}
		remove(vnode.anchor, container);
	},
	slot(n1, n2, container, anchor, parentComponent, parentSuspense) {
		if (!n1) {
			const slotBlock = renderVaporSlot(n2, parentComponent, parentSuspense);
			const selfAnchor = (isFragment(slotBlock) ? slotBlock.anchor : void 0) || /* @__PURE__ */ createTextNode();
			insert(n2.el = n2.anchor = selfAnchor, container, anchor);
			insert(n2.vb = slotBlock, container, selfAnchor);
		} else if (n2.vs.slot !== n1.vs.slot) {
			const selfAnchor = n1.anchor;
			const parent = selfAnchor.parentNode;
			const nextSibling = selfAnchor.nextSibling;
			const oldBlockOwnsAnchor = isFragment(n1.vb) && n1.vb.anchor === selfAnchor;
			remove(n1.vb, parent);
			stopVaporSlotScope(n1);
			const slotBlock = renderVaporSlot(n2, parentComponent, parentSuspense);
			let newAnchor = isFragment(slotBlock) ? slotBlock.anchor : void 0;
			let insertAnchor = nextSibling;
			if (newAnchor) {
				if (!oldBlockOwnsAnchor) remove(selfAnchor, parent);
			} else if (oldBlockOwnsAnchor) newAnchor = /* @__PURE__ */ createTextNode();
			else {
				newAnchor = selfAnchor;
				insertAnchor = selfAnchor;
			}
			insert(n2.el = n2.anchor = newAnchor, parent, insertAnchor);
			insert(n2.vb = slotBlock, parent, newAnchor);
		} else {
			n2.el = n2.anchor = n1.anchor;
			n2.vb = n1.vb;
			(n2.vs.ref = n1.vs.ref).value = n2.props;
			n2.vs.scope = n1.vs.scope;
		}
	},
	move(vnode, container, anchor, moveType) {
		move(vnode.vb || vnode.component, container, anchor, moveType);
		move(vnode.anchor, container, anchor, moveType);
	},
	hydrate(vnode, node, container, anchor, parentComponent, parentSuspense, onBeforeMount, onVnodeBeforeMount) {
		if (!isHydrating$1 && !isHydrating) return node;
		hydrateNode(node, () => this.mount(vnode, container, anchor, parentComponent, parentSuspense, onBeforeMount, onVnodeBeforeMount));
		return /* @__PURE__ */ _next(node);
	},
	hydrateSlot(vnode, node, parentComponent, parentSuspense) {
		if (!isHydrating$1 && !isHydrating) return node;
		hydrateNode(node, () => {
			vnode.vb = renderVaporSlot(vnode, parentComponent, parentSuspense);
			vnode.anchor = vnode.el = isFragment(vnode.vb) && vnode.vb.anchor ? vnode.vb.anchor : currentHydrationNode;
			if (!!(process.env.NODE_ENV !== "production") && !vnode.anchor) throw new Error(`Failed to locate slot anchor. this is likely a Vue internal bug.`);
		});
		return isComment(node, "[") ? vnode.anchor.nextSibling : vnode.anchor;
	},
	setTransitionHooks(component, hooks) {
		setTransitionHooks$1(component, hooks);
	},
	activate(vnode, container, anchor, parentComponent) {
		const cached = parentComponent.ctx.getCachedComponent(vnode);
		vnode.el = cached.el;
		vnode.component = cached.component;
		vnode.anchor = cached.anchor;
		const instance = vnode.component;
		const vnodeHookState = ensureVNodeHookState(instance, vnode);
		const rootEl = getRootElement(instance);
		if (rootEl) vnode.el = rootEl;
		if (vnode.dirs && !rootEl) {
			if (!!(process.env.NODE_ENV !== "production")) warn("Runtime directive used on component with non-element root node. The directives will not function as intended.");
			vnode.dirs = null;
		}
		if (shouldUpdateComponent(cached, vnode)) {
			vnodeHookState.skipVnodeHooks = true;
			instance.rawPropsRef.value = filterReservedProps(vnode.props);
			instance.rawSlotsRef.value = vnode.children;
			const vnodeBeforeUpdateHook = vnode.props && vnode.props.onVnodeBeforeUpdate;
			if (vnodeBeforeUpdateHook) callWithAsyncErrorHandling(vnodeBeforeUpdateHook, parentComponent, 7, [vnode, cached]);
			if (vnode.dirs) invokeDirectiveHook(vnode, cached, parentComponent, "beforeUpdate");
			queuePostFlushCb(() => {
				syncVNodeEl(vnode, instance);
				if (vnode.dirs) invokeDirectiveHook(vnode, cached, parentComponent, "updated");
				const vnodeUpdatedHook = vnode.props && vnode.props.onVnodeUpdated;
				if (vnodeUpdatedHook) callWithAsyncErrorHandling(vnodeUpdatedHook, parentComponent, 7, [vnode, cached]);
				if (!instance.isUpdating) vnodeHookState.skipVnodeHooks = false;
			});
		}
		activate$1(instance, container, anchor);
		insert(vnode.anchor, container, anchor);
		const vnodeMountedHook = vnode.props && vnode.props.onVnodeMounted;
		if (vnodeMountedHook) queuePostFlushCb(() => {
			callWithAsyncErrorHandling(vnodeMountedHook, parentComponent, 7, [vnode]);
		});
	},
	deactivate(vnode, container) {
		const instance = vnode.component;
		deactivate$1(instance, container);
		insert(vnode.anchor, container);
		queuePostFlushCb(() => {
			const vnodeHook = vnode.props && vnode.props.onVnodeUnmounted;
			if (vnodeHook) callWithAsyncErrorHandling(vnodeHook, instance.parent, 7, [vnode]);
		});
	}
};
const vaporSlotPropsProxyHandler = {
	get(target, key) {
		return target.value[key];
	},
	has(target, key) {
		return key in target.value;
	},
	ownKeys(target) {
		return Reflect.ownKeys(target.value);
	},
	getOwnPropertyDescriptor(target, key) {
		if (key in target.value) return {
			enumerable: true,
			configurable: true
		};
	}
};
const vaporSlotWrappersCache = /* @__PURE__ */ new WeakMap();
const vaporSlotsProxyHandler = { get(target, key) {
	const slot = target[key];
	if (isFunction(slot)) {
		slot.__vapor = true;
		let wrappers = vaporSlotWrappersCache.get(target);
		if (!wrappers) vaporSlotWrappersCache.set(target, wrappers = /* @__PURE__ */ new Map());
		const cached = wrappers.get(key);
		if (cached && cached.slot === slot) return cached.wrapped;
		const wrapped = (props) => [renderSlot({ [key]: slot }, key, props)];
		wrapped.__vs = slot;
		wrappers.set(key, {
			slot,
			wrapped
		});
		return wrapped;
	}
	return slot;
} };
let vdomHydrateNode;
function resolveVNodeRange(vnode) {
	const { type, shapeFlag, el, anchor } = vnode;
	if (shapeFlag & 64 && el && anchor && anchor !== el) return [el, anchor];
	if ((type === Static || type === Fragment) && el && anchor && anchor !== el) return [el, anchor];
	if (!(shapeFlag & 6)) return;
	const subTree = vnode.component && vnode.component.subTree;
	const subEl = subTree && subTree.el;
	const subAnchor = subTree && subTree.anchor;
	if (subTree && (subTree.type === Static || subTree.type === Fragment) && subEl && subAnchor && subAnchor !== subEl) return [subEl, subAnchor];
}
function resolveVNodeNodes(vnode) {
	const vnodeRange = resolveVNodeRange(vnode);
	if (vnodeRange) {
		const nodeRange = [];
		let n = vnodeRange[0];
		while (n) {
			nodeRange.push(n);
			if (n === vnodeRange[1]) break;
			n = n.nextSibling;
		}
		return nodeRange;
	}
	return vnode.el;
}
function appendVnodeUpdatedHook(vnode, hook) {
	const props = vnode.props || (vnode.props = {});
	const existing = props.onVnodeUpdated;
	props.onVnodeUpdated = existing ? isArray(existing) ? [...existing, hook] : [existing, hook] : hook;
}
function trackFragmentVNodeUpdates(frag, vnode) {
	const refresh = () => {
		frag.nodes = resolveVNodeNodes(vnode);
		if (frag.onUpdated) frag.onUpdated.forEach((m) => m());
	};
	appendVnodeUpdatedHook(vnode, refresh);
}
/**
* Mount VNode in vapor
*/
function mountVNode(internals, vnode, parentComponent) {
	const suspense = parentSuspense || parentComponent && parentComponent.suspense;
	const frag = new VaporFragment([]);
	frag.vnode = vnode;
	frag.$key = vnode.key;
	trackFragmentVNodeUpdates(frag, vnode);
	let isMounted = false;
	const unmount = (parentNode, transition) => {
		if (transition) setTransitionHooks(vnode, transition);
		if (vnode.shapeFlag & 256) if (vnode.type.__vapor) deactivate$1(vnode.component, parentComponent.ctx.getStorageContainer());
		else deactivate(vnode, parentComponent.ctx.getStorageContainer(), internals, parentComponent, null);
		else internals.um(vnode, parentComponent, null, !!parentNode);
	};
	frag.hydrate = () => {
		if (!isHydrating$1) return;
		hydrateVNode(vnode, parentComponent);
		onScopeDispose(unmount, true);
		isMounted = true;
		frag.nodes = resolveVNodeNodes(vnode);
	};
	frag.insert = (parentNode, anchor, transition) => {
		if (isHydrating$1) return;
		if (vnode.shapeFlag & 512) {
			if (vnode.type.__vapor) activate$1(vnode.component, parentNode, anchor);
			else activate(vnode, parentNode, anchor, internals, parentComponent, null, void 0, false);
			return;
		} else {
			const prev = currentInstance;
			simpleSetCurrentInstance(parentComponent);
			if (!isMounted) {
				if (transition) setTransitionHooks(vnode, transition);
				internals.p(null, vnode, parentNode, anchor, parentComponent, suspense, void 0, vnode.slotScopeIds);
				onScopeDispose(unmount, true);
				isMounted = true;
			} else internals.m(vnode, parentNode, anchor, 2, parentComponent);
			simpleSetCurrentInstance(prev);
		}
		frag.nodes = resolveVNodeNodes(vnode);
		if (isMounted && frag.onUpdated) frag.onUpdated.forEach((m) => m());
	};
	frag.remove = unmount;
	return frag;
}
/**
* Mount vdom component in vapor
*/
function createVDOMComponent(internals, component, parentComponent, rawProps, rawSlots) {
	const suspense = parentSuspense || parentComponent && parentComponent.suspense;
	const useBridge = shouldUseRendererBridge(component);
	const comp = useBridge ? ensureRendererBridge(component) : component;
	const frag = new VaporFragment([]);
	const vnode = frag.vnode = createVNode(comp, rawProps && extend({}, new Proxy(rawProps, rawPropsProxyHandlers)));
	frag.$key = vnode.key;
	trackFragmentVNodeUpdates(frag, vnode);
	if (currentKeepAliveCtx) {
		currentKeepAliveCtx.processShapeFlag(frag);
		if (component.__asyncLoader) {
			const keepAliveCtx = currentKeepAliveCtx;
			let disposed = false;
			onScopeDispose(() => disposed = true);
			component.__asyncLoader().then(() => {
				if (!disposed) keepAliveCtx.cacheBlock(frag);
			}).catch(NOOP);
		}
		setCurrentKeepAliveCtx(null);
	}
	const wrapper = new VaporComponentInstance(useBridge ? comp : { props: component.props }, rawProps, rawSlots, parentComponent ? parentComponent.appContext : void 0, void 0);
	vnode.vi = (instance) => {
		instance.props = shallowReactive(wrapper.props);
		const attrs = createInternalObject();
		const isFilteredEmit = (key) => typeof key === "string" && isEmitListener(instance.emitsOptions, key);
		instance.attrs = new Proxy(attrs, {
			get(_, key) {
				if (isFilteredEmit(key)) return;
				return wrapper.attrs[key];
			},
			has(_, key) {
				return !isFilteredEmit(key) && key in wrapper.attrs;
			},
			ownKeys() {
				return Reflect.ownKeys(wrapper.attrs).filter((key) => !isFilteredEmit(key));
			},
			getOwnPropertyDescriptor(_, key) {
				if (!isFilteredEmit(key) && key in wrapper.attrs) return {
					enumerable: true,
					configurable: true
				};
			}
		});
		instance.slots = wrapper.slots === EMPTY_OBJ ? EMPTY_OBJ : new Proxy(wrapper.slots, vaporSlotsProxyHandler);
	};
	let rawRef = null;
	let isMounted = false;
	const unmount = (parentNode, transition) => {
		if (rawRef) setRef(rawRef, null, null, vnode, true);
		if (transition) setTransitionHooks(vnode, transition);
		if (vnode.shapeFlag & 256) {
			deactivate(vnode, parentComponent.ctx.getStorageContainer(), internals, parentComponent, null);
			return;
		}
		internals.umt(vnode.component, null, !!parentNode);
	};
	frag.hydrate = () => {
		if (!isHydrating$1) return;
		hydrateVNode(vnode, parentComponent);
		onScopeDispose(unmount, true);
		isMounted = true;
		frag.nodes = resolveVNodeNodes(vnode);
	};
	vnode.scopeId = getCurrentScopeId() || null;
	vnode.slotScopeIds = currentSlotScopeIds;
	frag.insert = (parentNode, anchor, transition) => {
		if (isHydrating$1) return;
		if (vnode.shapeFlag & 512) activate(vnode, parentNode, anchor, internals, parentComponent, null, void 0, false);
		else {
			const prev = currentInstance;
			simpleSetCurrentInstance(parentComponent);
			if (!isMounted) {
				if (transition) setTransitionHooks(vnode, transition);
				internals.mt(vnode, parentNode, anchor, parentComponent, suspense, void 0, false);
				if (rawRef) setRef(rawRef, null, null, vnode);
				onScopeDispose(unmount, true);
				isMounted = true;
			} else internals.m(vnode, parentNode, anchor, 2, parentComponent);
			simpleSetCurrentInstance(prev);
		}
		frag.nodes = resolveVNodeNodes(vnode);
		if (isMounted && frag.onUpdated) frag.onUpdated.forEach((m) => m());
	};
	frag.remove = unmount;
	frag.setRef = (instance, ref, refFor, refKey) => {
		const oldRawRef = rawRef;
		rawRef = normalizeRef({
			ref,
			ref_for: refFor,
			ref_key: refKey
		}, instance);
		if (isMounted) {
			if (rawRef) setRef(rawRef, oldRawRef, null, vnode);
			else if (oldRawRef) setRef(oldRawRef, null, null, vnode, true);
		}
	};
	return frag;
}
const rendererBridgeCache = /* @__PURE__ */ new WeakMap();
/**
* Teleport/Suspense are renderer primitives (`__isTeleport` / `__isSuspense`),
* not regular components with their own render pipeline.
*
* We wrap them with a tiny functional bridge so they can pass through the
* interop component mount path while preserving built-in vnode semantics.
*/
function shouldUseRendererBridge(component) {
	return !!(component.__isTeleport || component.__isSuspense);
}
function ensureRendererBridge(component) {
	let bridge = rendererBridgeCache.get(component);
	if (!bridge) rendererBridgeCache.set(component, bridge = (props, { slots }) => createVNode(component, props, slots));
	return bridge;
}
function trackSlotVNodeUpdates(frag, vnode) {
	const refresh = () => {
		frag.nodes = resolveVNodeNodes(vnode);
		if (frag.onUpdated) frag.onUpdated.forEach((m) => m());
	};
	const track = (node) => {
		appendVnodeUpdatedHook(node, refresh);
		if (node.type === Fragment && isArray(node.children)) node.children.forEach((child) => {
			if (isVNode(child)) track(child);
		});
	};
	track(vnode);
}
/**
* Mount vdom slot in vapor
*/
function renderVDOMSlot(internals, slotsRef, name, props, parentComponent, fallback) {
	const suspense = parentSuspense || parentComponent.suspense;
	const frag = new VaporFragment([]);
	const instance = currentInstance;
	const slotOwner = currentSlotOwner;
	if (fallback && !frag.fallback) frag.fallback = fallback;
	let isMounted = false;
	let currentBlock = null;
	let currentVNode = null;
	let currentParentNode;
	let currentAnchor;
	let scope = effectScope();
	frag.insert = (parentNode, anchor) => {
		if (isHydrating$1) return;
		currentParentNode = parentNode;
		currentAnchor = anchor;
		if (!isMounted) {
			scope.run(render);
			isMounted = true;
		} else if (currentVNode) internals.m(currentVNode, parentNode, anchor, 2, parentComponent);
		else if (currentBlock) insert(currentBlock, parentNode, anchor);
		if (isMounted && frag.onUpdated) frag.onUpdated.forEach((m) => m());
	};
	frag.remove = (parentNode) => {
		scope.stop();
		if (currentBlock) remove(currentBlock, parentNode);
		else if (currentVNode) internals.um(currentVNode, parentComponent, null, true);
	};
	const render = () => {
		const prev = currentInstance;
		simpleSetCurrentInstance(instance);
		try {
			renderEffect(() => {
				const prevSlotOwner = setCurrentSlotOwner(slotOwner);
				try {
					const effectiveFallback = frag.fallback || fallback;
					let slotContent;
					let isEmpty = true;
					let emptyFrag = null;
					if (slotsRef.value) {
						slotContent = renderSlot(slotsRef.value, isFunction(name) ? name() : name, props);
						if (isVNode(slotContent)) {
							const children = slotContent.children;
							ensureVaporSlotFallback(children, effectiveFallback);
							isEmpty = children.length === 0;
						} else {
							if (effectiveFallback && slotContent) emptyFrag = attachSlotFallback(slotContent, () => effectiveFallback(internals, parentComponent));
							isEmpty = !isValidBlock(slotContent);
						}
					}
					let resolved = slotContent;
					if (isEmpty && effectiveFallback) if (isVNode(slotContent)) resolved = effectiveFallback(internals, parentComponent);
					else if (slotContent) resolved = renderSlotFallback(slotContent, () => effectiveFallback(internals, parentComponent), emptyFrag);
					else resolved = effectiveFallback(internals, parentComponent);
					if (isHydrating$1) {
						if (isVNode(resolved)) {
							frag.vnode = resolved;
							frag.$key = getVNodeKey(resolved);
							trackSlotVNodeUpdates(frag, resolved);
							hydrateVNode(resolved, parentComponent);
							currentVNode = resolved;
							currentBlock = null;
							frag.nodes = resolveVNodeNodes(resolved);
						} else if (resolved) {
							currentBlock = resolved;
							currentVNode = null;
							frag.nodes = resolved;
						} else {
							currentBlock = null;
							currentVNode = null;
							frag.nodes = [];
						}
						return;
					}
					if (isVNode(resolved)) {
						frag.vnode = resolved;
						frag.$key = getVNodeKey(resolved);
						trackSlotVNodeUpdates(frag, resolved);
						if (currentBlock) {
							remove(currentBlock, currentParentNode);
							currentBlock = null;
						}
						internals.p(currentVNode, resolved, currentParentNode, currentAnchor, parentComponent, suspense, void 0, resolved.slotScopeIds);
						currentVNode = resolved;
						frag.nodes = resolveVNodeNodes(resolved);
						return;
					}
					if (resolved) {
						frag.vnode = null;
						frag.$key = void 0;
						if (currentVNode) {
							internals.um(currentVNode, parentComponent, null, true);
							currentVNode = null;
						}
						if (currentBlock) remove(currentBlock, currentParentNode);
						insert(resolved, currentParentNode, currentAnchor);
						currentBlock = resolved;
						frag.nodes = resolved;
						return;
					}
					if (currentBlock) {
						remove(currentBlock, currentParentNode);
						currentBlock = null;
					}
					if (currentVNode) {
						internals.um(currentVNode, parentComponent, null, true);
						currentVNode = null;
					}
					frag.vnode = null;
					frag.$key = void 0;
					frag.nodes = [];
				} finally {
					setCurrentSlotOwner(prevSlotOwner);
				}
			});
		} finally {
			simpleSetCurrentInstance(prev);
		}
	};
	frag.hydrate = () => {
		if (!isHydrating$1) return;
		scope.run(render);
		currentParentNode = currentHydrationNode.parentNode;
		currentAnchor = currentHydrationNode;
		isMounted = true;
	};
	return frag;
}
const vaporInteropPlugin = (app) => {
	setInteropEnabled();
	const internals = ensureRenderer().internals;
	app._context.vapor = extend(vaporInteropImpl, {
		vdomMount: createVDOMComponent.bind(null, internals),
		vdomUnmount: internals.umt,
		vdomSlot: renderVDOMSlot.bind(null, internals),
		vdomMountVNode: mountVNode.bind(null, internals)
	});
	const mount = app.mount;
	app.mount = ((...args) => {
		optimizePropertyLookup();
		return mount(...args);
	});
};
function hydrateVNode(vnode, parentComponent) {
	const node = currentHydrationNode;
	if (!vdomHydrateNode) vdomHydrateNode = ensureHydrationRenderer().hydrateNode;
	const nextNode = vdomHydrateNode(node, vnode, parentComponent, null, null, false);
	if (nextNode) setCurrentHydrationNode(nextNode);
	else advanceHydrationNode(node);
}
function createVaporFallback(fallback, parentComponent) {
	const internals = ensureRenderer().internals;
	return () => {
		const block = createFallback(fallback)(internals, parentComponent);
		if (isHydrating$1 && isFragment(block) && block.hydrate) block.hydrate();
		return block;
	};
}
const createFallback = (fallback) => (internals, parentComponent) => {
	const fallbackNodes = fallback();
	if (isArray(fallbackNodes) && fallbackNodes.every(isVNode)) return createVNodeChildrenFragment(internals, () => fallback(), parentComponent);
	return fallbackNodes;
};
function renderVaporSlot(vnode, parentComponent, parentSuspense) {
	const prev = currentInstance;
	let prevSuspense = null;
	simpleSetCurrentInstance(parentComponent);
	if (parentSuspense) prevSuspense = setParentSuspense(parentSuspense);
	try {
		const { fallback } = vnode.vs;
		if (isHydrating$1 && fallback) {
			const frag = new SlotFragment();
			frag.updateSlot(() => invokeVaporSlot(vnode), createVaporFallback(fallback, parentComponent));
			return frag;
		}
		if (!fallback) return invokeVaporSlot(vnode);
		let slotBlock = invokeVaporSlot(vnode);
		const vaporFallback = createVaporFallback(fallback, parentComponent);
		if (!slotBlock) return vaporFallback();
		const emptyFrag = attachSlotFallback(slotBlock, vaporFallback);
		if (!isValidBlock(slotBlock)) slotBlock = renderSlotFallback(slotBlock, vaporFallback, emptyFrag);
		return slotBlock;
	} finally {
		if (parentSuspense) setParentSuspense(prevSuspense);
		simpleSetCurrentInstance(prev);
	}
}
function stopVaporSlotScope(vnode) {
	if (vnode.vs && vnode.vs.scope) {
		vnode.vs.scope.stop();
		vnode.vs.scope = void 0;
	}
}
/**
* Slot functions can create renderEffects while evaluating their block.
* Those effects live in this dedicated scope so slot re-mount/unmount can
* dispose them immediately instead of waiting for the parent component.
*/
function invokeVaporSlot(vnode) {
	const propsRef = vnode.vs.ref = shallowRef(vnode.props);
	const scope = effectScope();
	vnode.vs.scope = scope;
	try {
		return scope.run(() => vnode.vs.slot(new Proxy(propsRef, vaporSlotPropsProxyHandler)));
	} catch (e) {
		vnode.vs.scope = void 0;
		scope.stop();
		throw e;
	}
}
function syncVNodeEl(vnode, instance) {
	const rootEl = getRootElement(instance);
	if (rootEl) vnode.el = rootEl;
	else {
		vnode.el = vnode.anchor;
		vnode.dirs = null;
	}
}
const vnodeHookStateMap = /* @__PURE__ */ new WeakMap();
function ensureVNodeHookState(instance, vnode) {
	let state = vnodeHookStateMap.get(instance);
	if (!state) {
		state = {
			vnode,
			skipVnodeHooks: false
		};
		vnodeHookStateMap.set(instance, state);
		(instance.bu || (instance.bu = [])).push(() => {
			if (state.skipVnodeHooks) return;
			const vnodeHook = state.vnode.props && state.vnode.props.onVnodeBeforeUpdate;
			if (vnodeHook) callWithAsyncErrorHandling(vnodeHook, instance.parent, 7, [state.vnode, state.vnode]);
		});
		(instance.u || (instance.u = [])).unshift(() => syncVNodeEl(state.vnode, instance));
		instance.u.push(() => {
			if (state.skipVnodeHooks) {
				state.skipVnodeHooks = false;
				return;
			}
			const vnodeHook = state.vnode.props && state.vnode.props.onVnodeUpdated;
			if (vnodeHook) callWithAsyncErrorHandling(vnodeHook, instance.parent, 7, [state.vnode, state.vnode]);
		});
	} else state.vnode = vnode;
	return state;
}
function createVNodeChildrenFragment(internals, render, parentComponent) {
	const suspense = parentSuspense || parentComponent && parentComponent.suspense;
	const frag = new VaporFragment([]);
	let currentVNode = null;
	let currentChildren = [];
	let currentParentNode = null;
	let currentAnchor = null;
	let isMounted = false;
	const scope = effectScope();
	const renderContent = () => {
		const prev = currentInstance;
		simpleSetCurrentInstance(parentComponent);
		try {
			renderEffect(() => {
				const nextChildren = render();
				if (isHydrating$1) {
					nextChildren.forEach((vnode) => hydrateVNode(vnode, parentComponent));
					currentChildren = nextChildren;
					currentVNode = createVNode(Fragment, null, nextChildren);
					currentParentNode = currentHydrationNode.parentNode;
					currentAnchor = currentHydrationNode;
				} else if (!currentVNode) {
					currentChildren = nextChildren;
					currentVNode = createVNode(Fragment, null, nextChildren);
					if (nextChildren.length) internals.mc(nextChildren, currentParentNode, currentAnchor, parentComponent, suspense, void 0, null, false);
				} else {
					const nextVNode = createVNode(Fragment, null, nextChildren);
					internals.pc(currentVNode, nextVNode, currentParentNode, currentAnchor, parentComponent, suspense, void 0, null, false);
					currentChildren = nextChildren;
					currentVNode = nextVNode;
				}
				if (currentChildren.length === 0) frag.nodes = [];
				else if (currentChildren.length === 1) frag.nodes = resolveVNodeNodes(currentChildren[0]);
				else frag.nodes = currentChildren.map(resolveVNodeNodes);
				if (isMounted && frag.onUpdated) frag.onUpdated.forEach((hook) => hook());
			});
		} finally {
			simpleSetCurrentInstance(prev);
		}
	};
	frag.insert = (parentNode, anchor) => {
		if (isHydrating$1) return;
		currentParentNode = parentNode;
		currentAnchor = anchor;
		if (!isMounted) {
			scope.run(renderContent);
			isMounted = true;
		} else currentChildren.forEach((vnode) => {
			internals.m(vnode, parentNode, anchor, 2, parentComponent);
		});
	};
	frag.remove = (parentNode) => {
		scope.stop();
		currentChildren.forEach((vnode) => {
			internals.um(vnode, parentComponent, null, !!parentNode);
		});
	};
	frag.hydrate = () => {
		if (!isHydrating$1) return;
		scope.run(renderContent);
		isMounted = true;
	};
	return frag;
}
//#endregion
//#region packages/runtime-vapor/src/componentEmits.ts
/**
* The logic from core isn't too reusable so it's better to duplicate here
*/
function normalizeEmitsOptions(comp) {
	const cached = comp.__emitsOptions;
	if (cached) return cached;
	const raw = comp.emits;
	if (!raw) return null;
	let normalized;
	if (isArray(raw)) {
		normalized = {};
		for (const key of raw) normalized[key] = null;
	} else normalized = raw;
	return comp.__emitsOptions = normalized;
}
function emit(instance, event, ...rawArgs) {
	baseEmit(instance, instance.rawProps || EMPTY_OBJ, propGetter, event, ...rawArgs);
}
function propGetter(rawProps, key) {
	const dynamicSources = rawProps.$;
	if (dynamicSources) {
		let i = dynamicSources.length;
		while (i--) {
			const source = resolveSource(dynamicSources[i]);
			if (hasOwn(source, key)) return dynamicSources[interopKey] || isOn(key) && isFunction(dynamicSources[i]) ? source[key] : resolveSource(source[key]);
		}
	}
	return rawProps[key] && resolveSource(rawProps[key]);
}
//#endregion
//#region packages/runtime-vapor/src/componentProps.ts
function resolveSource(source) {
	return isFunction(source) ? resolveFunctionSource(source) : source;
}
function resolveFunctionSource(source) {
	if (source._cache) return source._cache.value;
	const parent = currentInstance && currentInstance.parent;
	if (parent) {
		source._cache = computed(() => {
			const prev = setCurrentInstance(parent);
			try {
				return source();
			} finally {
				setCurrentInstance(...prev);
			}
		});
		onScopeDispose$1(() => source._cache = void 0);
		return source._cache.value;
	}
	return source();
}
function getPropsProxyHandlers(comp, once) {
	if (comp.__propsHandlers) return comp.__propsHandlers;
	const propsOptions = normalizePropsOptions(comp)[0];
	const emitsOptions = normalizeEmitsOptions(comp);
	const isProp = propsOptions ? (key) => isString(key) && hasOwn(propsOptions, camelize(key)) : NO;
	const isAttr = propsOptions ? (key) => isString(key) && key !== "$" && !isProp(key) && !isEmitListener(emitsOptions, key) : (key) => isString(key);
	const getProp = (instance, key) => {
		if (key === "__v_isReactive") return true;
		if (!isProp(key)) return;
		const rawProps = instance.rawProps;
		const dynamicSources = rawProps.$;
		if (dynamicSources) {
			let i = dynamicSources.length;
			let source, isDynamic, rawKey;
			while (i--) {
				source = dynamicSources[i];
				isDynamic = isFunction(source);
				source = isDynamic ? resolveFunctionSource(source) : source;
				for (rawKey in source) if (camelize(rawKey) === key) return resolvePropValue(propsOptions, key, isDynamic ? source[rawKey] : resolveFunctionSource(source[rawKey]), instance, resolveDefault);
			}
		}
		for (const rawKey in rawProps) if (camelize(rawKey) === key) return resolvePropValue(propsOptions, key, resolveSource(rawProps[rawKey]), instance, resolveDefault);
		return resolvePropValue(propsOptions, key, void 0, instance, resolveDefault, true);
	};
	const withOnceCache = (getter) => {
		return ((instance, key) => {
			const cache = instance.oncePropsCache || (instance.oncePropsCache = {});
			if (!(key in cache)) {
				pauseTracking();
				try {
					cache[key] = getter(instance, key);
				} finally {
					resetTracking();
				}
			}
			return cache[key];
		});
	};
	const getOnceProp = withOnceCache(getProp);
	const propsHandlers = propsOptions ? {
		get: (target, key) => (once ? getOnceProp : getProp)(target, key),
		has: (_, key) => isProp(key),
		ownKeys: () => Object.keys(propsOptions),
		getOwnPropertyDescriptor(target, key) {
			if (isProp(key)) return {
				configurable: true,
				enumerable: true,
				get: () => (once ? getOnceProp : getProp)(target, key)
			};
		}
	} : null;
	if (!!(process.env.NODE_ENV !== "production") && propsOptions) Object.assign(propsHandlers, {
		set: propsSetDevTrap,
		deleteProperty: propsDeleteDevTrap
	});
	const getAttr = (target, key) => {
		if (isString(key) && !isProp(key) && !isEmitListener(emitsOptions, key)) return getAttrFromRawProps(target, key);
	};
	const hasAttr = (target, key) => {
		if (isAttr(key)) return hasAttrFromRawProps(target, key);
		else return false;
	};
	const getOnceAttr = withOnceCache((instance, key) => getAttr(instance.rawProps, key));
	const attrsHandlers = {
		get: (target, key) => once ? getOnceAttr(target, key) : getAttr(target.rawProps, key),
		has: (target, key) => hasAttr(target.rawProps, key),
		ownKeys: (target) => getKeysFromRawProps(target.rawProps).filter(isAttr),
		getOwnPropertyDescriptor(target, key) {
			if (isString(key) && hasAttr(target.rawProps, key)) return {
				configurable: true,
				enumerable: true,
				get: () => once ? getOnceAttr(target, key) : getAttr(target.rawProps, key)
			};
		}
	};
	if (!!(process.env.NODE_ENV !== "production")) Object.assign(attrsHandlers, {
		set: propsSetDevTrap,
		deleteProperty: propsDeleteDevTrap
	});
	return comp.__propsHandlers = [propsHandlers, attrsHandlers];
}
function getAttrFromRawProps(rawProps, key) {
	if (key === "$") return;
	const merged = key === "class" || key === "style" ? [] : void 0;
	const dynamicSources = rawProps.$;
	if (dynamicSources) {
		let i = dynamicSources.length;
		let source, isDynamic;
		while (i--) {
			source = dynamicSources[i];
			isDynamic = isFunction(source);
			source = isDynamic ? resolveFunctionSource(source) : source;
			if (source && hasOwn(source, key)) {
				const value = isDynamic ? source[key] : resolveFunctionSource(source[key]);
				if (merged) merged.push(value);
				else return value;
			}
		}
	}
	if (hasOwn(rawProps, key)) {
		const value = resolveSource(rawProps[key]);
		if (merged) merged.push(value);
		else return value;
	}
	if (merged && merged.length) return merged;
}
function hasAttrFromRawProps(rawProps, key) {
	if (key === "$") return false;
	const dynamicSources = rawProps.$;
	if (dynamicSources) {
		let i = dynamicSources.length;
		while (i--) {
			const source = resolveSource(dynamicSources[i]);
			if (source && hasOwn(source, key)) return true;
		}
	}
	return hasOwn(rawProps, key);
}
function getKeysFromRawProps(rawProps) {
	const keys = [];
	for (const key in rawProps) if (key !== "$") keys.push(key);
	const dynamicSources = rawProps.$;
	if (dynamicSources) {
		let i = dynamicSources.length;
		let source;
		while (i--) {
			source = isFunction(dynamicSources[i]) ? resolveFunctionSource(dynamicSources[i]) : dynamicSources[i];
			for (const key in source) keys.push(key);
		}
	}
	return Array.from(new Set(keys));
}
function normalizePropsOptions(comp) {
	const cached = comp.__propsOptions;
	if (cached) return cached;
	const raw = comp.props;
	if (!raw) return EMPTY_ARR;
	const normalized = {};
	const needCastKeys = [];
	baseNormalizePropsOptions(raw, normalized, needCastKeys);
	return comp.__propsOptions = [normalized, needCastKeys];
}
function resolveDefault(factory, instance) {
	const prev = setCurrentInstance(instance);
	const res = factory.call(null, instance.props);
	setCurrentInstance(...prev);
	return res;
}
function hasFallthroughAttrs(comp, rawProps) {
	if (rawProps) if (rawProps.$ || !comp.props) return true;
	else {
		const propsOptions = normalizePropsOptions(comp)[0];
		for (const key in rawProps) if (!hasOwn(propsOptions, camelize(key))) return true;
	}
	return false;
}
/**
* dev only
*/
function setupPropsValidation(instance) {
	const rawProps = instance.rawProps;
	if (!rawProps) return;
	renderEffect(() => {
		pushWarningContext(instance);
		validateProps(resolveDynamicProps(rawProps), instance.props, normalizePropsOptions(instance.type)[0]);
		popWarningContext();
	}, true);
}
function resolveDynamicProps(props) {
	const mergedRawProps = {};
	for (const key in props) if (key !== "$") mergedRawProps[key] = resolveSource(props[key]);
	if (props.$) for (const source of props.$) {
		const isDynamic = isFunction(source);
		const resolved = isDynamic ? resolveFunctionSource(source) : source;
		for (const key in resolved) {
			const value = isDynamic ? resolved[key] : resolveFunctionSource(source[key]);
			if (key === "class" || key === "style") {
				const existing = mergedRawProps[key];
				if (isArray(existing)) existing.push(value);
				else mergedRawProps[key] = [existing, value];
			} else mergedRawProps[key] = value;
		}
	}
	return mergedRawProps;
}
function propsSetDevTrap(_, key) {
	warn(`Attempt to mutate prop ${JSON.stringify(key)} failed. Props are readonly.`);
	return true;
}
function propsDeleteDevTrap(_, key) {
	warn(`Attempt to delete prop ${JSON.stringify(key)} failed. Props are readonly.`);
	return true;
}
const rawPropsProxyHandlers = {
	get: getAttrFromRawProps,
	has: hasAttrFromRawProps,
	ownKeys: getKeysFromRawProps,
	getOwnPropertyDescriptor(target, key) {
		if (hasAttrFromRawProps(target, key)) return {
			configurable: true,
			enumerable: true,
			get: () => getAttrFromRawProps(target, key)
		};
	}
};
//#endregion
//#region packages/runtime-vapor/src/scopeId.ts
function setScopeId(block, scopeIds) {
	if (block instanceof Element) for (const id of scopeIds) block.setAttribute(id, "");
	else if (isVaporComponent(block)) setScopeId(block.block, scopeIds);
	else if (isArray(block)) for (const b of block) setScopeId(b, scopeIds);
	else if (isFragment(block)) setScopeId(block.nodes, scopeIds);
}
function setComponentScopeId(instance) {
	const { parent, scopeId } = instance;
	if (!parent || !scopeId) return;
	if (isArray(instance.block) && instance.block.length > 1) return;
	const scopeIds = [];
	const parentScopeId = parent && parent.type.__scopeId;
	if (parentScopeId !== scopeId) scopeIds.push(scopeId);
	else if (parentScopeId) scopeIds.push(parentScopeId);
	if (isInteropEnabled && parent.subTree && parent.subTree.component === instance && parent.vnode.scopeId) {
		scopeIds.push(parent.vnode.scopeId);
		const inheritedScopeIds = getInheritedScopeIds(parent.vnode, parent.parent);
		scopeIds.push(...inheritedScopeIds);
	}
	if (scopeIds.length > 0) setScopeId(instance.block, scopeIds);
}
//#endregion
//#region packages/runtime-vapor/src/componentSlots.ts
/**
* Flag to indicate if we are executing a once slot.
* When true, renderEffect should skip creating reactive effect.
*/
let inOnceSlot = false;
/**
* Current slot scopeIds for vdom interop
*/
let currentSlotScopeIds = null;
function setCurrentSlotScopeIds(scopeIds) {
	try {
		return currentSlotScopeIds;
	} finally {
		currentSlotScopeIds = scopeIds;
	}
}
const dynamicSlotsProxyHandlers = {
	get: getSlot,
	has: (target, key) => !!getSlot(target, key),
	getOwnPropertyDescriptor(target, key) {
		const slot = getSlot(target, key);
		if (slot) return {
			configurable: true,
			enumerable: true,
			value: slot
		};
	},
	ownKeys(target) {
		let keys = Object.keys(target);
		const dynamicSources = target.$;
		if (dynamicSources) {
			keys = keys.filter((k) => k !== "$");
			for (const source of dynamicSources) if (isFunction(source)) {
				const slot = resolveFunctionSource(source);
				if (slot) if (isArray(slot)) for (const s of slot) keys.push(String(s.name));
				else keys.push(String(slot.name));
			} else keys.push(...Object.keys(source));
		}
		return keys;
	},
	set: NO,
	deleteProperty: NO
};
function getSlot(target, key) {
	if (key === "$") return;
	const dynamicSources = target.$;
	if (dynamicSources) {
		let i = dynamicSources.length;
		let source;
		while (i--) {
			source = dynamicSources[i];
			if (isFunction(source)) {
				const slot = resolveFunctionSource(source);
				if (slot) {
					if (isArray(slot)) {
						for (let j = slot.length - 1; j >= 0; j--) if (String(slot[j].name) === key) return slot[j].fn;
					} else if (String(slot.name) === key) return slot.fn;
				}
			} else if (hasOwn(source, key)) return source[key];
		}
	}
	if (hasOwn(target, key)) return target[key];
}
/**
* Tracks the slot owner (the component that defines the slot content).
* This is used for:
* 1. Getting the correct rawSlots in forwarded slots (via createSlot)
* 2. Inheriting the slot owner's scopeId
*/
let currentSlotOwner = null;
function setCurrentSlotOwner(owner) {
	try {
		return currentSlotOwner;
	} finally {
		currentSlotOwner = owner;
	}
}
/**
* Get the effective slot instance for accessing rawSlots and scopeId.
* Prefers currentSlotOwner (if inside a slot), falls back to currentInstance.
*/
function getScopeOwner() {
	return currentSlotOwner || currentInstance;
}
/**
* Wrap a slot function to track the slot owner.
*
* This ensures:
* 1. createSlot gets rawSlots from the correct instance (slot owner)
* 2. elements inherit the slot owner's scopeId
*/
function withVaporCtx(fn) {
	const owner = getScopeOwner();
	return (...args) => {
		const prevOwner = setCurrentSlotOwner(owner);
		try {
			return fn(...args);
		} finally {
			setCurrentSlotOwner(prevOwner);
		}
	};
}
function createSlot(name, rawProps, fallback, noSlotted, once) {
	const _insertionParent = insertionParent;
	const _insertionAnchor = insertionAnchor;
	const _isLastInsertion = isLastInsertion;
	if (!isHydrating$1) resetInsertionState();
	const instance = getScopeOwner();
	const rawSlots = instance.rawSlots;
	const slotProps = rawProps ? new Proxy(rawProps, rawPropsProxyHandlers) : EMPTY_OBJ;
	let fragment;
	if (isRef(rawSlots._) && isInteropEnabled) {
		if (isHydrating$1) locateHydrationNode();
		fragment = instance.appContext.vapor.vdomSlot(rawSlots._, name, slotProps, instance, fallback);
	} else {
		fragment = new SlotFragment();
		fragment.forwarded = currentSlotOwner != null && currentSlotOwner !== currentInstance;
		const isDynamicName = isFunction(name);
		const slotScopeIds = [];
		if (!noSlotted) {
			const scopeId = instance.type.__scopeId;
			if (scopeId) slotScopeIds.push(`${scopeId}-s`);
		}
		const renderSlot = () => {
			const slotName = isFunction(name) ? name() : name;
			if (instance.ce || instance.parent && isAsyncWrapper(instance.parent) && instance.parent.ce) {
				const el = /* @__PURE__ */ createElement("slot");
				renderEffect(() => {
					setDynamicProps(el, [slotProps, slotName !== "default" ? { name: slotName } : {}]);
				});
				if (fallback) insert(fallback(), el);
				fragment.nodes = el;
				return;
			}
			const slot = getSlot(rawSlots, slotName);
			if (slot) {
				const boundMap = slot._boundMap || (slot._boundMap = /* @__PURE__ */ new WeakMap());
				let bound = boundMap.get(fragment);
				if (!bound) {
					bound = () => {
						const prevSlotScopeIds = setCurrentSlotScopeIds(slotScopeIds.length > 0 ? slotScopeIds : null);
						const prev = inOnceSlot;
						try {
							if (once) inOnceSlot = true;
							return slot(slotProps);
						} finally {
							inOnceSlot = prev;
							setCurrentSlotScopeIds(prevSlotScopeIds);
						}
					};
					boundMap.set(fragment, bound);
				}
				fragment.updateSlot(bound, fallback);
			} else fragment.updateSlot(void 0, fallback);
		};
		if (!once && (isDynamicName || rawSlots.$)) renderEffect(renderSlot);
		else renderSlot();
	}
	if (!isHydrating$1) {
		if (!noSlotted) {
			const scopeId = instance.type.__scopeId;
			if (scopeId) setScopeId(fragment, [`${scopeId}-s`]);
		}
		if (_insertionParent) insert(fragment, _insertionParent, _insertionAnchor);
	} else {
		if (fragment.insert) fragment.hydrate();
		if (_isLastInsertion) advanceHydrationNode(_insertionParent);
	}
	return fragment;
}
//#endregion
//#region packages/runtime-vapor/src/renderEffect.ts
var RenderEffect = class extends ReactiveEffect {
	constructor(render) {
		super();
		this.render = render;
		const instance = currentInstance;
		if (!!(process.env.NODE_ENV !== "production") && !this.subs && !isVaporComponent(instance)) warn("renderEffect called without active EffectScope or Vapor instance.");
		const job = () => {
			if (this.dirty) this.run();
		};
		this.updateJob = () => {
			instance.isUpdating = false;
			instance.u && invokeArrayFns(instance.u);
		};
		if (instance) {
			if (!!(process.env.NODE_ENV !== "production")) {
				this.onTrack = instance.rtc ? (e) => invokeArrayFns(instance.rtc, e) : void 0;
				this.onTrigger = instance.rtg ? (e) => invokeArrayFns(instance.rtg, e) : void 0;
			}
			if (instance.type.ce) (instance.renderEffects || (instance.renderEffects = [])).push(this);
			job.i = instance;
		}
		this.job = job;
		this.i = instance;
		this.flags |= 128;
		this.job.flags |= 2;
	}
	fn() {
		const instance = this.i;
		const scope = this.subs ? this.subs.sub : void 0;
		const hasUpdateHooks = instance && (instance.bu || instance.u);
		if (!!(process.env.NODE_ENV !== "production") && instance) startMeasure(instance, `renderEffect`);
		const prev = setCurrentInstance(instance, scope);
		if (hasUpdateHooks && instance.isMounted && !instance.isUpdating) {
			instance.isUpdating = true;
			instance.bu && invokeArrayFns(instance.bu);
			this.render();
			queuePostFlushCb(this.updateJob);
		} else this.render();
		setCurrentInstance(...prev);
		if (!!(process.env.NODE_ENV !== "production") && instance) startMeasure(instance, `renderEffect`);
	}
	notify() {
		if (!(this.flags & 256)) queueJob(this.job, this.i ? this.i.uid : void 0);
	}
};
function renderEffect(fn, noLifecycle = false) {
	if (inOnceSlot) return fn();
	const effect = new RenderEffect(fn);
	if (noLifecycle) effect.fn = fn;
	effect.run();
}
//#endregion
//#region packages/runtime-vapor/src/helpers/setKey.ts
function setBlockKey(block, key) {
	if (!block) return;
	if (block instanceof Node) block.$key = key;
	else if (isVaporComponent(block)) {
		block.$key = key;
		if (block.block) setBlockKey(block.block, key);
	} else if (isArray(block)) {
		if (block.length === 1) setBlockKey(block[0], key);
	} else {
		block.$key = key;
		if (block.vnode) block.vnode.key = key;
		setBlockKey(block.nodes, key);
	}
}
//#endregion
//#region packages/runtime-vapor/src/fragment.ts
var VaporFragment = class {
	constructor(nodes) {
		this.vnode = null;
		this.slotOwner = currentSlotOwner;
		this.keepAliveCtx = currentKeepAliveCtx;
		this.nodes = nodes;
	}
	runWithRenderCtx(fn) {
		const prevSlotOwner = setCurrentSlotOwner(this.slotOwner);
		const prevKeepAliveCtx = setCurrentKeepAliveCtx(this.keepAliveCtx);
		try {
			return fn();
		} finally {
			setCurrentKeepAliveCtx(prevKeepAliveCtx);
			setCurrentSlotOwner(prevSlotOwner);
		}
	}
};
var ForFragment = class extends VaporFragment {
	constructor(nodes) {
		super(nodes);
	}
};
var ForBlock = class extends VaporFragment {
	constructor(nodes, scope, item, key, index, renderKey) {
		super(nodes);
		this.scope = scope;
		this.itemRef = item;
		this.keyRef = key;
		this.indexRef = index;
		this.key = renderKey;
	}
};
var DynamicFragment = class extends VaporFragment {
	constructor(anchorLabel, keyed = false, locate = true) {
		super([]);
		this.hydrate = (isEmpty = false, isSlot = false) => {
			if (!isHydrating$1) return;
			if (this.isAnchorPending) return;
			let advanceAfterRestore = null;
			let exitHydrationBoundary;
			try {
				if (isEmpty) {
					if (isComment(currentHydrationNode, "")) {
						this.anchor = markHydrationAnchor(currentHydrationNode);
						advanceHydrationNode(currentHydrationNode);
						return;
					}
				}
				if (this.anchorLabel && !isValidBlock(this.nodes) && this.nodes instanceof Comment && isReusableDynamicFragmentAnchor(this.nodes, this.anchorLabel) && /* @__PURE__ */ parentNode(this.nodes)) {
					this.anchor = markHydrationAnchor(this.nodes);
					this.nodes = [];
					if (currentHydrationNode !== this.anchor) {
						exitHydrationBoundary = enterHydrationBoundary(this.anchor);
						advanceAfterRestore = this.anchor;
					} else advanceHydrationNode(this.anchor);
					return;
				}
				if (this.anchorLabel && !isValidBlock(this.nodes) && this.nodes instanceof Comment && !/* @__PURE__ */ parentNode(this.nodes) && currentHydrationNode) {
					const parentNode$2 = /* @__PURE__ */ parentNode(currentHydrationNode);
					const nextNode = locateNextNode(currentHydrationNode);
					if (parentNode$2) {
						this.nodes = [];
						if (nextNode) exitHydrationBoundary = enterHydrationBoundary(nextNode);
						else {
							cleanupHydrationTail(currentHydrationNode);
							setCurrentHydrationNode(null);
						}
						queuePostFlushCb(() => {
							parentNode$2.insertBefore(this.anchor = markHydrationAnchor(!!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment(this.anchorLabel) : /* @__PURE__ */ createTextNode()), nextNode);
						});
						return;
					}
				}
				if (this.anchorLabel === "if" && currentSlotEndAnchor) {
					if (currentEmptyFragment !== void 0 && (!isValidBlock(this.nodes) || currentEmptyFragment === this)) {
						const endAnchor = currentSlotEndAnchor;
						this.isAnchorPending = true;
						queuePostFlushCb(() => endAnchor.parentNode.insertBefore(this.anchor = markHydrationAnchor(!!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment(this.anchorLabel) : /* @__PURE__ */ createTextNode()), endAnchor));
						return;
					}
				}
				const forwardedSlot = this.forwarded;
				const slotAnchor = isSlot ? currentSlotEndAnchor : null;
				if (isSlot && (!forwardedSlot || isValidBlock(this.nodes)) || this.anchorLabel === "if" && isArray(this.nodes) && this.nodes.length > 1) {
					const anchor = locateHydrationBoundaryClose(slotAnchor || currentHydrationNode, slotAnchor || null);
					if (isComment(anchor, "]")) {
						this.anchor = markHydrationAnchor(anchor);
						exitHydrationBoundary = enterHydrationBoundary(anchor);
						advanceHydrationNode(anchor);
						return;
					} else if (!!(process.env.NODE_ENV !== "production")) throw new Error(`Failed to locate ${this.anchorLabel} fragment anchor. this is likely a Vue internal bug.`);
				}
				let parentNode$1;
				let nextNode;
				if (forwardedSlot) {
					const anchor = markHydrationAnchor(slotAnchor);
					parentNode$1 = anchor.parentNode;
					nextNode = anchor.nextSibling;
				} else {
					const node = findBlockNode(this.nodes);
					parentNode$1 = node.parentNode;
					nextNode = node.nextNode;
				}
				queuePostFlushCb(() => {
					const anchor = nextNode && nextNode.parentNode === parentNode$1 ? nextNode : null;
					parentNode$1.insertBefore(this.anchor = markHydrationAnchor(!!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment(this.anchorLabel) : /* @__PURE__ */ createTextNode()), anchor);
				});
			} finally {
				exitHydrationBoundary && exitHydrationBoundary();
				if (advanceAfterRestore && currentHydrationNode === advanceAfterRestore) advanceHydrationNode(advanceAfterRestore);
			}
		};
		this.keyed = keyed;
		if (isHydrating$1) {
			this.anchorLabel = anchorLabel;
			if (locate) locateHydrationNode();
		} else {
			this.anchor = !!(process.env.NODE_ENV !== "production") && anchorLabel ? /* @__PURE__ */ createComment(anchorLabel) : /* @__PURE__ */ createTextNode();
			if (!!(process.env.NODE_ENV !== "production")) this.anchorLabel = anchorLabel;
		}
	}
	update(render, key = render) {
		if (key === this.current) {
			if (isHydrating$1 && this.anchorLabel !== "slot") this.hydrate(true);
			return;
		}
		const transition = this.$transition;
		if (transition && transition.state.isLeaving) {
			this.current = key;
			this.pending = {
				render,
				key
			};
			return;
		}
		const instance = currentInstance;
		const prevSub = setActiveSub();
		const parent = isHydrating$1 ? null : this.anchor.parentNode;
		if (this.scope) {
			let retainScope = false;
			const keepAliveCtx = this.keepAliveCtx;
			const cacheKey = keepAliveCtx ? this.keyed ? withCurrentCacheKey(this.current, () => keepAliveCtx.processShapeFlag(this.nodes)) : keepAliveCtx.processShapeFlag(this.nodes) : false;
			if (cacheKey !== false) {
				keepAliveCtx.cacheScope(cacheKey, this.current, this.scope);
				retainScope = true;
			}
			if (!retainScope) this.scope.stop();
			const mode = transition && transition.mode;
			if (mode && (mode !== "in-out" || mode === "in-out" && render) && (mode !== "out-in" || isValidBlock(this.nodes))) {
				applyTransitionLeaveHooks(this.nodes, transition, () => {
					const prevInstance = setCurrentInstance(instance);
					try {
						const pending = this.pending;
						if (pending) {
							this.pending = void 0;
							this.renderBranch(pending.render, transition, parent, pending.key);
						} else this.renderBranch(render, transition, parent, key);
					} finally {
						setCurrentInstance(...prevInstance);
					}
				});
				parent && remove(this.nodes, parent);
				if (mode === "out-in") {
					setActiveSub(prevSub);
					return;
				}
			} else parent && remove(this.nodes, parent);
		}
		if (isHydrating$1 && render && this.anchorLabel !== "slot" && !isValidBlock(this.nodes)) {
			const anchor = this.anchor || (currentHydrationNode === currentSlotEndAnchor ? currentSlotEndAnchor : null);
			if (anchor) setCurrentHydrationNode(markHydrationAnchor(anchor));
		}
		this.renderBranch(render, transition, parent, key);
		setActiveSub(prevSub);
		if (isHydrating$1 && this.anchorLabel !== "slot") this.hydrate(render == null);
	}
	renderBranch(render, transition, parent, key) {
		this.current = key;
		if (render) {
			const keepAliveCtx = this.keepAliveCtx;
			const scope = keepAliveCtx && keepAliveCtx.getScope(this.current);
			if (scope) this.scope = scope;
			else this.scope = new EffectScope$1();
			const renderBranch = () => {
				try {
					this.nodes = this.runWithRenderCtx(() => this.scope.run(render) || []);
				} finally {
					const key = this.keyed ? this.current : this.$key;
					if (key !== void 0) setBlockKey(this.nodes, key);
					if (transition) this.$transition = applyTransitionHooks(this.nodes, transition);
					if (keepAliveCtx) keepAliveCtx.processShapeFlag(this.nodes);
				}
			};
			if (keepAliveCtx && this.keyed) withCurrentCacheKey(key, renderBranch);
			else renderBranch();
			if (parent) {
				if (this.attrs) {
					if (this.nodes instanceof Element) this.scope.run(() => {
						renderEffect(() => applyFallthroughProps(this.nodes, this.attrs));
					});
					else if (!!(process.env.NODE_ENV !== "production") && (this.anchorLabel === "slot" || isArray(this.nodes) && this.nodes.length)) warnExtraneousAttributes(this.attrs);
				}
				insert(this.nodes, parent, this.anchor);
				if (keepAliveCtx && transition && transition.mode === "out-in") keepAliveCtx.cacheBlock();
				if (this.onUpdated) this.onUpdated.forEach((hook) => hook(this.nodes));
			}
		} else {
			this.scope = void 0;
			this.nodes = [];
		}
	}
};
function isReusableDynamicFragmentAnchor(node, anchorLabel) {
	return isComment(node, anchorLabel) || isComment(node, "") && (anchorLabel === "dynamic-component" || anchorLabel === "async component" || anchorLabel === "keyed");
}
let currentSlotEndAnchor = null;
function setCurrentSlotEndAnchor(end) {
	try {
		return currentSlotEndAnchor;
	} finally {
		currentSlotEndAnchor = end;
	}
}
let currentEmptyFragment;
var SlotFragment = class extends DynamicFragment {
	constructor() {
		super(isHydrating$1 || !!(process.env.NODE_ENV !== "production") ? "slot" : void 0, false, false);
		this.forwarded = false;
	}
	updateSlot(render, fallback, key = render || fallback) {
		let prevEndAnchor = null;
		let pushedEndAnchor = false;
		let exitHydrationBoundary;
		let deferHydrationBoundary = false;
		if (isHydrating$1) {
			locateHydrationNode();
			if (isComment(currentHydrationNode, "[")) {
				const endAnchor = locateEndAnchor(currentHydrationNode);
				setCurrentHydrationNode(currentHydrationNode.nextSibling);
				prevEndAnchor = setCurrentSlotEndAnchor(endAnchor);
				pushedEndAnchor = true;
				exitHydrationBoundary = enterHydrationBoundary(endAnchor);
			}
		}
		try {
			if (!render || !fallback) this.update(render || fallback, key);
			else {
				const wrapped = () => {
					const prev = currentEmptyFragment;
					if (isHydrating$1) currentEmptyFragment = null;
					try {
						let block = render();
						const emptyFrag = attachSlotFallback(block, fallback);
						if (!isValidBlock(block)) {
							if (isHydrating$1 && emptyFrag instanceof DynamicFragment) currentEmptyFragment = emptyFrag;
							block = renderSlotFallback(block, fallback, emptyFrag);
						}
						return block;
					} finally {
						if (isHydrating$1) currentEmptyFragment = prev;
					}
				};
				this.update(wrapped, key);
			}
			if (isHydrating$1) {
				this.hydrate(render == null, true);
				deferHydrationBoundary = !!exitHydrationBoundary && currentEmptyFragment !== void 0 && !isValidBlock(this.nodes);
			}
		} finally {
			if (isHydrating$1) {
				if (pushedEndAnchor) setCurrentSlotEndAnchor(prevEndAnchor);
				if (deferHydrationBoundary) this.deferredHydrationBoundary = exitHydrationBoundary;
				else exitHydrationBoundary && exitHydrationBoundary();
			}
		}
	}
};
function isFragment(val) {
	return val instanceof VaporFragment;
}
function isDynamicFragment(val) {
	return val instanceof DynamicFragment;
}
function renderSlotFallback(block, fallback, emptyFragment) {
	const frag = emptyFragment || findDeepestEmptyFragment(block);
	if (frag) {
		if (frag instanceof ForFragment) frag.nodes[0] = [fallback() || []];
		else if (frag instanceof DynamicFragment) {
			frag.update(fallback);
			if (isHydrating$1 && frag instanceof SlotFragment) {
				const deferredHydrationBoundary = frag.deferredHydrationBoundary;
				if (deferredHydrationBoundary) {
					frag.deferredHydrationBoundary = void 0;
					deferredHydrationBoundary();
				}
			}
		}
		return block;
	}
	return fallback();
}
function attachSlotFallback(block, fallback) {
	const state = { emptyFrag: null };
	traverseForFallback(block, fallback, state);
	return state.emptyFrag;
}
const slotFallbackState = /* @__PURE__ */ new WeakMap();
function traverseForFallback(block, fallback, state, forOwner) {
	if (isVaporComponent(block)) {
		if (block.block) traverseForFallback(block.block, fallback, state, forOwner);
		return;
	}
	if (isArray(block)) {
		for (const item of block) traverseForFallback(item, fallback, state, forOwner);
		return;
	}
	if (block instanceof ForFragment) {
		block.fallback = chainFallback(block.fallback, fallback);
		if (!isValidBlock(block.nodes)) state.emptyFrag = block;
		traverseForFallback(block.nodes, fallback, state, block);
		return;
	}
	if (block instanceof ForBlock) {
		traverseForFallback(block.nodes, fallback, state, forOwner);
		return;
	}
	if (block instanceof VaporFragment && block.insert) {
		block.fallback = chainFallback(block.fallback, fallback);
		if (!isValidBlock(block.nodes)) state.emptyFrag = block;
		traverseForFallback(block.nodes, fallback, state, forOwner);
		return;
	}
	if (block instanceof DynamicFragment) {
		let slotState = slotFallbackState.get(block);
		if (slotState) slotState.fallback = chainFallback(slotState.fallback, fallback);
		else slotFallbackState.set(block, slotState = {
			fallback,
			wrapped: false,
			forOwner
		});
		slotState.forOwner = forOwner || slotState.forOwner;
		if (!slotState.wrapped) {
			slotState.wrapped = true;
			const original = block.update.bind(block);
			block.update = (render, key) => {
				original(render, key);
				const emptyFrag = attachSlotFallback(block.nodes, slotState.fallback);
				if (render !== slotState.fallback && !isValidBlock(block.nodes) && (!slotState.forOwner || !isValidBlock(slotState.forOwner.nodes))) renderSlotFallback(block, slotState.fallback, emptyFrag);
			};
		}
		if (!isValidBlock(block.nodes)) state.emptyFrag = block;
		traverseForFallback(block.nodes, fallback, state, forOwner);
	}
}
function findDeepestEmptyFragment(block) {
	let emptyFrag = null;
	traverseForEmptyFragment(block, (frag) => emptyFrag = frag);
	return emptyFrag;
}
function traverseForEmptyFragment(block, onFound) {
	if (isVaporComponent(block)) {
		if (block.block) traverseForEmptyFragment(block.block, onFound);
		return;
	}
	if (isArray(block)) {
		for (const item of block) traverseForEmptyFragment(item, onFound);
		return;
	}
	if (isFragment(block)) {
		if (!isValidBlock(block.nodes)) onFound(block);
		traverseForEmptyFragment(block.nodes, onFound);
	}
}
function chainFallback(existing, next) {
	if (!existing) return next;
	return (...args) => {
		const res = existing(...args);
		return !isValidBlock(res) ? next(...args) : res;
	};
}
//#endregion
//#region packages/runtime-vapor/src/components/Teleport.ts
const VaporTeleportImpl = {
	name: "VaporTeleport",
	__isTeleport: true,
	__vapor: true,
	process(props, slots) {
		return new TeleportFragment(props, slots);
	}
};
var TeleportFragment = class extends VaporFragment {
	constructor(props, slots) {
		super([]);
		this.__isTeleportFragment = true;
		this.isMounted = false;
		this.childrenInitialized = false;
		this.ownerInstance = currentInstance$1;
		this.insert = (container, anchor) => {
			if (isHydrating$1) return;
			if (!this.placeholder) this.placeholder = !!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment("teleport start") : /* @__PURE__ */ createTextNode();
			insert(this.placeholder, container, anchor);
			insert(this.anchor, container, anchor);
			this.handlePropsUpdate();
		};
		this.remove = (parent = this.parent) => {
			if (this.mountToTargetJob) {
				this.mountToTargetJob.flags |= 4;
				this.mountToTargetJob = void 0;
			}
			if (this.nodes && this.mountContainer) {
				remove(this.nodes, this.mountContainer);
				this.nodes = [];
			}
			this.isMounted = false;
			if (this.targetStart) {
				remove(this.targetStart, /* @__PURE__ */ parentNode(this.targetStart));
				this.targetStart = void 0;
			}
			if (this.targetAnchor) {
				remove(this.targetAnchor, /* @__PURE__ */ parentNode(this.targetAnchor));
				this.targetAnchor = void 0;
			}
			if (this.anchor) {
				remove(this.anchor, /* @__PURE__ */ parentNode(this.anchor));
				this.anchor = void 0;
			}
			if (this.placeholder) {
				remove(this.placeholder, parent);
				this.placeholder = void 0;
			}
			this.mountContainer = void 0;
			this.mountAnchor = void 0;
		};
		this.hydrate = () => {
			if (!isHydrating$1) return;
			const target = this.target = resolveTeleportTarget(this.resolvedProps, querySelector);
			const disabled = isTeleportDisabled(this.resolvedProps);
			this.placeholder = currentHydrationNode;
			if (target) {
				const targetNode = target._lpa || target.firstChild;
				if (disabled) this.hydrateDisabledTeleport(target, targetNode);
				else {
					this.anchor = markHydrationAnchor(locateTeleportEndAnchor(currentHydrationNode.nextSibling));
					this.mountContainer = target;
					this.hydrateTargetAnchors(target, targetNode);
					this.mountAnchor = this.targetAnchor;
					if (targetNode) setCurrentHydrationNode(targetNode.nextSibling);
					if (!this.targetAnchor) this.mountChildren(target);
					else this.initChildren();
				}
			} else if (disabled) this.hydrateDisabledTeleport(null, null);
			else this.anchor = markHydrationAnchor(locateTeleportEndAnchor(currentHydrationNode.nextSibling));
			if (target || disabled) updateCssVars(this);
			advanceHydrationNode(this.anchor);
		};
		this.rawProps = props;
		this.rawSlots = slots;
		this.parentComponent = getScopeOwner();
		this.anchor = isHydrating$1 ? void 0 : !!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment("teleport end") : /* @__PURE__ */ createTextNode();
		renderEffect(() => {
			const prevTo = this.resolvedProps && this.resolvedProps.to;
			const wasDisabled = this.isDisabled;
			this.resolvedProps = extend({}, new Proxy(this.rawProps, rawPropsProxyHandlers));
			this.isDisabled = isTeleportDisabled(this.resolvedProps);
			if (wasDisabled !== this.isDisabled || !this.isDisabled && prevTo !== this.resolvedProps.to) this.handlePropsUpdate();
		});
	}
	get parent() {
		return this.anchor ? /* @__PURE__ */ parentNode(this.anchor) : null;
	}
	initChildren() {
		const prevInstance = setCurrentInstance(this.ownerInstance);
		try {
			this.childrenInitialized = true;
			renderEffect(() => this.runWithRenderCtx(() => this.handleChildrenUpdate(this.rawSlots && this.rawSlots.default ? this.rawSlots.default() : [])));
			this.bindChildren(this.nodes);
		} finally {
			setCurrentInstance(...prevInstance);
		}
	}
	ensureChildrenInitialized() {
		if (!this.childrenInitialized) this.initChildren();
	}
	registerUpdateCssVars(block) {
		if (isFragment(block)) {
			(block.onUpdated || (block.onUpdated = [])).push(() => updateCssVars(this));
			this.registerUpdateCssVars(block.nodes);
		} else if (isVaporComponent(block)) this.registerUpdateCssVars(block.block);
		else if (isArray(block)) block.forEach((node) => this.registerUpdateCssVars(node));
	}
	bindChildren(block) {
		if (this.parentComponent && this.parentComponent.ut) this.registerUpdateCssVars(block);
		if (!!(process.env.NODE_ENV !== "production")) {
			if (isVaporComponent(block)) block.parentTeleport = this;
			else if (isArray(block)) block.forEach((node) => isVaporComponent(node) && (node.parentTeleport = this));
		}
	}
	handleChildrenUpdate(children) {
		if (isHydrating$1 || !this.parent || !this.mountContainer) {
			this.nodes = children;
			return;
		}
		remove(this.nodes, this.mountContainer);
		insert(this.nodes = children, this.mountContainer, this.mountAnchor);
		this.bindChildren(this.nodes);
		updateCssVars(this);
	}
	mount(parent, anchor) {
		if (this.$transition && !this.isMounted) applyTransitionHooks(this.nodes, this.$transition);
		if (this.isMounted) move(this.nodes, this.mountContainer = parent, this.mountAnchor = anchor, 2);
		else {
			insert(this.nodes, this.mountContainer = parent, this.mountAnchor = anchor);
			this.isMounted = true;
		}
		updateCssVars(this);
	}
	mountToTarget() {
		const target = this.target = resolveTeleportTarget(this.resolvedProps, querySelector);
		if (target) {
			if (!this.targetAnchor || /* @__PURE__ */ parentNode(this.targetAnchor) !== target) {
				if (this.targetStart) remove(this.targetStart, /* @__PURE__ */ parentNode(this.targetStart));
				if (this.targetAnchor) remove(this.targetAnchor, /* @__PURE__ */ parentNode(this.targetAnchor));
				insert(this.targetStart = /* @__PURE__ */ createTextNode(""), target);
				insert(this.targetAnchor = /* @__PURE__ */ createTextNode(""), target);
			}
			this.ensureChildrenInitialized();
			if (this.parentComponent && this.parentComponent.isCE) (this.parentComponent.ce._teleportTargets || (this.parentComponent.ce._teleportTargets = /* @__PURE__ */ new Set())).add(target);
			this.mount(target, this.targetAnchor);
		} else if (!!(process.env.NODE_ENV !== "production")) warn(`Invalid Teleport target on ${this.targetAnchor ? "update" : "mount"}:`, target, `(${typeof target})`);
	}
	clearMainViewChildren() {
		if (!this.placeholder || !this.anchor) return;
		let node = this.placeholder.nextSibling;
		while (node && node !== this.anchor) {
			const next = node.nextSibling;
			remove(node, /* @__PURE__ */ parentNode(node));
			node = next;
		}
		this.isMounted = false;
		this.mountContainer = null;
	}
	handlePropsUpdate() {
		if (!this.parent || isHydrating$1) return;
		if (this.isDisabled) {
			this.ensureChildrenInitialized();
			this.mount(this.parent, this.anchor);
		} else {
			if (this.placeholder && this.anchor && this.placeholder.nextSibling !== this.anchor) this.clearMainViewChildren();
			if (isTeleportDeferred(this.resolvedProps) || !this.parent.isConnected) {
				if (!this.mountToTargetJob || this.mountToTargetJob.flags & 4) this.mountToTargetJob = () => {
					this.mountToTargetJob = void 0;
					if (this.isDisabled || !this.anchor) return;
					this.mountToTarget();
				};
				queuePostFlushCb(this.mountToTargetJob);
			} else this.mountToTarget();
		}
	}
	hydrateTargetAnchors(target, targetNode) {
		let targetAnchor = targetNode;
		while (targetAnchor) {
			if (targetAnchor.nodeType === 8) {
				if (targetAnchor.data === "teleport start anchor") this.targetStart = targetAnchor;
				else if (targetAnchor.data === "teleport anchor") {
					this.targetAnchor = markHydrationAnchor(targetAnchor);
					target._lpa = this.targetAnchor.nextSibling;
					break;
				}
			}
			targetAnchor = targetAnchor.nextSibling;
		}
	}
	hydrateDisabledTeleport(target, targetNode) {
		if (!isHydrating$1) return;
		let nextNode = this.placeholder.nextSibling;
		setCurrentHydrationNode(nextNode);
		this.mountAnchor = this.anchor = markHydrationAnchor(locateTeleportEndAnchor(nextNode));
		this.mountContainer = /* @__PURE__ */ parentNode(this.anchor);
		if (target) this.hydrateTargetAnchors(target, targetNode);
		else {
			this.targetStart = targetNode;
			this.targetAnchor = targetNode && targetNode.nextSibling;
		}
		this.initChildren();
	}
	mountChildren(target) {
		if (!isHydrating$1) return;
		target.appendChild(this.targetStart = /* @__PURE__ */ createTextNode(""));
		target.appendChild(this.mountAnchor = this.targetAnchor = markHydrationAnchor(/* @__PURE__ */ createTextNode("")));
		if (!isMismatchAllowed(target, 1)) {
			if (!!(process.env.NODE_ENV !== "production") || __VUE_PROD_HYDRATION_MISMATCH_DETAILS__) warn(`Hydration children mismatch on`, target, `\nServer rendered element contains fewer child nodes than client nodes.`);
			logMismatchError();
		}
		runWithoutHydration(this.initChildren.bind(this));
	}
};
const VaporTeleport = VaporTeleportImpl;
/**
* Use duck typing to check for VaporTeleport instead of direct reference
* to VaporTeleportImpl, allowing tree-shaking when Teleport is not used.
*/
function isVaporTeleport(value) {
	return !!(value && value.__isTeleport && value.__vapor);
}
/**
* Use duck typing to check for TeleportFragment instead of instanceof,
* allowing tree-shaking when Teleport is not used.
*/
function isTeleportFragment(value) {
	return !!(value && value.__isTeleportFragment);
}
function locateTeleportEndAnchor(node = currentHydrationNode) {
	let depth = 0;
	while (node) {
		if (isComment(node, "teleport start")) depth++;
		else if (isComment(node, "teleport end")) if (depth === 0) return node;
		else depth--;
		node = node.nextSibling;
	}
	return null;
}
function updateCssVars(frag) {
	const ctx = frag.parentComponent;
	if (ctx && ctx.ut) {
		let node, anchor;
		if (frag.isDisabled) {
			node = frag.placeholder;
			anchor = frag.anchor;
		} else {
			node = frag.targetStart;
			anchor = frag.targetAnchor;
		}
		while (node && node !== anchor) {
			if (node.nodeType === 1) node.setAttribute("data-v-owner", String(ctx.uid));
			node = node.nextSibling;
		}
		pauseTracking();
		try {
			ctx.ut();
		} finally {
			resetTracking();
		}
	}
}
//#endregion
//#region packages/runtime-vapor/src/block.ts
function isBlock(val) {
	return val instanceof Node || isArray(val) || isVaporComponent(val) || isFragment(val);
}
function isValidBlock(block) {
	if (block instanceof Node) return !(block instanceof Comment);
	else if (isVaporComponent(block)) return isValidBlock(block.block);
	else if (isArray(block)) return block.length > 0 && block.some(isValidBlock);
	else return isValidBlock(block.nodes);
}
function insert(block, parent, anchor = null, parentSuspense) {
	anchor = anchor === 0 ? parent.$fc || /* @__PURE__ */ _child(parent) : anchor;
	if (block instanceof Node) {
		if (!isHydrating$1) if (block instanceof Element && block.$transition && !block.$transition.disabled) performTransitionEnter(block, block.$transition, () => parent.insertBefore(block, anchor), parentSuspense);
		else parent.insertBefore(block, anchor);
	} else if (isVaporComponent(block)) if (block.isMounted && !block.isDeactivated) insert(block.block, parent, anchor);
	else mountComponent(block, parent, anchor);
	else if (isArray(block)) for (const b of block) insert(b, parent, anchor);
	else {
		if (block.anchor) {
			insert(block.anchor, parent, anchor);
			anchor = block.anchor;
		}
		if (block.insert) block.insert(parent, anchor, block.$transition);
		else insert(block.nodes, parent, anchor, parentSuspense);
	}
}
function move(block, parent, anchor = null, moveType = 1, parentComponent, parentSuspense) {
	anchor = anchor === 0 ? parent.$fc || /* @__PURE__ */ _child(parent) : anchor;
	if (block instanceof Node) if (block instanceof Element && block.$transition && !block.$transition.disabled && moveType !== 2) if (moveType === 0) performTransitionEnter(block, block.$transition, () => parent.insertBefore(block, anchor), parentSuspense, true);
	else performTransitionLeave(block, block.$transition, () => {
		if (moveType === 1 && parentComponent && parentComponent.isUnmounted) block.remove();
		else parent.insertBefore(block, anchor);
	}, parentSuspense, true);
	else parent.insertBefore(block, anchor);
	else if (isVaporComponent(block)) if (block.isMounted) move(block.block, parent, anchor, moveType, parentComponent, parentSuspense);
	else mountComponent(block, parent, anchor);
	else if (isArray(block)) for (const b of block) move(b, parent, anchor, moveType, parentComponent, parentSuspense);
	else {
		if (block.anchor) {
			move(block.anchor, parent, anchor, moveType, parentComponent, parentSuspense);
			anchor = block.anchor;
		}
		if (block.insert) block.insert(parent, anchor, block.$transition);
		else move(block.nodes, parent, anchor, moveType, parentComponent, parentSuspense);
	}
}
function prepend(parent, ...blocks) {
	let i = blocks.length;
	while (i--) insert(blocks[i], parent, 0);
}
function remove(block, parent) {
	if (block instanceof Node) if (block.$transition && block instanceof Element) performTransitionLeave(block, block.$transition, () => parent && parent.removeChild(block));
	else parent && parent.removeChild(block);
	else if (isVaporComponent(block)) unmountComponent(block, parent);
	else if (isArray(block)) for (let i = 0; i < block.length; i++) remove(block[i], parent);
	else {
		if (block.remove) block.remove(parent, block.$transition);
		else remove(block.nodes, parent);
		if (block.anchor) remove(block.anchor, parent);
		if (block.scope) block.scope.stop();
	}
}
/**
* dev / test only
*/
function normalizeBlock(block) {
	if (!!!(process.env.NODE_ENV !== "production") && true) throw new Error("normalizeBlock should not be used in production code paths");
	const nodes = [];
	if (block instanceof Node) nodes.push(block);
	else if (isArray(block)) block.forEach((child) => nodes.push(...normalizeBlock(child)));
	else if (isVaporComponent(block)) nodes.push(...normalizeBlock(block.block));
	else if (isTeleportFragment(block)) nodes.push(block.placeholder, block.anchor);
	else {
		nodes.push(...normalizeBlock(block.nodes));
		block.anchor && nodes.push(block.anchor);
	}
	return nodes;
}
function findBlockNode(block) {
	const lastChild = findLastChild(block);
	let { parentNode, nextSibling: nextNode } = lastChild;
	if (nextNode && isComment(nextNode, "]") && isFragmentBlock(block) && !isComment(lastChild, "]")) nextNode = nextNode.nextSibling;
	return {
		parentNode,
		nextNode
	};
}
function findLastChild(node) {
	if (node && node instanceof Node) return node;
	else if (isArray(node)) return findLastChild(node[node.length - 1]);
	else if (isVaporComponent(node)) return findLastChild(node.block);
	else {
		if (node.anchor) return node.anchor;
		return findLastChild(node.nodes);
	}
}
function isFragmentBlock(block) {
	if (isArray(block)) return true;
	else if (isVaporComponent(block)) return isFragmentBlock(block.block);
	else if (isFragment(block)) return isFragmentBlock(block.nodes);
	return false;
}
//#endregion
//#region packages/runtime-vapor/src/hmr.ts
function hmrRerender(instance) {
	const normalized = normalizeBlock(instance.block);
	const parent = normalized[0].parentNode;
	const anchor = normalized[normalized.length - 1].nextSibling;
	instance.scope.reset();
	remove(instance.block, parent);
	const prev = setCurrentInstance(instance);
	pushWarningContext(instance);
	devRender(instance);
	popWarningContext();
	setCurrentInstance(...prev);
	insert(instance.block, parent, anchor);
}
function hmrReload(instance, newComp) {
	if (instance.parent && isKeepAlive(instance.parent)) {
		instance.parent.hmrRerender();
		return;
	}
	const normalized = normalizeBlock(instance.block);
	const parent = normalized[0].parentNode;
	const anchor = normalized[normalized.length - 1].nextSibling;
	unmountComponent(instance, parent);
	const parentInstance = instance.parent;
	const prev = setCurrentInstance(parentInstance);
	const newInstance = createComponent(newComp, instance.rawProps, instance.rawSlots, instance.isSingleRoot, void 0, instance.appContext);
	setCurrentInstance(...prev);
	mountComponent(newInstance, parent, anchor);
	updateParentBlockOnHmrReload(parentInstance, instance, newInstance);
	updateParentTeleportOnHmrReload(instance, newInstance);
}
/**
* dev only
* update parentInstance.block to ensure that the correct parent and
* anchor are found during parentInstance HMR rerender/reload, as
* `normalizeBlock` relies on the current instance.block
*/
function updateParentBlockOnHmrReload(parentInstance, instance, newInstance) {
	if (parentInstance) parentInstance.block = replaceBlockInstance(parentInstance.block, instance, newInstance);
}
/**
* dev only
* during root component HMR reload, since the old component will be unmounted
* and a new one will be mounted, we need to update the teleport's nodes
* to ensure that the correct parent and anchor are found during parentInstance
* HMR rerender/reload, as `normalizeBlock` relies on the current instance.block
*/
function updateParentTeleportOnHmrReload(instance, newInstance) {
	const teleport = instance.parentTeleport;
	if (teleport) {
		newInstance.parentTeleport = teleport;
		teleport.nodes = replaceBlockInstance(teleport.nodes, instance, newInstance);
	}
}
function replaceBlockInstance(block, instance, newInstance) {
	if (block === instance) return newInstance;
	if (isArray(block)) {
		for (let i = 0; i < block.length; i++) block[i] = replaceBlockInstance(block[i], instance, newInstance);
		return block;
	}
	if (isVaporComponent(block)) {
		block.block = replaceBlockInstance(block.block, instance, newInstance);
		return block;
	}
	if (isFragment(block)) {
		block.nodes = replaceBlockInstance(block.nodes, instance, newInstance);
		return block;
	}
	return block;
}
//#endregion
//#region packages/runtime-vapor/src/component.ts
function createComponent(component, rawProps, rawSlots, isSingleRoot, once, appContext = currentInstance && currentInstance.appContext || emptyContext, managedMount = false) {
	const _insertionParent = insertionParent;
	const _insertionAnchor = insertionAnchor;
	const _isLastInsertion = isLastInsertion;
	let hydrationClose = null;
	let exitHydrationBoundary;
	let deferHydrationBoundary = false;
	const finalizeHydrationBoundary = () => {
		exitHydrationBoundary && exitHydrationBoundary();
		if (hydrationClose && currentHydrationNode === hydrationClose) advanceHydrationNode(hydrationClose);
	};
	if (isHydrating$1) {
		locateHydrationNode();
		if (component.__multiRoot && isComment(currentHydrationNode, "[")) {
			hydrationClose = locateEndAnchor(currentHydrationNode);
			exitHydrationBoundary = enterHydrationBoundary(hydrationClose && markHydrationAnchor(hydrationClose));
			setCurrentHydrationNode(currentHydrationNode.nextSibling);
		}
	} else resetInsertionState();
	try {
		let prevSuspense = null;
		if (currentInstance && currentInstance.suspense) prevSuspense = setParentSuspense(currentInstance.suspense);
		if ((isSingleRoot || currentInstance && isVaporTransition(currentInstance.type)) && component.inheritAttrs !== false && isVaporComponent(currentInstance) && currentInstance.hasFallthrough) {
			const attrs = currentInstance.attrs;
			if (rawProps && rawProps !== EMPTY_OBJ) (rawProps.$ || (rawProps.$ = [])).push(() => attrs);
			else rawProps = { $: [() => attrs] };
		}
		if (currentInstance && currentInstance.vapor && isKeepAlive(currentInstance)) {
			const cached = currentInstance.ctx.getCachedComponent(component);
			if (cached) return cached;
		}
		if (isInteropEnabled && appContext.vapor && !component.__vapor) {
			const frag = appContext.vapor.vdomMount(component, currentInstance, rawProps, rawSlots);
			if (!isHydrating$1) {
				if (_insertionParent) insert(frag, _insertionParent, _insertionAnchor);
			} else {
				frag.hydrate();
				if (_isLastInsertion) advanceHydrationNode(_insertionParent);
			}
			return frag;
		}
		if (isVaporTeleport(component)) {
			const frag = component.process(rawProps, rawSlots);
			if (!isHydrating$1) {
				if (_insertionParent) insert(frag, _insertionParent, _insertionAnchor);
			} else {
				frag.hydrate();
				if (_isLastInsertion) advanceHydrationNode(_insertionParent);
			}
			return frag;
		}
		const instance = new VaporComponentInstance(component, rawProps, rawSlots, appContext, once);
		if (currentKeepAliveCtx && !isAsyncWrapper(instance)) {
			currentKeepAliveCtx.processShapeFlag(instance);
			setCurrentKeepAliveCtx(null);
		}
		const prevSlotOwner = setCurrentSlotOwner(null);
		if (!!(process.env.NODE_ENV !== "production")) {
			registerHMR(instance);
			instance.isSingleRoot = isSingleRoot;
			instance.hmrRerender = hmrRerender.bind(null, instance);
			instance.hmrReload = hmrReload.bind(null, instance);
			pushWarningContext(instance);
			startMeasure(instance, `init`);
			instance.propsOptions = normalizePropsOptions(component);
			instance.emitsOptions = normalizeEmitsOptions(component);
		}
		if (isHydrating$1 && isAsyncWrapper(instance) && component.__asyncHydrate && !component.__asyncResolved) component.__asyncHydrate(currentHydrationNode, instance, () => setupComponent(instance, component));
		else setupComponent(instance, component);
		if (!!(process.env.NODE_ENV !== "production")) {
			popWarningContext();
			endMeasure(instance, "init");
		}
		if (currentInstance && currentInstance.suspense) setParentSuspense(prevSuspense);
		setCurrentSlotOwner(prevSlotOwner);
		onScopeDispose$1(() => unmountComponent(instance), true);
		if (!managedMount && (_insertionParent || isHydrating$1)) mountComponent(instance, _insertionParent, _insertionAnchor);
		if (isHydrating$1 && _insertionAnchor !== void 0) advanceHydrationNode(_insertionParent);
		if (isHydrating$1 && hydrationClose && instance.suspense && instance.asyncDep && !instance.asyncResolved && instance.restoreAsyncContext) {
			deferHydrationBoundary = true;
			instance.deferredHydrationBoundary = () => {
				if (instance.block && hydrationClose && findBlockNode(instance.block).nextNode === hydrationClose.nextSibling) setCurrentHydrationNode(hydrationClose);
				finalizeHydrationBoundary();
			};
		}
		return instance;
	} finally {
		if (isHydrating$1 && !deferHydrationBoundary) finalizeHydrationBoundary();
	}
}
function setupComponent(instance, component) {
	const prevInstance = setCurrentInstance(instance);
	const prevSub = setActiveSub();
	if (!!(process.env.NODE_ENV !== "production")) setupPropsValidation(instance);
	const setupFn = isFunction(component) ? component : component.setup;
	const setupResult = setupFn ? callWithErrorHandling(setupFn, instance, 0, [instance.props, instance]) || EMPTY_OBJ : EMPTY_OBJ;
	const isAsyncSetup = isPromise(setupResult);
	if ((isAsyncSetup || instance.sp) && !isAsyncWrapper(instance)) markAsyncBoundary(instance);
	if (isAsyncSetup) {
		instance.asyncDep = setupResult;
		if (!!(process.env.NODE_ENV !== "production") && !instance.suspense) {
			var _getComponentName;
			warn(`Component <${(_getComponentName = getComponentName(component)) !== null && _getComponentName !== void 0 ? _getComponentName : "Anonymous"}>: setup function returned a promise, but no <Suspense> boundary was found in the parent component tree. A component with async setup() must be nested in a <Suspense> in order to be rendered.`);
		}
	} else handleSetupResult(setupResult, component, instance);
	setActiveSub(prevSub);
	setCurrentInstance(...prevInstance);
}
let isApplyingFallthroughProps = false;
function applyFallthroughProps(el, attrs) {
	isApplyingFallthroughProps = true;
	setDynamicProps(el, [attrs]);
	isApplyingFallthroughProps = false;
}
/**
* dev only
*/
function createDevSetupStateProxy(instance) {
	const { setupState } = instance;
	return new Proxy(setupState, { get(target, key, receiver) {
		if (isString(key) && !key.startsWith("__v") && !hasOwn(toRaw(setupState), key)) warn(`Property ${JSON.stringify(key)} was accessed during render but is not defined on instance.`);
		return Reflect.get(target, key, receiver);
	} });
}
/**
* dev only
*/
function devRender(instance) {
	instance.block = (instance.type.render ? callWithErrorHandling(instance.type.render, instance, 1, [
		instance.setupState,
		instance.props,
		instance.emit,
		instance.attrs,
		instance.slots
	]) : callWithErrorHandling(isFunction(instance.type) ? instance.type : instance.type.setup, instance, 0, [instance.props, {
		slots: instance.slots,
		attrs: instance.attrs,
		emit: instance.emit,
		expose: instance.expose
	}])) || [];
}
const emptyContext = {
	app: null,
	config: {},
	provides: /* @__PURE__ */ Object.create(null)
};
var VaporComponentInstance = class {
	constructor(comp, rawProps, rawSlots, appContext, once) {
		this.accessedAttrs = false;
		this.vapor = true;
		this.uid = nextUid();
		this.type = comp;
		this.parent = currentInstance;
		if (currentInstance) {
			this.root = currentInstance.root;
			this.appContext = currentInstance.appContext;
			this.provides = currentInstance.provides;
			this.ids = currentInstance.ids;
		} else {
			this.root = this;
			this.appContext = appContext || emptyContext;
			this.provides = Object.create(this.appContext.provides);
			this.ids = [
				"",
				0,
				0
			];
		}
		this.block = null;
		this.scope = new EffectScope(true);
		this.emit = emit.bind(null, this);
		this.expose = expose.bind(null, this);
		this.refs = EMPTY_OBJ;
		this.emitted = this.exposed = this.exposeProxy = this.propsDefaults = null;
		this.suspense = parentSuspense;
		this.suspenseId = parentSuspense ? parentSuspense.pendingId : 0;
		this.asyncDep = null;
		this.asyncResolved = false;
		this.isMounted = this.isUnmounted = this.isUpdating = this.isDeactivated = false;
		this.rawProps = rawProps || EMPTY_OBJ;
		this.hasFallthrough = hasFallthroughAttrs(comp, rawProps);
		if (rawProps || comp.props) {
			const [propsHandlers, attrsHandlers] = getPropsProxyHandlers(comp, once);
			this.attrs = new Proxy(this, attrsHandlers);
			this.props = comp.props ? new Proxy(this, propsHandlers) : isFunction(comp) ? this.attrs : EMPTY_OBJ;
		} else this.props = this.attrs = EMPTY_OBJ;
		this.rawSlots = rawSlots || EMPTY_OBJ;
		this.slots = rawSlots ? rawSlots.$ ? new Proxy(rawSlots, dynamicSlotsProxyHandlers) : rawSlots : EMPTY_OBJ;
		this.scopeId = getCurrentScopeId();
		if (comp.ce) comp.ce(this);
		if (!!(process.env.NODE_ENV !== "production")) if (this.props === this.attrs) this.accessedAttrs = true;
		else {
			const attrs = this.attrs;
			const instance = this;
			this.attrs = new Proxy(attrs, { get(target, key, receiver) {
				instance.accessedAttrs = true;
				return Reflect.get(target, key, receiver);
			} });
		}
	}
	/**
	* Expose `getKeysFromRawProps` on the instance so it can be used in code
	* paths where it's needed, e.g. `useModel`
	*/
	rawKeys() {
		return getKeysFromRawProps(this.rawProps);
	}
};
function isVaporComponent(value) {
	return value instanceof VaporComponentInstance;
}
/**
* Used when a component cannot be resolved at compile time
* and needs rely on runtime resolution - where it might fallback to a plain
* element if the resolution fails.
*/
function createComponentWithFallback(comp, rawProps, rawSlots, isSingleRoot, once, appContext) {
	if (comp === NULL_DYNAMIC_COMPONENT) {
		if (isHydrating$1 && currentHydrationNode) {
			if (isReusableNullComponentAnchor(currentHydrationNode)) {
				const node = currentHydrationNode;
				if (isComment(node, "")) advanceHydrationNode(node);
				return node;
			}
			const nextAnchor = locateNextNode(currentHydrationNode);
			if (nextAnchor && isReusableNullComponentAnchor(nextAnchor)) return nextAnchor;
		}
		return !!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment("ndc") : /* @__PURE__ */ createTextNode("");
	}
	if (!isString(comp)) return createComponent(comp, rawProps, rawSlots, isSingleRoot, once, appContext);
	return createPlainElement(comp, rawProps, rawSlots, isSingleRoot, once);
}
function isReusableNullComponentAnchor(node) {
	return isComment(node, "") || isComment(node, "dynamic-component") || isComment(node, "async component") || isComment(node, "keyed");
}
function createPlainElement(comp, rawProps, rawSlots, isSingleRoot, once) {
	const _insertionParent = insertionParent;
	const _insertionAnchor = insertionAnchor;
	const _isLastInsertion = isLastInsertion;
	if (isHydrating$1) locateHydrationNode();
	else resetInsertionState();
	const el = isHydrating$1 ? adoptTemplate(currentHydrationNode, `<${comp}/>`) : /* @__PURE__ */ createElement(comp);
	el.$root = isSingleRoot;
	if (!isHydrating$1) {
		const scopeId = getCurrentScopeId();
		if (scopeId) setScopeId(el, [scopeId]);
	}
	if (rawProps) {
		const setFn = () => setDynamicProps(el, [resolveDynamicProps(rawProps)]);
		if (once) setFn();
		else renderEffect(setFn);
	}
	if (rawSlots) {
		let nextNode = null;
		if (isHydrating$1) {
			nextNode = locateNextNode(el);
			setCurrentHydrationNode(el.firstChild);
		}
		if (rawSlots.$) {
			const frag = new DynamicFragment(isHydrating$1 ? "" : !!(process.env.NODE_ENV !== "production") ? "slot" : void 0);
			renderEffect(() => frag.update(getSlot(rawSlots, "default")));
			if (!isHydrating$1) insert(frag, el);
		} else {
			let slot = getSlot(rawSlots, "default");
			if (slot) {
				const block = slot();
				if (!isHydrating$1) insert(block, el);
			}
		}
		if (isHydrating$1) setCurrentHydrationNode(nextNode);
	}
	if (!isHydrating$1) {
		if (_insertionParent) insert(el, _insertionParent, _insertionAnchor);
	} else if (_isLastInsertion) advanceHydrationNode(_insertionParent);
	return el;
}
function mountComponent(instance, parent, anchor) {
	if (instance.suspense && instance.asyncDep && !instance.asyncResolved) {
		const component = instance.type;
		instance.suspense.registerDep(instance, (setupResult) => {
			const reset = instance.restoreAsyncContext && instance.restoreAsyncContext();
			try {
				handleSetupResult(setupResult, component, instance);
				mountComponent(instance, parent, anchor);
				if (isHydrating$1) instance.deferredHydrationBoundary && instance.deferredHydrationBoundary();
			} finally {
				if (isHydrating$1) instance.deferredHydrationBoundary = void 0;
				instance.restoreAsyncContext = void 0;
				if (reset) reset();
			}
		});
		return;
	}
	if (instance.shapeFlag & 512) {
		instance.parent.ctx.activate(instance, parent, anchor);
		return;
	}
	const { root, type } = instance;
	if (root && root.ce && root.ce._hasShadowRoot()) root.ce._injectChildStyle(type, instance.parent ? instance.parent.type : void 0);
	if (!!(process.env.NODE_ENV !== "production")) startMeasure(instance, `mount`);
	if (instance.bm) invokeArrayFns(instance.bm);
	if (!isHydrating$1) {
		insert(instance.block, parent, anchor);
		setComponentScopeId(instance);
	}
	if (instance.m) queuePostFlushCb(instance.m);
	if (instance.shapeFlag & 256 && instance.a) queuePostFlushCb(instance.a);
	instance.isMounted = true;
	if (!!(process.env.NODE_ENV !== "production")) endMeasure(instance, `mount`);
}
function unmountComponent(instance, parentNode) {
	if (instance.shapeFlag & 256 && instance.parent && instance.parent.vapor && instance.parent.ctx) {
		if (parentNode) instance.parent.ctx.deactivate(instance);
		return;
	}
	if (instance.isMounted && !instance.isUnmounted) {
		if (!!(process.env.NODE_ENV !== "production")) unregisterHMR(instance);
		invalidateMount(instance.m);
		invalidateMount(instance.a);
		if (instance.bum) invokeArrayFns(instance.bum);
		instance.scope.stop();
		if (instance.um) queuePostFlushCb(instance.um);
		instance.isUnmounted = true;
	}
	if (parentNode) remove(instance.block, parentNode);
}
function getExposed(instance) {
	if (instance.exposed) return instance.exposeProxy || (instance.exposeProxy = new Proxy(markRaw(instance.exposed), { get: (target, key) => unref(target[key]) }));
}
function getRootElement(block, onDynamicFragment, recurse = true) {
	if (block instanceof Element) return block;
	if (recurse && isVaporComponent(block)) return getRootElement(block.block, onDynamicFragment, recurse);
	if (isFragment(block) && !isTeleportFragment(block)) {
		if (block instanceof DynamicFragment && onDynamicFragment) onDynamicFragment(block);
		const { nodes } = block;
		if (nodes instanceof Element && nodes.$root) return nodes;
		return getRootElement(nodes, onDynamicFragment, recurse);
	}
	if (isArray(block)) {
		let singleRoot;
		let hasComment = false;
		for (const b of block) {
			if (b instanceof Comment) {
				hasComment = true;
				continue;
			}
			const thisRoot = getRootElement(b, onDynamicFragment, recurse);
			if (!thisRoot || singleRoot) return;
			singleRoot = thisRoot;
		}
		return hasComment ? singleRoot : void 0;
	}
}
function isVaporTransition(component) {
	return getComponentName(component) === "VaporTransition";
}
function handleSetupResult(setupResult, component, instance) {
	if (!!(process.env.NODE_ENV !== "production")) pushWarningContext(instance);
	if (!!(process.env.NODE_ENV !== "production") && !isBlock(setupResult)) if (isFunction(component)) {
		warn(`Functional vapor component must return a block directly.`);
		instance.block = [];
	} else if (!component.render) {
		warn(`Vapor component setup() returned non-block value, and has no render function.`);
		instance.block = [];
	} else {
		if (!!(process.env.NODE_ENV !== "production") || __VUE_PROD_DEVTOOLS__) instance.devtoolsRawSetupState = setupResult;
		instance.setupState = proxyRefs(setupResult);
		if (!!(process.env.NODE_ENV !== "production")) instance.setupState = createDevSetupStateProxy(instance);
		devRender(instance);
	}
	else if (setupResult === EMPTY_OBJ && component.render) instance.block = callWithErrorHandling(component.render, instance, 1);
	else instance.block = setupResult;
	if (instance.hasFallthrough && component.inheritAttrs !== false && Object.keys(instance.attrs).length) {
		const root = getRootElement(instance.block, (frag) => frag.attrs = instance.attrs, false);
		if (root) renderEffect(() => {
			const attrs = isFunction(component) && !isVaporTransition(component) ? getFunctionalFallthrough(instance.attrs) : instance.attrs;
			if (attrs) applyFallthroughProps(root, attrs);
		});
		else if (!!(process.env.NODE_ENV !== "production") && (!instance.accessedAttrs && isArray(instance.block) && instance.block.length || isTeleportFragment(instance.block))) warnExtraneousAttributes(instance.attrs);
	}
	if (!!(process.env.NODE_ENV !== "production")) popWarningContext();
}
function getCurrentScopeId() {
	const scopeOwner = getScopeOwner();
	return scopeOwner ? scopeOwner.type.__scopeId : void 0;
}
//#endregion
//#region packages/runtime-vapor/src/apiCreateApp.ts
let _createApp;
const mountApp = (app, container) => {
	optimizePropertyLookup();
	if (container.nodeType === 1) {
		if (!!(process.env.NODE_ENV !== "production") && container.childNodes.length) warn("mount target container is not empty and will be cleared.");
		container.textContent = "";
	}
	const instance = app._ceComponent || createComponent(app._component, app._props, null, false, false, app._context);
	mountComponent(instance, container);
	flushOnAppMount();
	return instance;
};
let _hydrateApp;
const hydrateApp = (app, container) => {
	optimizePropertyLookup();
	let instance;
	withHydration(container, () => {
		instance = app._ceComponent || createComponent(app._component, app._props, null, false, false, app._context, true);
		mountComponent(instance, container);
		flushOnAppMount();
	});
	return instance;
};
const unmountApp = (app) => {
	unmountComponent(app._instance, app._container);
};
function prepareApp() {
	initFeatureFlags();
	const target = getGlobalThis();
	target.__VUE__ = true;
	if (!!(process.env.NODE_ENV !== "production") || __VUE_PROD_DEVTOOLS__) setDevtoolsHook(target.__VUE_DEVTOOLS_GLOBAL_HOOK__, target);
}
function postPrepareApp(app) {
	app.vapor = true;
	const mount = app.mount;
	app.mount = (container, ...args) => {
		container = normalizeContainer(container);
		const proxy = mount(container, ...args);
		if (container instanceof Element) {
			container.removeAttribute("v-cloak");
			container.setAttribute("data-v-app", "");
		}
		return proxy;
	};
}
const createVaporApp = (comp, props) => {
	prepareApp();
	if (!_createApp) _createApp = createAppAPI(mountApp, unmountApp, getExposed);
	const app = _createApp(comp, props);
	postPrepareApp(app);
	return app;
};
const createVaporSSRApp = (comp, props) => {
	setIsHydratingEnabled(true);
	prepareApp();
	if (!_hydrateApp) _hydrateApp = createAppAPI(hydrateApp, unmountApp, getExposed);
	const app = _hydrateApp(comp, props);
	postPrepareApp(app);
	return app;
};
//#endregion
//#region packages/runtime-vapor/src/apiDefineAsyncComponent.ts
function defineVaporAsyncComponent(source) {
	const { load, getResolvedComp, setPendingRequest, source: { loadingComponent, errorComponent, delay, hydrate: hydrateStrategy, timeout, suspensible = true } } = createAsyncComponentContext(source);
	return /* @__PURE__ */ defineVaporComponent({
		name: "VaporAsyncComponentWrapper",
		__asyncLoader: load,
		__asyncHydrate(el, instance, hydrate) {
			if (!isHydrating$1) return;
			if (isComment(el, "[")) {
				const end = /* @__PURE__ */ _next(locateEndAnchor(el));
				const block = instance.block = [el];
				let cur = el;
				while (true) {
					let n = /* @__PURE__ */ _next(cur);
					if (n && n !== end) block.push(cur = n);
					else break;
				}
			} else instance.block = el;
			instance.isMounted = true;
			setCurrentHydrationNode(isComment(el, "[") ? locateEndAnchor(el) : el.nextSibling);
			performAsyncHydrate(el, instance, () => hydrateNode(el, hydrate), getResolvedComp, load, hydrateStrategy);
		},
		get __asyncResolved() {
			return getResolvedComp();
		},
		setup() {
			const instance = currentInstance;
			markAsyncBoundary(instance);
			const frag = !!(process.env.NODE_ENV !== "production") || isHydrating$1 ? new DynamicFragment("async component") : new DynamicFragment();
			let resolvedComp = getResolvedComp();
			if (resolvedComp) {
				frag.update(() => createInnerComp(resolvedComp, instance));
				return frag;
			}
			const onError = (err) => {
				setPendingRequest(null);
				handleError(err, instance, 13, !errorComponent);
			};
			if (suspensible && instance.suspense) return load().then(() => {
				resolvedComp = getResolvedComp();
				if (resolvedComp) frag.update(() => createInnerComp(resolvedComp, instance));
				return frag;
			}).catch((err) => {
				onError(err);
				if (errorComponent) frag.update(() => createInnerComp(errorComponent, instance, { error: () => err }, {}));
				return frag;
			});
			const { loaded, error, delayed } = useAsyncComponentState(delay, timeout, onError);
			load().then(() => {
				loaded.value = true;
			}).catch((err) => {
				onError(err);
				error.value = err;
			});
			renderEffect(() => {
				resolvedComp = getResolvedComp();
				let render;
				if (loaded.value && resolvedComp) render = () => createInnerComp(resolvedComp, instance);
				else if (error.value && errorComponent) render = () => createComponent(errorComponent, { error: () => error.value });
				else if (loadingComponent && !delayed.value) render = () => createComponent(loadingComponent);
				frag.update(render);
				if (frag.keepAliveCtx) frag.keepAliveCtx.cacheBlock();
			});
			return frag;
		}
	});
}
function createInnerComp(comp, parent, rawProps = parent.rawProps, rawSlots = parent.rawSlots) {
	const prevInstance = setCurrentInstance(parent);
	try {
		return createComponent(comp, rawProps, rawSlots, void 0, void 0, parent.appContext);
	} finally {
		setCurrentInstance(...prevInstance);
	}
}
//#endregion
//#region packages/runtime-vapor/src/apiDefineCustomElement.ts
const vaporCustomElementHydrates = /* @__PURE__ */ new WeakMap();
/* @__NO_SIDE_EFFECTS__ */
function defineVaporCustomElement(options, extraOptions, _createApp, _hydrate) {
	let Comp = /* @__PURE__ */ defineVaporComponent(options, extraOptions);
	if (isPlainObject(Comp)) Comp = extend({}, Comp, extraOptions);
	class VaporCustomElement extends VaporElement {
		constructor(initialProps) {
			super(Comp, initialProps, _createApp);
		}
	}
	VaporCustomElement.def = Comp;
	if (_hydrate) vaporCustomElementHydrates.set(VaporCustomElement, _hydrate);
	return VaporCustomElement;
}
const defineVaporSSRCustomElement = ((options, extraOptions) => {
	return /* @__PURE__ */ defineVaporCustomElement(options, extraOptions, createVaporSSRApp, withHydration);
});
var VaporElement = class extends VueElementBase {
	constructor(def, props = {}, createAppFn = createVaporApp) {
		super(def, props, createAppFn);
	}
	_needsHydration() {
		const hydrate = vaporCustomElementHydrates.get(this.constructor);
		if (this.shadowRoot && hydrate) return true;
		else if (!!(process.env.NODE_ENV !== "production") && this.shadowRoot) warn("Custom element has pre-rendered declarative shadow root but is not defined as hydratable. Use `defineVaporSSRCustomElement`.");
		return false;
	}
	_mount(def) {
		if ((!!(process.env.NODE_ENV !== "production") || __VUE_PROD_DEVTOOLS__) && !def.name) def.name = "VaporElement";
		this._app = this._createApp(this._def);
		this._inheritParentContext();
		if (this._def.configureApp) this._def.configureApp(this._app);
		const hydrate = vaporCustomElementHydrates.get(this.constructor);
		if (this.shadowRoot && hydrate) hydrate(this._root, this._createComponent.bind(this));
		else this._createComponent();
		this._app.mount(this._root);
		if (!this.shadowRoot) this._renderSlots();
	}
	_update() {
		if (!this._app) return;
		const renderEffects = this._instance.renderEffects;
		if (renderEffects) renderEffects.forEach((e) => e.run());
	}
	_unmount() {
		this._app.unmount();
		if (this._instance && this._instance.ce) this._instance.ce = void 0;
		this._app = this._instance = null;
	}
	/**
	* Only called when shadowRoot is false
	*/
	_updateSlotNodes(replacements) {
		this._updateFragmentNodes(this._instance.block, replacements);
	}
	/**
	* Replace slot nodes with their replace content
	* @internal
	*/
	_updateFragmentNodes(block, replacements) {
		if (Array.isArray(block)) {
			block.forEach((item) => this._updateFragmentNodes(item, replacements));
			return;
		}
		if (!isFragment(block)) return;
		const { nodes } = block;
		if (Array.isArray(nodes)) {
			const newNodes = [];
			for (const node of nodes) if (node instanceof HTMLSlotElement) newNodes.push(...replacements.get(node));
			else {
				this._updateFragmentNodes(node, replacements);
				newNodes.push(node);
			}
			block.nodes = newNodes;
		} else if (nodes instanceof HTMLSlotElement) block.nodes = replacements.get(nodes);
		else this._updateFragmentNodes(nodes, replacements);
	}
	_createComponent() {
		this._def.ce = (instance) => {
			this._app._ceComponent = this._instance = instance;
			if (!this.shadowRoot) this._instance.u = [this._renderSlots.bind(this)];
			this._processInstance();
		};
		createComponent(this._def, this._props, void 0, void 0, void 0, this._app._context);
	}
};
//#endregion
//#region packages/runtime-vapor/src/dom/template.ts
let t;
/* @__NO_SIDE_EFFECTS__ */
function template(html, root, ns) {
	let node;
	return () => {
		if (isHydrating$1) {
			const adopted = adoptTemplate(currentHydrationNode, html);
			if (root) adopted.$root = true;
			return adopted;
		}
		if (html[0] !== "<") return /* @__PURE__ */ createTextNode(html);
		if (!node) {
			t = t || document.createElement("template");
			if (ns) {
				const tag = ns === 1 ? "svg" : "math";
				t.innerHTML = `<${tag}>${html}</${tag}>`;
				node = /* @__PURE__ */ _child(/* @__PURE__ */ _child(t.content));
			} else {
				t.innerHTML = html;
				node = /* @__PURE__ */ _child(t.content);
			}
		}
		const ret = node.cloneNode(true);
		if (root) ret.$root = true;
		return ret;
	};
}
//#endregion
//#region packages/runtime-vapor/src/apiCreateIf.ts
function createIf(condition, b1, b2, blockShape, once, index) {
	const _insertionParent = insertionParent;
	const _insertionAnchor = insertionAnchor;
	const _isLastInsertion = isLastInsertion;
	if (!isHydrating$1) resetInsertionState();
	let frag;
	if (once) {
		const ok = condition();
		if (isHydrating$1) locateHydrationNode(decodeIfShape(blockShape, ok) === 2);
		frag = ok ? b1() : b2 ? b2() : [!!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment("if") : /* @__PURE__ */ createTextNode()];
	} else {
		const keyed = index != null;
		frag = isHydrating$1 || !!(process.env.NODE_ENV !== "production") ? new DynamicFragment("if", keyed, false) : new DynamicFragment(void 0, keyed, false);
		renderEffect(() => {
			const ok = condition();
			if (isHydrating$1) locateHydrationNode(decodeIfShape(blockShape, ok) === 2);
			frag.update(ok ? b1 : b2, keyed ? `${index}${ok ? 0 : 1}` : void 0);
		});
	}
	if (!isHydrating$1) {
		if (_insertionParent) insert(frag, _insertionParent, _insertionAnchor);
	} else if (_isLastInsertion) advanceHydrationNode(_insertionParent);
	return frag;
}
function decodeIfShape(shape, ok) {
	return shape >> (ok ? 0 : 2) & 3;
}
//#endregion
//#region packages/runtime-vapor/src/apiCreateFragment.ts
/**
* Create a dynamic fragment keyed by a reactive value for Vapor transitions.
* The fragment is re-rendered when the key changes to trigger enter/leave
* animations.
*
* Example:
* <VaporTransition>
*   <h1 :key="count">{{ count }}</h1>
* </VaporTransition>
*/
function createKeyedFragment(key, render) {
	const _insertionParent = insertionParent;
	const _insertionAnchor = insertionAnchor;
	const _isLastInsertion = isLastInsertion;
	if (!isHydrating$1) resetInsertionState();
	const frag = !!(process.env.NODE_ENV !== "production") ? new DynamicFragment("keyed", true) : new DynamicFragment(void 0, true);
	renderEffect(() => frag.update(render, key()));
	if (!isHydrating$1) {
		if (_insertionParent) insert(frag, _insertionParent, _insertionAnchor);
	} else if (_isLastInsertion) advanceHydrationNode(_insertionParent);
	return frag;
}
//#endregion
//#region packages/runtime-vapor/src/apiCreateFor.ts
const createFor = (src, renderItem, getKey, flags = 0, setup) => {
	const _insertionParent = insertionParent;
	const _insertionAnchor = insertionAnchor;
	const _insertionIndex = insertionIndex;
	const _isLastInsertion = isLastInsertion;
	if (isHydrating$1) locateHydrationNode(true);
	else resetInsertionState();
	let isMounted = false;
	let oldBlocks = [];
	let newBlocks;
	let parent;
	let currentKey;
	let parentAnchor;
	let pendingHydrationAnchor = false;
	if (!isHydrating$1) parentAnchor = !!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment("for") : /* @__PURE__ */ createTextNode();
	const frag = new ForFragment(oldBlocks);
	const instance = currentInstance$1;
	const canUseFastRemove = !!(flags & 1);
	const isComponent = !!(flags & 2);
	const selectors = [];
	const slotOwner = currentSlotOwner;
	if (!!(process.env.NODE_ENV !== "production") && !instance) warn("createFor() can only be used inside setup()");
	const renderList = () => {
		const source = normalizeSource(src());
		const newLength = source.values.length;
		const oldLength = oldBlocks.length;
		newBlocks = new Array(newLength);
		let isFallback = false;
		const prevSub = setActiveSub();
		if (!isMounted) {
			isMounted = true;
			if (isHydrating$1) {
				const hydrationStart = currentHydrationNode;
				let exitHydrationBoundary;
				let nextNode;
				const emptyLocalRange = isComment(hydrationStart, "]") && isComment(hydrationStart.previousSibling, "[");
				const slotFallbackRange = currentEmptyFragment !== void 0 && currentSlotEndAnchor;
				try {
					if (emptyLocalRange && newLength) {
						parentAnchor = markHydrationAnchor(hydrationStart);
						exitHydrationBoundary = enterHydrationBoundary(parentAnchor);
						for (let i = 0; i < newLength; i++) mount(source, i);
						setCurrentHydrationNode(parentAnchor);
					} else {
						for (let i = 0; i < newLength; i++) {
							if (isComment(currentHydrationNode, "]")) {
								nextNode = markHydrationAnchor(currentHydrationNode);
								setCurrentHydrationNode(nextNode);
							} else nextNode = locateNextNode(currentHydrationNode);
							mount(source, i);
							if (nextNode) setCurrentHydrationNode(nextNode);
						}
						if (slotFallbackRange && !isValidBlock(newBlocks)) {
							const anchor = currentHydrationNode !== hydrationStart ? currentHydrationNode : hydrationStart !== currentSlotEndAnchor ? hydrationStart.nextSibling : currentSlotEndAnchor;
							parentAnchor = markHydrationAnchor(!!(process.env.NODE_ENV !== "production") ? /* @__PURE__ */ createComment("for") : /* @__PURE__ */ createTextNode());
							pendingHydrationAnchor = true;
							if (currentHydrationNode === hydrationStart || currentHydrationNode === currentSlotEndAnchor) setCurrentHydrationNode(hydrationStart);
							queuePostFlushCb(() => anchor.parentNode.insertBefore(parentAnchor, anchor));
						} else {
							parentAnchor = markHydrationAnchor(locateHydrationBoundaryClose(currentHydrationNode));
							exitHydrationBoundary = enterHydrationBoundary(parentAnchor);
							if (!!(process.env.NODE_ENV !== "production") && !isComment(parentAnchor, "]")) throw new Error(`v-for fragment anchor node was not found. this is likely a Vue internal bug.`);
							if (_insertionParent && isComment(parentAnchor, "]")) {
								parentAnchor.$idx = _insertionIndex || 0;
								_insertionParent.$llc = parentAnchor;
							}
						}
					}
				} finally {
					exitHydrationBoundary && exitHydrationBoundary();
				}
			} else for (let i = 0; i < newLength; i++) mount(source, i);
		} else {
			parent = parent || parentAnchor.parentNode;
			if (!oldLength) {
				if (frag.fallback && frag.nodes[0].length > 0) remove(frag.nodes[0], parent);
				for (let i = 0; i < newLength; i++) mount(source, i);
			} else if (!newLength) {
				for (const selector of selectors) selector.cleanup();
				const doRemove = !canUseFastRemove;
				for (let i = 0; i < oldLength; i++) unmount(oldBlocks[i], doRemove, false);
				if (canUseFastRemove) {
					parent.textContent = "";
					parent.appendChild(parentAnchor);
				}
				if (frag.fallback) {
					insert(frag.nodes[0] = frag.fallback(), parent, parentAnchor);
					isFallback = true;
				}
			} else if (!getKey) {
				const commonLength = Math.min(newLength, oldLength);
				for (let i = 0; i < commonLength; i++) update(newBlocks[i] = oldBlocks[i], getItem(source, i)[0]);
				for (let i = oldLength; i < newLength; i++) mount(source, i);
				for (let i = newLength; i < oldLength; i++) unmount(oldBlocks[i]);
			} else {
				if (!!(process.env.NODE_ENV !== "production")) {
					const keyToIndexMap = /* @__PURE__ */ new Map();
					for (let i = 0; i < newLength; i++) {
						const key = getKey(...getItem(source, i));
						if (key != null) {
							if (keyToIndexMap.has(key)) warn(`Duplicate keys found during update:`, JSON.stringify(key), `Make sure keys are unique.`);
							keyToIndexMap.set(key, i);
						}
					}
				}
				const commonLength = Math.min(oldLength, newLength);
				const oldKeyIndexPairs = new Array(oldLength);
				const queuedBlocks = new Array(newLength);
				let endOffset = 0;
				let queuedBlocksLength = 0;
				let oldKeyIndexPairsLength = 0;
				while (endOffset < commonLength) {
					const index = newLength - endOffset - 1;
					const item = getItem(source, index);
					const key = getKey(...item);
					const existingBlock = oldBlocks[oldLength - endOffset - 1];
					if (existingBlock.key !== key) break;
					update(existingBlock, ...item);
					newBlocks[index] = existingBlock;
					endOffset++;
				}
				const e1 = commonLength - endOffset;
				const e2 = oldLength - endOffset;
				const e3 = newLength - endOffset;
				for (let i = 0; i < e1; i++) {
					const currentItem = getItem(source, i);
					const currentKey = getKey(...currentItem);
					const oldBlock = oldBlocks[i];
					const oldKey = oldBlock.key;
					if (oldKey === currentKey) update(newBlocks[i] = oldBlock, currentItem[0]);
					else {
						queuedBlocks[queuedBlocksLength++] = [
							i,
							currentItem,
							currentKey
						];
						oldKeyIndexPairs[oldKeyIndexPairsLength++] = [oldKey, i];
					}
				}
				for (let i = e1; i < e2; i++) oldKeyIndexPairs[oldKeyIndexPairsLength++] = [oldBlocks[i].key, i];
				for (let i = e1; i < e3; i++) {
					const blockItem = getItem(source, i);
					const blockKey = getKey(...blockItem);
					queuedBlocks[queuedBlocksLength++] = [
						i,
						blockItem,
						blockKey
					];
				}
				queuedBlocks.length = queuedBlocksLength;
				oldKeyIndexPairs.length = oldKeyIndexPairsLength;
				const oldKeyIndexMap = new Map(oldKeyIndexPairs);
				const opers = new Array(queuedBlocks.length);
				let mountCounter = 0;
				let opersLength = 0;
				for (let i = queuedBlocks.length - 1; i >= 0; i--) {
					const [index, item, key] = queuedBlocks[i];
					const oldIndex = oldKeyIndexMap.get(key);
					if (oldIndex !== void 0) {
						oldKeyIndexMap.delete(key);
						const reusedBlock = newBlocks[index] = oldBlocks[oldIndex];
						update(reusedBlock, ...item);
						opers[opersLength++] = {
							index,
							block: reusedBlock
						};
					} else {
						mountCounter++;
						opers[opersLength++] = {
							source,
							index,
							item,
							key
						};
					}
				}
				const useFastRemove = mountCounter === newLength;
				for (const leftoverIndex of oldKeyIndexMap.values()) unmount(oldBlocks[leftoverIndex], !(useFastRemove && canUseFastRemove), !useFastRemove);
				if (useFastRemove) {
					for (const selector of selectors) selector.cleanup();
					if (canUseFastRemove) {
						parent.textContent = "";
						parent.appendChild(parentAnchor);
					}
				}
				if (opers.length === mountCounter) for (const { source, index, item, key } of opers) mount(source, index, index < newLength - 1 ? normalizeAnchor(newBlocks[index + 1].nodes) : parentAnchor, item, key);
				else if (opers.length) {
					let anchor = oldBlocks[0];
					let blocksTail;
					for (let i = 0; i < oldLength; i++) {
						const block = oldBlocks[i];
						if (oldKeyIndexMap.has(block.key)) continue;
						block.prevAnchor = anchor;
						anchor = oldBlocks[i + 1];
						if (blocksTail !== void 0) {
							blocksTail.next = block;
							block.prev = blocksTail;
						}
						blocksTail = block;
					}
					for (const action of opers) {
						const { index } = action;
						if (index < newLength - 1) {
							const nextBlock = newBlocks[index + 1];
							let anchorNode = normalizeAnchor(nextBlock.prevAnchor.nodes);
							if (!anchorNode.parentNode) anchorNode = normalizeAnchor(nextBlock.nodes);
							if ("source" in action) {
								const { item, key } = action;
								moveLink(mount(source, index, anchorNode, item, key), nextBlock.prev, nextBlock);
							} else if (action.block.next !== nextBlock) {
								insert(action.block, parent, anchorNode);
								moveLink(action.block, nextBlock.prev, nextBlock);
							}
						} else if ("source" in action) {
							const { item, key } = action;
							const block = mount(source, index, parentAnchor, item, key);
							moveLink(block, blocksTail);
							blocksTail = block;
						} else if (action.block.next !== void 0) {
							let anchorNode = anchor ? normalizeAnchor(anchor.nodes) : parentAnchor;
							if (!anchorNode.parentNode) anchorNode = parentAnchor;
							insert(action.block, parent, anchorNode);
							moveLink(action.block, blocksTail);
							blocksTail = action.block;
						}
					}
					for (const block of newBlocks) block.prevAnchor = block.next = block.prev = void 0;
				}
			}
		}
		if (!isFallback) {
			frag.nodes = [oldBlocks = newBlocks];
			if (parentAnchor) frag.nodes.push(parentAnchor);
		} else oldBlocks = [];
		if (isMounted && frag.onUpdated) frag.onUpdated.forEach((m) => m());
		setActiveSub(prevSub);
	};
	const needKey = renderItem.length > 1;
	const needIndex = renderItem.length > 2;
	const mount = (source, idx, anchor = parentAnchor, [item, key, index] = getItem(source, idx), key2 = getKey && getKey(item, key, index)) => {
		const itemRef = shallowRef$1(item);
		const keyRef = needKey ? shallowRef$1(key) : void 0;
		const indexRef = needIndex ? shallowRef$1(index) : void 0;
		currentKey = key2;
		let nodes;
		let scope;
		if (isComponent) nodes = renderItem(itemRef, keyRef, indexRef);
		else {
			scope = new EffectScope$1();
			nodes = scope.run(() => renderItem(itemRef, keyRef, indexRef));
		}
		const block = newBlocks[idx] = new ForBlock(nodes, scope, itemRef, keyRef, indexRef, key2);
		if (frag.$transition) applyTransitionHooks(block.nodes, frag.$transition);
		if (parent) insert(block.nodes, parent, anchor);
		return block;
	};
	const update = ({ itemRef, keyRef, indexRef }, newItem, newKey, newIndex) => {
		if (newItem !== itemRef.value) itemRef.value = newItem;
		if (keyRef && newKey !== void 0 && newKey !== keyRef.value) keyRef.value = newKey;
		if (indexRef && newIndex !== void 0 && newIndex !== indexRef.value) indexRef.value = newIndex;
	};
	const unmount = (block, doRemove = true, doDeregister = true) => {
		if (!isComponent) block.scope.stop();
		if (doRemove) remove(block.nodes, parent);
		if (doDeregister) for (const selector of selectors) selector.deregister(block.key);
	};
	if (setup) setup({ createSelector });
	if (flags & 4) renderList();
	else renderEffect(() => {
		if (!isMounted) return renderList();
		const prevOwner = setCurrentSlotOwner(slotOwner);
		try {
			renderList();
		} finally {
			setCurrentSlotOwner(prevOwner);
		}
	});
	if (!isHydrating$1) {
		if (_insertionParent) insert(frag, _insertionParent, _insertionAnchor);
	} else if (!pendingHydrationAnchor) advanceHydrationNode(_isLastInsertion ? _insertionParent : parentAnchor);
	return frag;
	function createSelector(source) {
		let operMap = /* @__PURE__ */ new Map();
		let activeKey = source();
		let activeOpers;
		watch$1(source, (newValue) => {
			if (activeOpers !== void 0) for (const oper of activeOpers) oper();
			queuePostFlushCb(() => {
				activeKey = newValue;
				activeOpers = operMap.get(newValue);
				if (activeOpers !== void 0) for (const oper of activeOpers) oper();
			});
		});
		selectors.push({
			deregister,
			cleanup
		});
		return register;
		function cleanup() {
			operMap = /* @__PURE__ */ new Map();
			activeOpers = void 0;
		}
		function register(oper) {
			oper();
			let opers = operMap.get(currentKey);
			if (opers !== void 0) opers.push(oper);
			else {
				opers = [oper];
				operMap.set(currentKey, opers);
				if (currentKey === activeKey) activeOpers = opers;
			}
		}
		function deregister(key) {
			operMap.delete(key);
			if (key === activeKey) activeOpers = void 0;
		}
	}
};
function moveLink(block, newPrev, newNext) {
	const { prev: oldPrev, next: oldNext } = block;
	if (oldPrev) oldPrev.next = oldNext;
	if (oldNext) {
		oldNext.prev = oldPrev;
		if (block.prevAnchor !== block) oldNext.prevAnchor = block.prevAnchor;
	}
	if (newPrev) newPrev.next = block;
	if (newNext) newNext.prev = block;
	block.prev = newPrev;
	block.next = newNext;
	block.prevAnchor = block;
}
function createForSlots(rawSource, getSlot) {
	const source = normalizeSource(rawSource);
	const sourceLength = source.values.length;
	const slots = new Array(sourceLength);
	for (let i = 0; i < sourceLength; i++) slots[i] = getSlot(...getItem(source, i));
	return slots;
}
function normalizeSource(source) {
	let values = source;
	let needsWrap = false;
	let isReadonlySource = false;
	let keys;
	if (isArray(source)) {
		if (isReactive(source)) {
			needsWrap = !isShallow(source);
			values = shallowReadArray(source);
			isReadonlySource = isReadonly(source);
		}
	} else if (isString(source)) values = source.split("");
	else if (typeof source === "number") {
		if (!!(process.env.NODE_ENV !== "production") && !Number.isInteger(source)) warn(`The v-for range expect an integer value but got ${source}.`);
		values = new Array(source);
		for (let i = 0; i < source; i++) values[i] = i + 1;
	} else if (isObject(source)) if (source[Symbol.iterator]) values = Array.from(source);
	else {
		keys = Object.keys(source);
		values = new Array(keys.length);
		for (let i = 0, l = keys.length; i < l; i++) values[i] = source[keys[i]];
	}
	else values = [];
	return {
		values,
		needsWrap,
		isReadonlySource,
		keys
	};
}
function getItem({ keys, values, needsWrap, isReadonlySource }, idx) {
	const value = needsWrap ? isReadonlySource ? toReadonly(toReactive(values[idx])) : toReactive(values[idx]) : values[idx];
	if (keys) return [
		value,
		keys[idx],
		idx
	];
	else return [
		value,
		idx,
		void 0
	];
}
function normalizeAnchor(node) {
	if (node instanceof Node) return node;
	else if (isArray(node)) return normalizeAnchor(node[0]);
	else if (isVaporComponent(node)) return normalizeAnchor(node.block);
	else return normalizeAnchor(node.nodes);
}
function getRestElement(val, keys) {
	const res = {};
	for (const key in val) if (!keys.includes(key)) res[key] = val[key];
	return res;
}
function getDefaultValue(val, defaultVal) {
	return val === void 0 ? defaultVal : val;
}
function isForBlock(block) {
	return block instanceof ForBlock;
}
//#endregion
//#region packages/runtime-vapor/src/apiTemplateRef.ts
function ensureCleanup(el) {
	let cleanupRef = refCleanups.get(el);
	if (!cleanupRef) {
		refCleanups.set(el, cleanupRef = { fn: NOOP });
		onScopeDispose$1(() => {
			invalidatePendingRef(el);
			cleanupRef.fn();
			refCleanups.delete(el);
		});
	}
	return cleanupRef;
}
function createTemplateRefSetter() {
	const instance = currentInstance$1;
	const oldRefMap = /* @__PURE__ */ new WeakMap();
	const setRefMap = /* @__PURE__ */ new WeakMap();
	return (el, ref, refFor, refKey) => {
		if (isDynamicFragment(el) || isVaporComponent(el) && isAsyncWrapper(el)) {
			const frag = isDynamicFragment(el) ? el : el.block;
			const doSet = () => {
				if (isVaporComponent(el) && el.isDeactivated) return;
				oldRefMap.set(el, setRef$1(instance, el, ref, oldRefMap.get(el), refFor, refKey));
			};
			const prevSet = setRefMap.get(frag);
			if (prevSet && frag.onUpdated) remove$1(frag.onUpdated, prevSet);
			(frag.onUpdated || (frag.onUpdated = [])).push(doSet);
			setRefMap.set(frag, doSet);
		}
		const oldRef = setRef$1(instance, el, ref, oldRefMap.get(el), refFor, refKey);
		oldRefMap.set(el, oldRef);
		return oldRef;
	};
}
/**
* Function for handling a template ref
*/
function setRef$1(instance, el, ref, oldRef, refFor = false, refKey) {
	if (!instance || instance.isUnmounted) return;
	const setupState = !!(process.env.NODE_ENV !== "production") ? instance.setupState || {} : null;
	const refValue = getRefValue(el);
	if (isInteropEnabled) {
		const target = isFragment(el) && el.setRef ? el : refValue && isFragment(refValue) && refValue.setRef ? refValue : null;
		if (target) {
			target.setRef(instance, ref, refFor, refKey);
			return ref;
		}
	}
	const refs = instance.refs === EMPTY_OBJ ? instance.refs = {} : instance.refs;
	const canSetSetupRef = !!(process.env.NODE_ENV !== "production") ? createCanSetSetupRefChecker(setupState, refs) : NO;
	const canSetRef = (ref, key) => {
		if (!!(process.env.NODE_ENV !== "production") && knownTemplateRefs.has(ref)) return false;
		if (key && isTemplateRefKey(refs, key)) return false;
		return true;
	};
	if (oldRef != null && oldRef !== ref) {
		invalidatePendingRef(el);
		if (isString(oldRef)) {
			refs[oldRef] = null;
			if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(oldRef)) setupState[oldRef] = null;
		} else if (isRef$1(oldRef)) {
			if (canSetRef(oldRef)) oldRef.value = null;
		} else if (isFunction(oldRef) && isDynamicFragment(el)) callWithErrorHandling(oldRef, instance, 12, [null, refs]);
	} else if (oldRef != null && isDynamicFragment(el)) {
		if (isFunction(oldRef)) callWithErrorHandling(oldRef, instance, 12, [null, refs]);
		else if (refFor) unsetRef(el);
	}
	if (ref == null) return ref;
	if (isFunction(ref)) {
		const invokeRefSetter = (value) => {
			callWithErrorHandling(ref, instance, 12, [value, refs]);
		};
		invokeRefSetter(refValue);
		ensureCleanup(el).fn = () => invokeRefSetter(null);
	} else {
		const _isString = isString(ref);
		const _isRef = isRef$1(ref);
		let existing;
		if (_isString || _isRef) {
			const doSet = () => {
				if (refFor) {
					if (refValue == null) return;
					existing = _isString ? !!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref) ? setupState[ref] : refs[ref] : canSetRef(ref) || !refKey ? ref.value : refs[refKey];
					if (!isArray(existing)) {
						existing = [refValue];
						if (_isString) {
							refs[ref] = existing;
							if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref)) {
								setupState[ref] = refs[ref];
								existing = setupState[ref];
							}
						} else {
							if (canSetRef(ref, refKey)) ref.value = existing;
							if (refKey) refs[refKey] = existing;
						}
					} else if (!existing.includes(refValue)) existing.push(refValue);
				} else if (_isString) {
					refs[ref] = refValue;
					if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref)) setupState[ref] = refValue;
				} else if (_isRef) {
					if (canSetRef(ref, refKey)) ref.value = refValue;
					if (refKey) refs[refKey] = refValue;
				} else if (!!(process.env.NODE_ENV !== "production")) warn("Invalid template ref type:", ref, `(${typeof ref})`);
			};
			const cleanup = ensureCleanup(el);
			cleanup.fn = () => {
				if (refFor) {
					if (isArray(existing)) remove$1(existing, refValue);
				} else if (_isString) {
					refs[ref] = null;
					if (!!(process.env.NODE_ENV !== "production") && canSetSetupRef(ref)) setupState[ref] = null;
				} else if (_isRef) {
					if (canSetRef(ref, refKey)) ref.value = null;
					if (refKey) refs[refKey] = null;
				}
			};
			invalidatePendingRef(el);
			if (refValue != null) {
				const job = () => {
					doSet();
					if (cleanup.job === job) cleanup.job = void 0;
				};
				cleanup.job = job;
				queuePostFlushCb(job, -1);
			} else doSet();
		} else if (!!(process.env.NODE_ENV !== "production")) warn("Invalid template ref type:", ref, `(${typeof ref})`);
	}
	return ref;
}
const getRefValue = (el) => {
	if (isVaporComponent(el)) {
		if (isAsyncWrapper(el)) {
			if (!el.type.__asyncResolved) return null;
			return getRefValue(el.block.nodes);
		}
		return getExposed(el) || el;
	} else if (isTeleportFragment(el)) return null;
	else if (isDynamicFragment(el)) {
		if (isArray(el.nodes)) return null;
		return getRefValue(el.nodes);
	}
	return el;
};
//#endregion
//#region packages/runtime-vapor/src/helpers/useCssVars.ts
function useVaporCssVars(getter) {
	const instance = currentInstance;
	baseUseCssVars(instance, () => resolveParentNode(instance.block), getter, (vars) => setVars(instance, vars));
}
function resolveParentNode(block) {
	if (block instanceof Node) return block.parentNode;
	else if (isArray(block)) return resolveParentNode(block[block.length - 1]);
	else if (isVaporComponent(block)) return resolveParentNode(block.block);
	else return resolveParentNode(block.anchor || block.nodes);
}
function setVars(instance, vars) {
	if (instance.ce) setVarsOnNode(instance.ce, vars);
	else setVarsOnBlock(instance.block, vars);
}
function setVarsOnBlock(block, vars) {
	if (block instanceof Node) setVarsOnNode(block, vars);
	else if (isArray(block)) block.forEach((child) => setVarsOnBlock(child, vars));
	else if (isVaporComponent(block)) setVarsOnBlock(block.block, vars);
	else if (isTeleportFragment(block)) return;
	else setVarsOnBlock(block.nodes, vars);
}
//#endregion
//#region packages/runtime-vapor/src/apiCreateDynamicComponent.ts
function createDynamicComponent(getter, rawProps, rawSlots, isSingleRoot, once) {
	const _insertionParent = insertionParent;
	const _insertionAnchor = insertionAnchor;
	const _isLastInsertion = isLastInsertion;
	if (!isHydrating$1) resetInsertionState();
	const frag = isHydrating$1 || !!(process.env.NODE_ENV !== "production") ? new DynamicFragment("dynamic-component") : new DynamicFragment();
	const scopeOwner = getScopeOwner();
	const renderFn = () => {
		const value = getter();
		const appContext = currentInstance && currentInstance.appContext || emptyContext;
		frag.update(() => {
			if (isBlock(value)) return value;
			if (isInteropEnabled && appContext.vapor && isVNode(value)) {
				if (isKeepAlive(currentInstance)) {
					const frag = currentInstance.ctx.getCachedComponent(value.type, value.key);
					if (frag) return frag;
				}
				const frag = appContext.vapor.vdomMountVNode(value, currentInstance);
				if (isHydrating$1) {
					locateHydrationNode(shouldConsumeFragmentStart(value));
					frag.hydrate();
					if (_isLastInsertion) advanceHydrationNode(_insertionParent);
				}
				return frag;
			}
			return createComponentWithFallback(withScopeOwner(scopeOwner, () => resolveDynamicComponent(value)), rawProps, rawSlots, isSingleRoot, once, appContext);
		}, value);
	};
	if (once) renderFn();
	else renderEffect(renderFn);
	if (!isHydrating$1) {
		if (_insertionParent) insert(frag, _insertionParent, _insertionAnchor);
	} else if (_isLastInsertion) advanceHydrationNode(_insertionParent);
	return frag;
}
function withScopeOwner(owner, fn) {
	const prev = setCurrentRenderingInstance(owner);
	try {
		return fn();
	} finally {
		setCurrentRenderingInstance(prev);
	}
}
function shouldConsumeFragmentStart(vnode) {
	if (vnode.type === Fragment) return false;
	if (vnode.shapeFlag & 6) {
		const type = vnode.type;
		return !!type.__vapor && !type.__multiRoot;
	}
	return true;
}
//#endregion
//#region packages/runtime-vapor/src/apiSetupHelpers.ts
function withAsyncContext(getAwaitable) {
	const instance = currentInstance;
	if (isHydrating$1 && instance && instance.vapor) {
		const hydrationNode = currentHydrationNode;
		instance.restoreAsyncContext = () => enterHydration(hydrationNode);
	}
	return withAsyncContext$1(getAwaitable);
}
//#endregion
//#region packages/runtime-vapor/src/directives/vModel.ts
function ensureMounted(cb) {
	if (currentInstance.isMounted) cb();
	else onMounted(cb);
}
const applyTextModel = (el, get, set, { trim, number, lazy } = {}) => {
	vModelTextInit(el, trim, number, lazy, set);
	ensureMounted(() => {
		let value;
		renderEffect(() => {
			vModelTextUpdate(el, value, value = get(), trim, number, lazy);
		});
	});
};
const applyCheckboxModel = (el, get, set) => {
	vModelCheckboxInit(el, set);
	ensureMounted(() => {
		let value;
		renderEffect(() => {
			vModelCheckboxUpdate(el, value, traverse(value = get()));
		});
	});
};
const applyRadioModel = (el, get, set) => {
	addEventListener(el, "change", () => set(vModelGetValue(el)));
	ensureMounted(() => {
		let value;
		renderEffect(() => {
			if (value !== (value = get())) el.checked = looseEqual(value, vModelGetValue(el));
		});
	});
};
const applySelectModel = (el, get, set, modifiers) => {
	vModelSelectInit(el, get(), modifiers && modifiers.number, set);
	ensureMounted(() => {
		renderEffect(() => vModelSetSelected(el, traverse(get())));
	});
};
const applyDynamicModel = (el, get, set, modifiers) => {
	let apply = applyTextModel;
	if (el.tagName === "SELECT") apply = applySelectModel;
	else if (el.tagName === "TEXTAREA") apply = applyTextModel;
	else if (el.type === "checkbox") apply = applyCheckboxModel;
	else if (el.type === "radio") apply = applyRadioModel;
	apply(el, get, set, modifiers);
};
//#endregion
//#region packages/runtime-vapor/src/directives/custom.ts
function withVaporDirectives(node, dirs) {
	const element = isVaporComponent(node) ? getRootElement(node.block) : node;
	if (!element) {
		if (!!(process.env.NODE_ENV !== "production")) warn("Runtime directive used on component with non-element root node. The directives will not function as intended.");
		return;
	}
	for (const [dir, value, argument, modifiers] of dirs) if (dir) {
		const ret = dir(element, value, argument, modifiers);
		if (ret) onScopeDispose(ret);
	}
}
//#endregion
//#region packages/runtime-vapor/src/components/TransitionGroup.ts
const positionMap = /* @__PURE__ */ new WeakMap();
const newPositionMap = /* @__PURE__ */ new WeakMap();
const decorate = (t) => {
	delete t.props.mode;
	return t;
};
const VaporTransitionGroup = /* @__PURE__ */ decorate(/* @__PURE__ */ defineVaporComponent({
	name: "VaporTransitionGroup",
	props: /* @__PURE__ */ extend({}, TransitionPropsValidators, {
		tag: String,
		moveClass: String
	}),
	setup(props, { slots, expose }) {
		expose();
		ensureTransitionHooksRegistered();
		const instance = currentInstance;
		const state = useTransitionState();
		let cssTransitionProps = resolveTransitionProps(props);
		const propsProxy = new Proxy({}, { get(_, key) {
			return cssTransitionProps[key];
		} });
		renderEffect(() => {
			cssTransitionProps = resolveTransitionProps(props);
		});
		let prevChildren;
		let slottedBlock = [];
		onBeforeUpdate(() => {
			prevChildren = [];
			const children = getTransitionBlocks(slottedBlock);
			for (let i = 0; i < children.length; i++) {
				const child = children[i];
				const el = isValidTransitionBlock(child) && child.$transition ? getTransitionElement(child) : void 0;
				if (el) {
					prevChildren.push(child);
					child.$transition.disabled = true;
					positionMap.set(child, el.getBoundingClientRect());
				}
			}
		});
		onUpdated(() => {
			if (!prevChildren.length) return;
			const moveClass = props.moveClass || `${props.name || "v"}-move`;
			const firstChild = getFirstConnectedChild(prevChildren);
			if (!firstChild || !hasCSSTransform(firstChild, firstChild.parentNode, moveClass)) {
				prevChildren = [];
				return;
			}
			prevChildren.forEach(callPendingCbs);
			prevChildren.forEach((child) => {
				child.$transition.disabled = false;
				recordPosition(child);
			});
			const movedChildren = prevChildren.filter(applyTranslation);
			forceReflow();
			movedChildren.forEach((c) => handleMovedChildren(getTransitionElement(c), moveClass));
			prevChildren = [];
		});
		const frag = new DynamicFragment("transition-group");
		let currentTag;
		let currentSlot;
		let isMounted = false;
		renderEffect(() => {
			const tag = props.tag;
			const slot = slots.default;
			if (isMounted && tag === currentTag && slot === currentSlot) return;
			const container = tag ? /* @__PURE__ */ createElement(tag) : void 0;
			let block = slottedBlock;
			frag.update(() => {
				block = slot && slot() || [];
				applyGroupTransitionHooks(block, propsProxy, state, instance);
				if (container) {
					insert(block, container);
					return container;
				}
				return block;
			});
			slottedBlock = block;
			currentTag = tag;
			currentSlot = slot;
			isMounted = true;
		});
		return frag;
	}
}));
function applyGroupTransitionHooks(block, props, state, instance) {
	const fragments = [];
	const children = getTransitionBlocks(block, (frag) => fragments.push(frag));
	for (let i = 0; i < children.length; i++) {
		const child = children[i];
		if (isValidTransitionBlock(child)) {
			if (child.$key != null) setTransitionHooks$1(child, resolveTransitionHooks(child, props, state, instance));
			else if (!!(process.env.NODE_ENV !== "production")) warn(`<transition-group> children must be keyed`);
		}
	}
	fragments.forEach((frag) => {
		const hooks = resolveTransitionHooks(frag, props, state, instance);
		hooks.applyGroup = applyGroupTransitionHooks;
		frag.$transition = hooks;
	});
}
function inheritKey(children, key) {
	if (key === void 0 || children.length === 0) return;
	for (let i = 0; i < children.length; i++) {
		const child = children[i];
		child.$key = String(key) + String(child.$key != null ? child.$key : i);
	}
}
function getTransitionBlocks(block, onFragment) {
	let children = [];
	if (block instanceof Element) children.push(block);
	else if (isVaporComponent(block)) {
		const blocks = getTransitionBlocks(block.block, onFragment);
		inheritKey(blocks, block.$key);
		children.push(...blocks);
	} else if (isArray(block)) for (let i = 0; i < block.length; i++) {
		const b = block[i];
		const blocks = getTransitionBlocks(b, onFragment);
		if (isForBlock(b)) blocks.forEach((block) => block.$key = b.key);
		children.push(...blocks);
	}
	else if (isFragment(block)) if (isInteropEnabled && block.vnode) children.push(block);
	else {
		if (onFragment) onFragment(block);
		const blocks = getTransitionBlocks(block.nodes, onFragment);
		inheritKey(blocks, block.$key);
		children.push(...blocks);
	}
	return children;
}
function isValidTransitionBlock(block) {
	return !!(block instanceof Element || isFragment(block) && block.vnode);
}
function getTransitionElement(block) {
	if (block instanceof Element) return block;
	if (isInteropEnabled && isFragment(block) && block.vnode) return getTransitionElementFromVNode(block.vnode);
}
function recordPosition(c) {
	const el = getTransitionElement(c);
	if (el) newPositionMap.set(c, el.getBoundingClientRect());
}
function applyTranslation(c) {
	const el = getTransitionElement(c);
	if (el && baseApplyTranslation(positionMap.get(c), newPositionMap.get(c), el)) return c;
}
function getFirstConnectedChild(children) {
	for (let i = 0; i < children.length; i++) {
		const child = children[i];
		const el = getTransitionElement(child);
		if (el && el.isConnected) return el;
	}
}
//#endregion
export { DynamicFragment, VaporElement, VaporFragment, VaporKeepAlive, VaporTeleport, VaporTransition, VaporTransitionGroup, applyCheckboxModel, applyDynamicModel, applyRadioModel, applySelectModel, applyTextModel, applyVShow, child, createComponent, createComponentWithFallback, createDynamicComponent, createFor, createForSlots, createIf, createInvoker, createKeyedFragment, createPlainElement, createSlot, createTemplateRefSetter, createTextNode, createVaporApp, createVaporSSRApp, defineVaporAsyncComponent, defineVaporComponent, defineVaporCustomElement, defineVaporSSRCustomElement, delegate, delegateEvents, getDefaultValue, getRestElement, insert, isFragment, isVaporComponent, next, nthChild, on, prepend, remove, renderEffect, setAttr, setBlockHtml, setBlockKey, setBlockText, setClass, setDOMProp, setDynamicEvents, setDynamicProps, setElementText, setHtml, setInsertionState, setProp, setStyle, setText, setValue, template, txt, useVaporCssVars, vaporInteropPlugin, withAsyncContext, withVaporCtx, withVaporDirectives };
